// @aposin/ng-aquila/schematics/ng-update/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateToV11 = exports.upgradeData = void 0;
const schematics_1 = require("@angular/cdk/schematics");
const data_1 = require("./data");
const misc_template_1 = require("./migrations/misc-template");
/** Upgrade data that will be used for the Angular Material ng-update schematic. */
exports.upgradeData = {
    attributeSelectors: data_1.attributeSelectors,
    classNames: data_1.classNames,
    constructorChecks: data_1.constructorChecks,
    cssSelectors: data_1.cssSelectors,
    elementSelectors: data_1.elementSelectors,
    inputNames: data_1.inputNames,
    methodCallChecks: data_1.methodCallChecks,
    outputNames: data_1.outputNames,
    propertyNames: data_1.propertyNames,
    symbolRemoval: data_1.symbolRemoval,
};
const customMigrations = [misc_template_1.MiscTemplateMigration];
/** Entry point for the migration schematics with target of Angular CDK 12.0.0 */
function updateToV11() {
    return (0, schematics_1.createMigrationSchematicRule)(schematics_1.TargetVersion.V11, customMigrations, exports.upgradeData, v11MarginsWarning);
}
exports.updateToV11 = updateToV11;
/** Function that will be called when the migration completed. */
function onMigrationComplete(context, targetVersion, hasFailures) {
    context.logger.info('');
    context.logger.info(`  ✓  Updated @aposin/ng-aquila to ${targetVersion}`);
    context.logger.info('');
    if (hasFailures) {
        context.logger.warn('  ⚠  Some issues were detected but could not be fixed automatically. Please check the output above and fix these issues manually.');
    }
}
function v11MarginsWarning(context, targetVersion, hasFailures) {
    // call the generic complete function for all schematics
    onMigrationComplete(context, targetVersion, hasFailures);
    // this major we want to throw this additional warning
    context.logger.warn(`
    ⚠ In ng-aquila v11 some components got their default outer margins removed please check your applications for visual changes.
    You can find more details at:
    https://allianz.github.io/ng-aquila/guides/releases
    `);
}
