// @aposin/ng-aquila/esm2020/utils/viewport-service.mjs
import { Injectable } from '@angular/core';
import { asyncScheduler, fromEvent, Subject } from 'rxjs';
import { map, startWith, takeUntil, throttleTime } from 'rxjs/operators';
import * as i0 from "@angular/core";
/** Available breakpoints to subscribe to. */
export var NxBreakpoints;
(function (NxBreakpoints) {
    /** Min size for BREAKPOINT_MOBILE */
    NxBreakpoints[NxBreakpoints["BREAKPOINT_XSMALL"] = 0] = "BREAKPOINT_XSMALL";
    NxBreakpoints[NxBreakpoints["BREAKPOINT_SMALL"] = 320] = "BREAKPOINT_SMALL";
    /** Min size for BREAKPOINT_TABLET */
    NxBreakpoints[NxBreakpoints["BREAKPOINT_MEDIUM"] = 704] = "BREAKPOINT_MEDIUM";
    /** Min size for BREAKPOINT_DESKTOP */
    NxBreakpoints[NxBreakpoints["BREAKPOINT_LARGE"] = 992] = "BREAKPOINT_LARGE";
    NxBreakpoints[NxBreakpoints["BREAKPOINT_XLARGE"] = 1280] = "BREAKPOINT_XLARGE";
    NxBreakpoints[NxBreakpoints["BREAKPOINT_2XLARGE"] = 1472] = "BREAKPOINT_2XLARGE";
    NxBreakpoints[NxBreakpoints["BREAKPOINT_3XLARGE"] = 1760] = "BREAKPOINT_3XLARGE";
})(NxBreakpoints || (NxBreakpoints = {}));
const DEFAULT_THROTTLE_TIME = 200;
/** Service subscribing to window resize events and providing breakpoint matching functions. */
export class NxViewportService {
    constructor() {
        this._destroyed = new Subject();
        this.initialViewportWidth = 0;
        this.viewportChange$ = fromEvent(window, 'resize').pipe(map(() => window.innerWidth), takeUntil(this._destroyed));
    }
    /** Returns whether the current viewport width is greater than or equal (>=) to minSize. */
    min(minSize, throttleTimeMs = DEFAULT_THROTTLE_TIME) {
        return this.viewportChange$.pipe(startWith(window.innerWidth), throttleTime(throttleTimeMs, asyncScheduler, { trailing: true }), map(windowInnerWidth => windowInnerWidth >= minSize));
    }
    /** Returns whether the current viewport width is lower (<) than maxSize. */
    max(maxSize, throttleTimeMs = DEFAULT_THROTTLE_TIME) {
        return this.viewportChange$.pipe(startWith(window.innerWidth), throttleTime(throttleTimeMs, asyncScheduler, { trailing: true }), map(windowInnerWidth => windowInnerWidth < maxSize));
    }
    /** Returns whether the current viewport width is greater than or equal (>=) to minSize and lower (<) than maxSize. */
    between(minSize, maxSize, throttleTimeMs = DEFAULT_THROTTLE_TIME) {
        return this.viewportChange$.pipe(startWith(window.innerWidth), throttleTime(throttleTimeMs, asyncScheduler, { trailing: true }), map(windowInnerWidth => windowInnerWidth >= minSize && windowInnerWidth < maxSize));
    }
    ngOnDestroy() {
        this._destroyed.next();
        this._destroyed.complete();
    }
}
NxViewportService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxViewportService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
NxViewportService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxViewportService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxViewportService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return []; } });
