// @aposin/ng-aquila/esm2020/tabs/tab-body.mjs
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { ChangeDetectionStrategy, Component, Input, ViewChild, ViewContainerRef } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "./tab-group-base";
import * as i2 from "@angular/cdk/a11y";
/** @docs-private */
export class NxTabBodyComponent {
    constructor(_tabGroup, _focusMonitor, _elementRef) {
        this._tabGroup = _tabGroup;
        this._focusMonitor = _focusMonitor;
        this._elementRef = _elementRef;
        this._active = false;
        this._destroyed = new Subject();
        this._focusMonitor.monitor(this._elementRef);
    }
    set active(value) {
        this._active = coerceBooleanProperty(value);
    }
    get active() {
        return this._active;
    }
    ngOnInit() {
        this._tabGroup._appearanceChange.pipe(takeUntil(this._destroyed)).subscribe(() => {
            this.detach();
        });
    }
    ngOnChanges(changes) {
        if ('active' in changes) {
            const change = changes.active;
            if (change.currentValue) {
                this.attach();
            }
            else {
                this.detach();
            }
        }
    }
    ngOnDestroy() {
        this._destroyed.next();
        this._destroyed.complete();
        this._focusMonitor.stopMonitoring(this._elementRef);
    }
    attach() {
        this._outlet.insert(this.tab.contentViewRef);
    }
    detach() {
        const index = this._outlet.indexOf(this.tab.contentViewRef);
        if (index !== -1) {
            this._outlet.detach(index);
        }
    }
}
NxTabBodyComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTabBodyComponent, deps: [{ token: i1.NxTabGroupBase }, { token: i2.FocusMonitor }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
NxTabBodyComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxTabBodyComponent, selector: "nx-tab-body", inputs: { tab: "tab", active: "active" }, viewQueries: [{ propertyName: "_outlet", first: true, predicate: ["outlet"], descendants: true, read: ViewContainerRef, static: true }], usesOnChanges: true, ngImport: i0, template: "<ng-container #outlet></ng-container>\n", styles: [":host{display:block}:host:focus{outline:none}:host.cdk-keyboard-focused{box-shadow:var(--focus-box-shadow);border-radius:4px}@media screen and (-ms-high-contrast: active),(forced-colors: active){:host.cdk-keyboard-focused{box-shadow:0 0 0 2px background,0 0 0 6px windowText;outline:4px solid CanvasText;outline-offset:2px}}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTabBodyComponent, decorators: [{
            type: Component,
            args: [{ selector: 'nx-tab-body', changeDetection: ChangeDetectionStrategy.OnPush, template: "<ng-container #outlet></ng-container>\n", styles: [":host{display:block}:host:focus{outline:none}:host.cdk-keyboard-focused{box-shadow:var(--focus-box-shadow);border-radius:4px}@media screen and (-ms-high-contrast: active),(forced-colors: active){:host.cdk-keyboard-focused{box-shadow:0 0 0 2px background,0 0 0 6px windowText;outline:4px solid CanvasText;outline-offset:2px}}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.NxTabGroupBase }, { type: i2.FocusMonitor }, { type: i0.ElementRef }]; }, propDecorators: { _outlet: [{
                type: ViewChild,
                args: ['outlet', { static: true, read: ViewContainerRef }]
            }], tab: [{
                type: Input
            }], active: [{
                type: Input
            }] } });
