// @aposin/ng-aquila/esm2020/tabs/tab-header-outlet.mjs
import { Component, Input, ViewChild, ViewContainerRef } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "./tab-group-base";
/** @docs-private */
export class NxTabHeaderOutletComponent {
    constructor(_tabGroup) {
        this._tabGroup = _tabGroup;
        this._destroyed = new Subject();
    }
    ngOnInit() {
        this._tabGroup._appearanceChange.pipe(takeUntil(this._destroyed)).subscribe(() => {
            this.detach();
        });
        this.attach();
    }
    attach() {
        this._outlet.insert(this.content);
    }
    detach() {
        const index = this._outlet.indexOf(this.content);
        if (index !== -1) {
            this._outlet.detach(index);
        }
    }
    ngOnDestroy() {
        this._destroyed.next();
        this._destroyed.complete();
    }
}
NxTabHeaderOutletComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTabHeaderOutletComponent, deps: [{ token: i1.NxTabGroupBase }], target: i0.ɵɵFactoryTarget.Component });
NxTabHeaderOutletComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxTabHeaderOutletComponent, selector: "nx-tab-header-outlet", inputs: { content: "content" }, viewQueries: [{ propertyName: "_outlet", first: true, predicate: ["outlet"], descendants: true, read: ViewContainerRef, static: true }], ngImport: i0, template: '<ng-container #outlet></ng-container>', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTabHeaderOutletComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'nx-tab-header-outlet',
                    template: '<ng-container #outlet></ng-container>',
                }]
        }], ctorParameters: function () { return [{ type: i1.NxTabGroupBase }]; }, propDecorators: { _outlet: [{
                type: ViewChild,
                args: ['outlet', { static: true, read: ViewContainerRef }]
            }], content: [{
                type: Input
            }] } });
