// @aposin/ng-aquila/esm2020/tabs/tabs.module.mjs
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxAccordionModule } from '@aposin/ng-aquila/accordion';
import { NxButtonModule } from '@aposin/ng-aquila/button';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxTabScrollIndicator } from './scroll-indicator/scroll-indicator';
import { NxTabComponent } from './tab';
import { NxTabBodyComponent } from './tab-body';
import { NxTabContentDirective } from './tab-content';
import { NxTabGroupComponent } from './tab-group';
import { NxTabHeaderComponent } from './tab-header';
import { NxTabHeaderOutletComponent } from './tab-header-outlet';
import { NxTabLabelDirective } from './tab-label';
import { NxTabLabelWrapperDirective } from './tab-label-wrapper';
import { NxTabLinkDirective, NxTabNavBarComponent } from './tab-nav-bar';
import * as i0 from "@angular/core";
export class NxTabsModule {
}
NxTabsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTabsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxTabsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTabsModule, declarations: [NxTabComponent,
        NxTabGroupComponent,
        NxTabLabelDirective,
        NxTabLabelWrapperDirective,
        NxTabHeaderComponent,
        NxTabBodyComponent,
        NxTabNavBarComponent,
        NxTabLinkDirective,
        NxTabContentDirective,
        NxTabHeaderOutletComponent,
        NxTabScrollIndicator], imports: [CommonModule, PortalModule, NxAccordionModule, NxIconModule, NxButtonModule], exports: [NxTabComponent,
        NxTabGroupComponent,
        NxTabLabelDirective,
        NxTabLabelWrapperDirective,
        NxTabHeaderComponent,
        NxTabBodyComponent,
        NxTabNavBarComponent,
        NxTabLinkDirective,
        NxTabContentDirective] });
NxTabsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTabsModule, imports: [[CommonModule, PortalModule, NxAccordionModule, NxIconModule, NxButtonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTabsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PortalModule, NxAccordionModule, NxIconModule, NxButtonModule],
                    exports: [
                        NxTabComponent,
                        NxTabGroupComponent,
                        NxTabLabelDirective,
                        NxTabLabelWrapperDirective,
                        NxTabHeaderComponent,
                        NxTabBodyComponent,
                        NxTabNavBarComponent,
                        NxTabLinkDirective,
                        NxTabContentDirective,
                    ],
                    declarations: [
                        NxTabComponent,
                        NxTabGroupComponent,
                        NxTabLabelDirective,
                        NxTabLabelWrapperDirective,
                        NxTabHeaderComponent,
                        NxTabBodyComponent,
                        NxTabNavBarComponent,
                        NxTabLinkDirective,
                        NxTabContentDirective,
                        NxTabHeaderOutletComponent,
                        NxTabScrollIndicator,
                    ],
                }]
        }] });
