// @aposin/ng-aquila/esm2020/modal/dialog/modal-close.directive.mjs
import { Directive, Input, Optional } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./modal-ref";
import * as i2 from "./dialog.service";
/**
 * Button that will close the current modal.
 */
export class NxModalCloseDirective {
    constructor(modalRef, _elementRef, _dialogService) {
        this.modalRef = modalRef;
        this._elementRef = _elementRef;
        this._dialogService = _dialogService;
        /** Default to "button" to prevents accidental form submits. */
        this.type = 'button';
    }
    ngOnInit() {
        if (!this.modalRef) {
            // When this directive is included in a modal via TemplateRef (rather than being
            // in a Component), the DialogRef isn't available via injection because embedded
            // views cannot be given a custom injector. Instead, we look up the DialogRef by
            // ID. This must occur in `onInit`, as the ID binding for the modal container won't
            // be resolved at constructor time.
            this.modalRef = getClosestDialog(this._elementRef, this._dialogService.openModals);
        }
    }
    ngOnChanges(changes) {
        const proxiedChange = changes['modalResult'];
        if (proxiedChange) {
            this.modalResult = proxiedChange.currentValue;
        }
    }
}
NxModalCloseDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxModalCloseDirective, deps: [{ token: i1.NxModalRef, optional: true }, { token: i0.ElementRef }, { token: i2.NxDialogService }], target: i0.ɵɵFactoryTarget.Directive });
NxModalCloseDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxModalCloseDirective, selector: "[nxModalClose]", inputs: { ariaLabel: ["aria-label", "ariaLabel"], type: "type", modalResult: ["nxModalClose", "modalResult"] }, host: { listeners: { "click": "modalRef.close(modalResult)" }, properties: { "attr.aria-label": "ariaLabel || null", "attr.type": "type" } }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxModalCloseDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxModalClose]',
                    host: {
                        '(click)': 'modalRef.close(modalResult)',
                        '[attr.aria-label]': 'ariaLabel || null',
                        '[attr.type]': 'type',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.NxModalRef, decorators: [{
                    type: Optional
                }] }, { type: i0.ElementRef }, { type: i2.NxDialogService }]; }, propDecorators: { ariaLabel: [{
                type: Input,
                args: ['aria-label']
            }], type: [{
                type: Input
            }], modalResult: [{
                type: Input,
                args: ['nxModalClose']
            }] } });
/**
 * Finds the closest NxModalRef to an element by looking at the DOM.
 * @param element Element relative to which to look for a modal.
 * @param openDialogs References to the currently-open modals.
 */
function getClosestDialog(element, openDialogs) {
    let parent = element.nativeElement.parentElement;
    while (parent && !parent.classList.contains('nx-modal__container')) {
        parent = parent.parentElement;
    }
    return parent ? openDialogs.find(modal => modal.id === parent.id) : null;
}
