// @aposin/ng-aquila/esm2020/modal/dialog/modal-config.mjs
/**
 * Configuration for opening a modal  with the NxDialogService.
 */
export class NxModalConfig {
    constructor() {
        /** The ARIA role of the modal element. */
        this.role = 'dialog';
        /** Custom class for the overlay pane. */
        this.panelClass = '';
        /** Whether the modal has a backdrop. */
        this.hasBackdrop = true;
        /** Custom class for the backdrop. */
        this.backdropClass = '';
        /** Whether the user can use escape or clicking on the backdrop to close the modal. */
        this.disableClose = false;
        /** Width of the modal. */
        this.width = '736px';
        /** Height of the modal. */
        this.height = '';
        /** Max-width of the modal. If a number is provided, assumes pixel units. Defaults to 100vw. */
        this.maxWidth = '736px';
        /** Data being injected into the child component. */
        this.data = null;
        /** ID of the element that describes the modal. */
        this.ariaDescribedBy = null;
        /** ID of the element that labels the modal. */
        this.ariaLabelledBy = null;
        /** Aria label to assign to the modal element. */
        this.ariaLabel = null;
        /** Whether the modal should focus the first focusable element on open. */
        this.autoFocus = true;
        /**
         * Whether the modal should restore focus to the
         * previously-focused element, after it's closed.
         */
        this.restoreFocus = true;
        /**
         * Whether the modal should close when the user goes backwards/forwards in history.
         * Note that this usually doesn't include clicking on links (unless the user is using
         * the `HashLocationStrategy`).
         */
        this.closeOnNavigation = true;
        /** Whether a close button with icon should be displayed in the top right corner of the modal.
         *
         * Default value is `false`.
         */
        this.showCloseIcon = false;
        /** Sets the 'aria-label' of the modal close button needed for accessibility.
         *
         * Default value is `'Close dialog'`.
         */
        this.closeIconButtonLabel = 'Close dialog';
        /** Sets locale direction for the modal
         *
         * Default value is `'ltr'`
         */
        this.direction = 'ltr';
    }
}
