// @aposin/ng-aquila/esm2020/overlay/overlay-config.mjs
export const VERTICAL_DIRECTIONS = ['top', 'bottom'];
export const HORIZONTAL_DIRECTIONS = ['left', 'right'];
export const CLOCKWISE_DIRECTIONS = ['top', 'right', 'bottom', 'left'];
export const BASE_OFFSET = 0; // TODO should be 16 once we have a "showArrow" option
/**
 * Configuration for opening a overlay  with the NxDialogService.
 */
export class NxOverlayConfig {
    constructor() {
        /** The ARIA role of the overlay element. */
        this.role = null;
        /** Custom class for the overlay pane. */
        this.panelClass = '';
        /** Whether the overlay has a backdrop. */
        this.hasBackdrop = false;
        /** Custom class for the backdrop. */
        this.backdropClass = '';
        /** Whether the user can use escape or clicking on the backdrop to close the overlay. */
        this.closeOnClickOutside = true;
        /** Width of the overlay. */
        this.width = '';
        /** Height of the overlay. */
        this.height = '';
        /** Max-width of the overlay. If a number is provided, assumes pixel units. Defaults to 100vw. */
        this.maxWidth = '';
        /** Position overrides. */
        this.direction = 'bottom';
        /** The fallbacks that are chosen. */
        this.fallbackOrientation = 'clockwise';
        /** ID of the element that describes the overlay. */
        this.ariaDescribedBy = null;
        /** ID of the element that labels the overlay. */
        this.ariaLabelledBy = null;
        /** Aria label to assign to the overlay element. */
        this.ariaLabel = null;
        /** Whether the overlay should focus the first focusable element on open. */
        this.autoFocus = true;
        /**
         * Whether the overlay should restore focus to the
         * previously-focused element, after it's closed.
         */
        this.restoreFocus = true;
        /**
         * Whether the overlay should close when the user goes backwards/forwards in history.
         * Note that this usually doesn't include clicking on links (unless the user is using
         * the `HashLocationStrategy`).
         */
        this.closeOnNavigation = true;
    }
}
