// @aposin/ng-aquila/esm2020/overlay/overlay.module.mjs
import { A11yModule } from '@angular/cdk/a11y';
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { NxOverlayContainerComponent } from './overlay-container.component';
import { NX_OVERLAY_SCROLL_STRATEGY_PROVIDER, NxOverlayService } from './overlay-service';
import { NxOverlayPositionBuilder } from './position-builder';
import * as i0 from "@angular/core";
export class NxOverlayModule {
}
NxOverlayModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxOverlayModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxOverlayModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxOverlayModule, declarations: [NxOverlayContainerComponent], imports: [OverlayModule, A11yModule, PortalModule, RouterModule], exports: [NxOverlayContainerComponent, OverlayModule] });
NxOverlayModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxOverlayModule, providers: [NxOverlayService, NxOverlayPositionBuilder, NX_OVERLAY_SCROLL_STRATEGY_PROVIDER], imports: [[OverlayModule, A11yModule, PortalModule, RouterModule], OverlayModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxOverlayModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [OverlayModule, A11yModule, PortalModule, RouterModule],
                    exports: [NxOverlayContainerComponent, OverlayModule],
                    declarations: [NxOverlayContainerComponent],
                    providers: [NxOverlayService, NxOverlayPositionBuilder, NX_OVERLAY_SCROLL_STRATEGY_PROVIDER],
                }]
        }] });
