// @aposin/ng-aquila/esm2020/message/message-toast/message-toast-config.mjs
/**
 * Configuration used when opening a message toast.
 */
export class NxMessageToastConfig {
    constructor() {
        /** The politeness level for the LiveAnnouncer announcement.
         *
         *  Default value: 'polite'.
         */
        this.politeness = 'polite';
        /**
         * Message to be announced by the LiveAnnouncer. When opening a toast message without a custom
         * component or template, the announcement message will default to the specified message.
         *
         * Default value: ''.
         */
        this.announcementMessage = '';
        /** The length of time in milliseconds to wait before automatically dismissing the message toast.
         *
         * Default value: 3000.
         */
        this.duration = 3000;
        /** Context of the message toast.
         *
         * Default value: 'info'. */
        this.context = 'info';
    }
}
/**
 * @docs-private
 * Needed so that the user text data can be injected in the message toastcomponent
 */
export class NxMessageToastData {
    constructor(data) {
        this.data = data;
    }
}
