// @aposin/ng-aquila/esm2020/config/nx-expert.module.mjs
import { NgModule } from '@angular/core';
import { ERROR_DEFAULT_OPTIONS, LABEL_DEFAULT_OPTIONS } from '@aposin/ng-aquila/base';
import { SELECTABLE_CARD_DEFAULT_OPTIONS } from '@aposin/ng-aquila/card';
import { CIRCLE_TOGGLE_GROUP_DEFAULT_OPTIONS } from '@aposin/ng-aquila/circle-toggle';
import { COMPARISON_TABLE_DEFAULT_OPTIONS } from '@aposin/ng-aquila/comparison-table';
import { DATA_DISPLAY_DEFAULT_OPTIONS } from '@aposin/ng-aquila/data-display';
import { DATEPICKER_DEFAULT_OPTIONS } from '@aposin/ng-aquila/datefield';
import { FORMFIELD_DEFAULT_OPTIONS } from '@aposin/ng-aquila/formfield';
import { SMALL_STAGE_DEFAULT_OPTIONS } from '@aposin/ng-aquila/small-stage';
import { TAB_GROUP_DEFAULT_OPTIONS, TAB_NAV_BAR_DEFAULT_OPTIONS } from '@aposin/ng-aquila/tabs';
import * as i0 from "@angular/core";
// expert presets
const comparisonTableExpertOptions = { useFullRowForExpandableArea: true };
const formfieldExpertOptions = {
    appearance: 'outline',
    nxFloatLabel: 'always',
};
const errorExpertOptions = { appearance: 'text' };
const labelExpertOptions = { size: 'small' };
const datepickerExpertOptions = { toggleIconTabindex: -1 };
const tabGroupOptions = { appearance: 'expert' };
const tabNavBarOptions = { appearance: 'expert' };
const selectableCardOptions = { appearance: 'expert' };
const smallStageOptions = { appearance: 'expert' };
const circleToggleGroupOptions = { appearance: 'expert' };
const dataDisplayOptions = { size: 'medium' };
// should not be a singleton since different tokens can be used in separate modules
export class NxExpertModule {
}
NxExpertModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxExpertModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxExpertModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxExpertModule });
NxExpertModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxExpertModule, providers: [
        { provide: COMPARISON_TABLE_DEFAULT_OPTIONS, useValue: comparisonTableExpertOptions },
        { provide: FORMFIELD_DEFAULT_OPTIONS, useValue: formfieldExpertOptions },
        { provide: LABEL_DEFAULT_OPTIONS, useValue: labelExpertOptions },
        { provide: DATEPICKER_DEFAULT_OPTIONS, useValue: datepickerExpertOptions },
        { provide: ERROR_DEFAULT_OPTIONS, useValue: errorExpertOptions },
        { provide: TAB_GROUP_DEFAULT_OPTIONS, useValue: tabGroupOptions },
        { provide: TAB_NAV_BAR_DEFAULT_OPTIONS, useValue: tabNavBarOptions },
        { provide: SELECTABLE_CARD_DEFAULT_OPTIONS, useValue: selectableCardOptions },
        { provide: SMALL_STAGE_DEFAULT_OPTIONS, useValue: smallStageOptions },
        { provide: CIRCLE_TOGGLE_GROUP_DEFAULT_OPTIONS, useValue: circleToggleGroupOptions },
        { provide: DATA_DISPLAY_DEFAULT_OPTIONS, useValue: dataDisplayOptions },
    ] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxExpertModule, decorators: [{
            type: NgModule,
            args: [{
                    providers: [
                        { provide: COMPARISON_TABLE_DEFAULT_OPTIONS, useValue: comparisonTableExpertOptions },
                        { provide: FORMFIELD_DEFAULT_OPTIONS, useValue: formfieldExpertOptions },
                        { provide: LABEL_DEFAULT_OPTIONS, useValue: labelExpertOptions },
                        { provide: DATEPICKER_DEFAULT_OPTIONS, useValue: datepickerExpertOptions },
                        { provide: ERROR_DEFAULT_OPTIONS, useValue: errorExpertOptions },
                        { provide: TAB_GROUP_DEFAULT_OPTIONS, useValue: tabGroupOptions },
                        { provide: TAB_NAV_BAR_DEFAULT_OPTIONS, useValue: tabNavBarOptions },
                        { provide: SELECTABLE_CARD_DEFAULT_OPTIONS, useValue: selectableCardOptions },
                        { provide: SMALL_STAGE_DEFAULT_OPTIONS, useValue: smallStageOptions },
                        { provide: CIRCLE_TOGGLE_GROUP_DEFAULT_OPTIONS, useValue: circleToggleGroupOptions },
                        { provide: DATA_DISPLAY_DEFAULT_OPTIONS, useValue: dataDisplayOptions },
                    ],
                }]
        }] });
