// @aposin/ng-aquila/esm2020/progressbar/progressbar.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxProgressbarComponent } from './progressbar.component';
import * as i0 from "@angular/core";
export class NxProgressbarModule {
}
NxProgressbarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressbarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxProgressbarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressbarModule, declarations: [NxProgressbarComponent], imports: [CommonModule], exports: [NxProgressbarComponent] });
NxProgressbarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressbarModule, providers: [], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressbarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [NxProgressbarComponent],
                    declarations: [NxProgressbarComponent],
                    providers: [],
                }]
        }] });
