// @aposin/ng-aquila/esm2020/page-search/page-search.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxAutocompleteModule } from '@aposin/ng-aquila/autocomplete';
import { NxButtonModule } from '@aposin/ng-aquila/button';
import { NxFormfieldModule } from '@aposin/ng-aquila/formfield';
import { NxGridModule } from '@aposin/ng-aquila/grid';
import { NxPageSearchComponent } from './page-search.component';
import * as i0 from "@angular/core";
export class NxPageSearchModule {
}
NxPageSearchModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPageSearchModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxPageSearchModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPageSearchModule, declarations: [NxPageSearchComponent], imports: [CommonModule, NxAutocompleteModule, NxFormfieldModule, NxButtonModule, NxGridModule], exports: [NxPageSearchComponent] });
NxPageSearchModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPageSearchModule, imports: [[CommonModule, NxAutocompleteModule, NxFormfieldModule, NxButtonModule, NxGridModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPageSearchModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, NxAutocompleteModule, NxFormfieldModule, NxButtonModule, NxGridModule],
                    declarations: [NxPageSearchComponent],
                    exports: [NxPageSearchComponent],
                }]
        }] });
