// @aposin/ng-aquila/esm2020/badge/badge.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxBadgeComponent } from './badge.component';
import * as i0 from "@angular/core";
export class NxBadgeModule {
}
NxBadgeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBadgeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxBadgeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBadgeModule, declarations: [NxBadgeComponent], imports: [CommonModule], exports: [NxBadgeComponent] });
NxBadgeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBadgeModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBadgeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [NxBadgeComponent],
                    exports: [NxBadgeComponent],
                }]
        }] });
