// @aposin/ng-aquila/esm2020/mask/mask.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxFormfieldModule } from '@aposin/ng-aquila/formfield';
import { NxIbanMaskDirective } from './iban-mask.directive';
import { NxMaskDirective } from './mask.directive';
import * as i0 from "@angular/core";
export class NxMaskModule {
}
NxMaskModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMaskModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxMaskModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMaskModule, declarations: [NxMaskDirective, NxIbanMaskDirective], imports: [CommonModule, NxFormfieldModule], exports: [NxFormfieldModule, NxMaskDirective, NxIbanMaskDirective] });
NxMaskModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMaskModule, imports: [[CommonModule, NxFormfieldModule], NxFormfieldModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMaskModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, NxFormfieldModule],
                    declarations: [NxMaskDirective, NxIbanMaskDirective],
                    exports: [NxFormfieldModule, NxMaskDirective, NxIbanMaskDirective],
                }]
        }] });
