// @aposin/ng-aquila/esm2020/notification-panel/notification-panel.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { NxOverlayModule } from '@aposin/ng-aquila/overlay';
import { NxNotificationPanelItemComponent } from './notification-item/notification-item.component';
import { NxNotificationItemActionsDirective, NxNotificationItemContentDirective, NxNotificationItemHeaderDirective, NxNotificationItemMetadataDirective, } from './notification-item/notification-item-slots';
import { NX_NOTIFICATION_PANEL_SCROLL_STRATEGY_PROVIDER, NxNotificationPanelTriggerDirective } from './notification-panel-trigger.directive';
import { NxNotificationPanelComponent } from './panel/notification-panel.component';
import * as i0 from "@angular/core";
export class NxNotificationPanelModule {
}
NxNotificationPanelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationPanelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxNotificationPanelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationPanelModule, declarations: [NxNotificationPanelComponent,
        NxNotificationPanelItemComponent,
        NxNotificationPanelTriggerDirective,
        NxNotificationItemActionsDirective,
        NxNotificationItemContentDirective,
        NxNotificationItemMetadataDirective,
        NxNotificationItemHeaderDirective], imports: [NxOverlayModule, RouterModule], exports: [NxNotificationPanelComponent,
        NxNotificationPanelItemComponent,
        NxNotificationPanelTriggerDirective,
        NxNotificationItemActionsDirective,
        NxNotificationItemContentDirective,
        NxNotificationItemMetadataDirective,
        NxNotificationItemHeaderDirective] });
NxNotificationPanelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationPanelModule, providers: [NX_NOTIFICATION_PANEL_SCROLL_STRATEGY_PROVIDER], imports: [[NxOverlayModule, RouterModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationPanelModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [NxOverlayModule, RouterModule],
                    exports: [
                        NxNotificationPanelComponent,
                        NxNotificationPanelItemComponent,
                        NxNotificationPanelTriggerDirective,
                        NxNotificationItemActionsDirective,
                        NxNotificationItemContentDirective,
                        NxNotificationItemMetadataDirective,
                        NxNotificationItemHeaderDirective,
                    ],
                    declarations: [
                        NxNotificationPanelComponent,
                        NxNotificationPanelItemComponent,
                        NxNotificationPanelTriggerDirective,
                        NxNotificationItemActionsDirective,
                        NxNotificationItemContentDirective,
                        NxNotificationItemMetadataDirective,
                        NxNotificationItemHeaderDirective,
                    ],
                    providers: [NX_NOTIFICATION_PANEL_SCROLL_STRATEGY_PROVIDER],
                }]
        }] });
