// @aposin/ng-aquila/esm2020/notification-panel/notification-panel-trigger.directive.mjs
import { Overlay } from '@angular/cdk/overlay';
import { Directive, Inject, InjectionToken, Input, Optional, Self } from '@angular/core';
import { take } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@aposin/ng-aquila/overlay";
/** Injection token that determines the scroll handling while a notification-panel is open. */
export const NX_NOTIFICATION_PANEL_SCROLL_STRATEGY = new InjectionToken('nx-notification-panel-scroll-strategy');
/** @docs-private */
export function NX_NOTIFICATION_PANEL_SCROLL_STRATEGY_PROVIDER_FACTORY(overlay) {
    return () => overlay.scrollStrategies.reposition();
}
/** @docs-private */
export const NX_NOTIFICATION_PANEL_SCROLL_STRATEGY_PROVIDER = {
    provide: NX_NOTIFICATION_PANEL_SCROLL_STRATEGY,
    useFactory: NX_NOTIFICATION_PANEL_SCROLL_STRATEGY_PROVIDER_FACTORY,
    deps: [Overlay],
};
const DEFAULT_CONFIG = {
    direction: 'bottom-start',
    fallbackOrientation: 'vertical',
    autoFocus: true,
    offset: 8,
};
export class NxNotificationPanelTriggerDirective {
    constructor(_nxOverlay, _element, _triggerButton, _defaultScrollStrategyFactory) {
        this._nxOverlay = _nxOverlay;
        this._element = _element;
        this._triggerButton = _triggerButton;
        this._defaultScrollStrategyFactory = _defaultScrollStrategyFactory;
        /** Strategy factory that will be used to handle scrolling while the notification-panel panel is open. */
        this._scrollStrategyFactory = this._defaultScrollStrategyFactory;
    }
    set notificationPanel(value) {
        this._panelTemplate = value;
    }
    get notificationPanel() {
        return this._panelTemplate;
    }
    open() {
        if (this._overlayRef) {
            return;
        }
        const config = { ...DEFAULT_CONFIG, scrollStrategy: this._scrollStrategyFactory(), triggerButton: this._triggerButton ?? undefined };
        this._overlayRef = this._nxOverlay.open(this._panelTemplate, this._element, config);
        this._overlayRef
            .afterClosed()
            .pipe(take(1))
            .subscribe(() => this.close());
    }
    close() {
        if (this._overlayRef) {
            this._overlayRef.close();
            this._overlayRef = null;
        }
    }
}
NxNotificationPanelTriggerDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationPanelTriggerDirective, deps: [{ token: i1.NxOverlayService }, { token: i0.ElementRef }, { token: i1.NxTriggerButton, optional: true, self: true }, { token: NX_NOTIFICATION_PANEL_SCROLL_STRATEGY }], target: i0.ɵɵFactoryTarget.Directive });
NxNotificationPanelTriggerDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxNotificationPanelTriggerDirective, selector: "[nxNotificationPanelTriggerFor]", inputs: { notificationPanel: ["nxNotificationPanelTriggerFor", "notificationPanel"] }, host: { listeners: { "click": "open()" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationPanelTriggerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxNotificationPanelTriggerFor]',
                    host: {
                        '(click)': 'open()',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.NxOverlayService }, { type: i0.ElementRef }, { type: i1.NxTriggerButton, decorators: [{
                    type: Optional
                }, {
                    type: Self
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [NX_NOTIFICATION_PANEL_SCROLL_STRATEGY]
                }] }]; }, propDecorators: { notificationPanel: [{
                type: Input,
                args: ['nxNotificationPanelTriggerFor']
            }] } });
