// @aposin/ng-aquila/esm2020/radio-button/radio-button.module.mjs
import { ObserversModule } from '@angular/cdk/observers';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxErrorModule, NxLabelModule } from '@aposin/ng-aquila/base';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxRadioComponent, NxRadioGroupComponent } from './radio-button';
import * as i0 from "@angular/core";
export class NxRadioModule {
}
NxRadioModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRadioModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxRadioModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRadioModule, declarations: [NxRadioComponent, NxRadioGroupComponent], imports: [CommonModule, NxIconModule, ObserversModule], exports: [NxRadioComponent, NxRadioGroupComponent, NxErrorModule, NxLabelModule] });
NxRadioModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRadioModule, imports: [[CommonModule, NxIconModule, ObserversModule], NxErrorModule, NxLabelModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRadioModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxRadioComponent, NxRadioGroupComponent],
                    exports: [NxRadioComponent, NxRadioGroupComponent, NxErrorModule, NxLabelModule],
                    imports: [CommonModule, NxIconModule, ObserversModule],
                }]
        }] });
