// @aposin/ng-aquila/esm2020/code-input/code-input.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { NxCodeInputComponent } from './code-input.component';
import { NxCodeInputIntl } from './code-input-intl';
import * as i0 from "@angular/core";
export class NxCodeInputModule {
}
NxCodeInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxCodeInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxCodeInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxCodeInputModule, declarations: [NxCodeInputComponent], imports: [CommonModule, FormsModule, ReactiveFormsModule], exports: [NxCodeInputComponent] });
NxCodeInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxCodeInputModule, providers: [NxCodeInputIntl], imports: [[CommonModule, FormsModule, ReactiveFormsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxCodeInputModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, FormsModule, ReactiveFormsModule],
                    declarations: [NxCodeInputComponent],
                    exports: [NxCodeInputComponent],
                    providers: [NxCodeInputIntl],
                }]
        }] });
