// @aposin/ng-aquila/esm2020/video/video.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxVideoComponent } from './video.component';
import * as i0 from "@angular/core";
export class NxVideoModule {
}
NxVideoModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxVideoModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxVideoModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxVideoModule, declarations: [NxVideoComponent], imports: [CommonModule, NxIconModule], exports: [NxVideoComponent] });
NxVideoModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxVideoModule, imports: [[CommonModule, NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxVideoModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxVideoComponent],
                    exports: [NxVideoComponent],
                    imports: [CommonModule, NxIconModule],
                }]
        }] });
