// @aposin/ng-aquila/esm2020/switcher/switcher.module.mjs
import { ObserversModule } from '@angular/cdk/observers';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxSwitcherComponent } from './switcher.component';
import * as i0 from "@angular/core";
export class NxSwitcherModule {
}
NxSwitcherModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSwitcherModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxSwitcherModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSwitcherModule, declarations: [NxSwitcherComponent], imports: [CommonModule, NxIconModule, ObserversModule], exports: [NxSwitcherComponent] });
NxSwitcherModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSwitcherModule, imports: [[CommonModule, NxIconModule, ObserversModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSwitcherModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxSwitcherComponent],
                    exports: [NxSwitcherComponent],
                    imports: [CommonModule, NxIconModule, ObserversModule],
                }]
        }] });
