// @aposin/ng-aquila/esm2020/timefield/timefield.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { NxRadioModule } from '@aposin/ng-aquila/radio-button';
import { NxTimefieldComponent } from './timefield.component';
import { NxTimefieldIntl } from './timefield-intl';
import * as i0 from "@angular/core";
export class NxTimefieldModule {
}
NxTimefieldModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTimefieldModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxTimefieldModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTimefieldModule, declarations: [NxTimefieldComponent], imports: [CommonModule, FormsModule, NxRadioModule], exports: [NxTimefieldComponent] });
NxTimefieldModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTimefieldModule, providers: [NxTimefieldIntl], imports: [[CommonModule, FormsModule, NxRadioModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTimefieldModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, FormsModule, NxRadioModule],
                    declarations: [NxTimefieldComponent],
                    exports: [NxTimefieldComponent],
                    providers: [NxTimefieldIntl],
                }]
        }] });
