// @aposin/ng-aquila/schematics/ng-update/migrations/misc-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MiscTemplateMigration = void 0;
const schematics_1 = require("@angular/cdk/schematics");
/**
 * Migration that walks through every template and reports if there are
 * instances of outdated Angular CDK API that can't be migrated automatically.
 *
 * TODO: make it reusable across major versions, e.g. with a custom data object
 * similar to the CDK UpgradeData
 */
class MiscTemplateMigration extends schematics_1.Migration {
    constructor() {
        super(...arguments);
        // For the first version we need it for V11 but we might need the same
        // functionality for future major versions as well
        this.enabled = this.targetVersion === schematics_1.TargetVersion.V11;
    }
    visitTemplate(template) {
        let relativeOffsets = [];
        const limitedTo = ['nx-single-stepper', 'nx-multi-stepper', 'nx-progress-stepper'];
        relativeOffsets.push(...(0, schematics_1.findInputsOnElementWithTag)(template.content, 'title', limitedTo));
        relativeOffsets
            .map(offset => template.start + offset)
            .forEach(start => {
            this.failures.push({
                filePath: template.filePath,
                position: template.getCharacterAndLineOfPosition(start),
                message: `Found deprecated input "title" on nx-single-stepper, nx-multi-stepper or ` +
                    `nx-progress-stepper. Please use a "nx-label" element inside the content instead.`,
            });
        });
        // nxAriaLabel on nx-dropdown
        relativeOffsets = [];
        relativeOffsets.push(...(0, schematics_1.findInputsOnElementWithTag)(template.content, 'nxAriaLabel', ['nx-dropdown']));
        relativeOffsets
            .map(offset => template.start + offset)
            .forEach(start => {
            this.failures.push({
                filePath: template.filePath,
                position: template.getCharacterAndLineOfPosition(start),
                message: `Found deprecated input "nxAriaLabel" on nx-dropdown. Please remove the input.`,
            });
        });
    }
}
exports.MiscTemplateMigration = MiscTemplateMigration;
