// @aposin/ng-aquila/esm2020/modal/modal.module.mjs
import { A11yModule } from '@angular/cdk/a11y';
/* new modal implementation */
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NX_MODAL_SCROLL_STRATEGY_PROVIDER, NxDialogService } from './dialog/dialog.service';
import { NxModalCloseDirective } from './dialog/modal-close.directive';
import { NxModalContainer } from './dialog/modal-container.component';
import { NxModalActionsDirective, NxModalComponent, NxModalContentDirective } from './modal.component';
import { NxModalService } from './modal.service';
import { NxOpenModalOnClickDirective } from './modal-open-on-click.directive';
import * as i0 from "@angular/core";
export class NxModalModule {
    // define the modal service on root level as an application wide singleton,
    // assuming that we only display one modal at a time, which seems quite reasonable
    // this is static and named "forRoot" by convention
    static forRoot() {
        return {
            ngModule: NxModalModule,
            providers: [NxModalService],
        };
    }
}
NxModalModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxModalModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxModalModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxModalModule, declarations: [NxModalComponent,
        NxOpenModalOnClickDirective,
        NxModalActionsDirective,
        NxModalContentDirective,
        /* new modal implementation */
        NxModalContainer,
        NxModalCloseDirective], imports: [A11yModule, OverlayModule, PortalModule, CommonModule, NxIconModule], exports: [NxModalComponent,
        NxOpenModalOnClickDirective,
        NxModalActionsDirective,
        NxModalContentDirective,
        /* new modal implementation */
        NxModalContainer,
        NxModalCloseDirective] });
NxModalModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxModalModule, providers: [NxDialogService, NX_MODAL_SCROLL_STRATEGY_PROVIDER], imports: [[A11yModule, OverlayModule, PortalModule, CommonModule, NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxModalModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [A11yModule, OverlayModule, PortalModule, CommonModule, NxIconModule],
                    declarations: [
                        NxModalComponent,
                        NxOpenModalOnClickDirective,
                        NxModalActionsDirective,
                        NxModalContentDirective,
                        /* new modal implementation */
                        NxModalContainer,
                        NxModalCloseDirective,
                    ],
                    exports: [
                        NxModalComponent,
                        NxOpenModalOnClickDirective,
                        NxModalActionsDirective,
                        NxModalContentDirective,
                        /* new modal implementation */
                        NxModalContainer,
                        NxModalCloseDirective,
                    ],
                    providers: [NxDialogService, NX_MODAL_SCROLL_STRATEGY_PROVIDER],
                }]
        }] });
