// @aposin/ng-aquila/esm2020/action/action.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxActionComponent } from './action.component';
import { NxActionIconDirective } from './action-icon.directive';
import * as i0 from "@angular/core";
export class NxActionModule {
}
NxActionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxActionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxActionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxActionModule, declarations: [NxActionComponent, NxActionIconDirective], imports: [NxIconModule, CommonModule], exports: [NxActionComponent, NxActionIconDirective] });
NxActionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxActionModule, imports: [[NxIconModule, CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxActionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [NxIconModule, CommonModule],
                    declarations: [NxActionComponent, NxActionIconDirective],
                    exports: [NxActionComponent, NxActionIconDirective],
                }]
        }] });
