// @aposin/ng-aquila/esm2020/checkbox/checkbox.module.mjs
import { ObserversModule } from '@angular/cdk/observers';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxErrorModule, NxLabelModule } from '@aposin/ng-aquila/base';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxCheckboxComponent, NxCheckboxGroupComponent } from './checkbox.component';
import * as i0 from "@angular/core";
export class NxCheckboxModule {
}
NxCheckboxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxCheckboxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxCheckboxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxCheckboxModule, declarations: [NxCheckboxComponent, NxCheckboxGroupComponent], imports: [CommonModule, NxIconModule, ObserversModule], exports: [NxCheckboxComponent, NxCheckboxGroupComponent, NxLabelModule, NxErrorModule] });
NxCheckboxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxCheckboxModule, imports: [[CommonModule, NxIconModule, ObserversModule], NxLabelModule, NxErrorModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxCheckboxModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxCheckboxComponent, NxCheckboxGroupComponent],
                    exports: [NxCheckboxComponent, NxCheckboxGroupComponent, NxLabelModule, NxErrorModule],
                    imports: [CommonModule, NxIconModule, ObserversModule],
                }]
        }] });
