// @aposin/ng-aquila/esm2020/file-uploader/file-uploader-hint.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
let nextId = 0;
// This Directive solely purpose is to mark given ng-content and project it into the required destination.
export class NxFileUploaderHintDirective {
    constructor() {
        /** Sets the id of the file upload hint. */
        this.id = `nx-formfield-hint-${nextId++}`;
    }
}
NxFileUploaderHintDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFileUploaderHintDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxFileUploaderHintDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxFileUploaderHintDirective, selector: "[nxFileUploadHint]", inputs: { id: "id" }, host: { properties: { "attr.id": "id", "class.nx-file-upload-hint": "true" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFileUploaderHintDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxFileUploadHint]',
                    host: {
                        '[attr.id]': 'id',
                        '[class.nx-file-upload-hint]': 'true',
                    },
                }]
        }], propDecorators: { id: [{
                type: Input
            }] } });
