// @aposin/ng-aquila/esm2020/file-uploader/file-uploader.model.mjs
import { EventEmitter } from '@angular/core';
/**
 * Class used to keep track of the current states of the file
 */
export class FileItem {
    constructor(file, index = 0) {
        /** Whether the file is being uploaded at the moment. */
        this.isUploading = false;
        /** Whether the file was successfully uploaded. */
        this.isUploaded = false;
        /** Whether an error occured when uploading the file. */
        this.isError = false;
        /** Sets the index of the file. */
        this.index = 0;
        /**
         * emits the FileItem on changed
         */
        this.onChange = new EventEmitter();
        this.file = file;
        this.name = file.name;
        this._size = file.size;
        this._type = file.type;
        this.index = index;
    }
    /** Returns the file size in bytes */
    get size() {
        return this._size;
    }
    /** Returns the file type */
    get type() {
        return this._type;
    }
    /**
     * sets the file to a uploaded state
     */
    setUploadedState() {
        this.isUploaded = true;
        this.isUploading = false;
        this.isError = false;
        this.onChange.emit(this);
    }
    /**
     * sets the file to a 'upload-in-progress' state
     */
    setUploadingState() {
        this.isUploaded = false;
        this.isUploading = true;
        this.isError = false;
        this.onChange.emit(this);
    }
    /**
     * sets the file to an error state
     */
    setErrorState() {
        this.isUploaded = false;
        this.isUploading = false;
        this.isError = true;
        this.onChange.emit(this);
    }
}
