// @aposin/ng-aquila/esm2020/file-uploader/file-uploader-intl.mjs
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
/** File uploader data that requires internationalization. */
export class NxFileUploaderIntl {
    constructor() {
        /**
         * Stream that emits whenever the labels here are changed. Use this to notify
         * components if the labels have changed after initialization.
         */
        this.changes = new Subject();
        /** A label for the uploading state used in the queue. */
        this.uploadingLabel = 'uploading';
        /** Aria-label for the delete icon. */
        this.deleteLabel = 'Delete';
        /** Aria-label for the 'successfully-uploaded' icon. */
        this.uploadedStateLabel = 'uploaded';
    }
}
NxFileUploaderIntl.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFileUploaderIntl, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
NxFileUploaderIntl.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFileUploaderIntl, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFileUploaderIntl, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
