// @aposin/ng-aquila/esm2020/file-uploader/file-uploader-trigger.directive.mjs
import { Directive, HostListener, Input } from '@angular/core';
import * as i0 from "@angular/core";
// This Directive solely purpose is to mark given ng-content and project it into the required destination.
export class NxFileUploaderTriggerDirective {
    /** References the file upload component instance that the trigger is associated with. */
    set fileUpload(value) {
        if (this._fileUpload !== value) {
            this._fileUpload = value;
        }
    }
    get fileUpload() {
        return this._fileUpload;
    }
    _onClick() {
        this._fileUpload.uploadFiles();
    }
}
NxFileUploaderTriggerDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFileUploaderTriggerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxFileUploaderTriggerDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxFileUploaderTriggerDirective, selector: "[nxFileUploadTriggerFor]", inputs: { fileUpload: ["nxFileUploadTriggerFor", "fileUpload"] }, host: { listeners: { "click": "_onClick()" } }, exportAs: ["nxFileUploadTrigger"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFileUploaderTriggerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxFileUploadTriggerFor]',
                    exportAs: 'nxFileUploadTrigger',
                }]
        }], propDecorators: { fileUpload: [{
                type: Input,
                args: ['nxFileUploadTriggerFor']
            }], _onClick: [{
                type: HostListener,
                args: ['click']
            }] } });
