// @aposin/ng-aquila/esm2020/file-uploader/file-uploader-button.directive.mjs
import { Directive, HostListener, Input } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
// This Directive solely purpose is to mark given ng-content and project it into the required destination.
export class NxFileUploaderButtonDirective {
    constructor(_cdr) {
        this._cdr = _cdr;
        this._clicked = new Subject();
        this._disabled = false;
    }
    /** Whether the file uploader is disabled */
    set disabled(value) {
        this._disabled = value;
        this._cdr.markForCheck();
    }
    get disabled() {
        return this._disabled;
    }
    /** @docs-private */
    setDescribedByIds(ids) {
        this._ariaDescribedby = ids.join(' ');
    }
    /** @docs-private */
    onClick() {
        this._clicked.next();
    }
    ngOnDestroy() {
        this._clicked.complete();
    }
}
NxFileUploaderButtonDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFileUploaderButtonDirective, deps: [{ token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Directive });
NxFileUploaderButtonDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxFileUploaderButtonDirective, selector: "[nxFileUploadButton]", inputs: { disabled: "disabled" }, host: { listeners: { "click": "onClick()" }, properties: { "attr.aria-describedby": "_ariaDescribedby || null", "attr.disabled": "disabled || null", "class.nx-file-upload-button": "true" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFileUploaderButtonDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxFileUploadButton]',
                    host: {
                        '[attr.aria-describedby]': '_ariaDescribedby || null',
                        '[attr.disabled]': 'disabled || null',
                        '[class.nx-file-upload-button]': 'true',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i0.ChangeDetectorRef }]; }, propDecorators: { disabled: [{
                type: Input
            }], onClick: [{
                type: HostListener,
                args: ['click']
            }] } });
