// @aposin/ng-aquila/esm2020/small-stage/small-stage.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxSmallStageImageBottomDirective } from './directives/bottom-image.directive';
import { NxSmallStageImageEndDirective } from './directives/end-image.directive';
import { NxSmallStageHeaderDirective } from './directives/header.directive';
import { NxSmallStageImageDirective } from './directives/image.directive';
import { NxSmallStageImageStartDirective } from './directives/start-image.directive';
import { NxSmallStageComponent } from './small-stage.component';
import * as i0 from "@angular/core";
export class NxSmallStageModule {
}
NxSmallStageModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSmallStageModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxSmallStageModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSmallStageModule, declarations: [NxSmallStageImageDirective,
        NxSmallStageImageStartDirective,
        NxSmallStageImageEndDirective,
        NxSmallStageHeaderDirective,
        NxSmallStageComponent,
        NxSmallStageImageBottomDirective], imports: [CommonModule], exports: [NxSmallStageImageDirective,
        NxSmallStageImageStartDirective,
        NxSmallStageImageEndDirective,
        NxSmallStageHeaderDirective,
        NxSmallStageComponent,
        NxSmallStageImageBottomDirective] });
NxSmallStageModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSmallStageModule, providers: [], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSmallStageModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [
                        NxSmallStageImageDirective,
                        NxSmallStageImageStartDirective,
                        NxSmallStageImageEndDirective,
                        NxSmallStageHeaderDirective,
                        NxSmallStageComponent,
                        NxSmallStageImageBottomDirective,
                    ],
                    declarations: [
                        NxSmallStageImageDirective,
                        NxSmallStageImageStartDirective,
                        NxSmallStageImageEndDirective,
                        NxSmallStageHeaderDirective,
                        NxSmallStageComponent,
                        NxSmallStageImageBottomDirective,
                    ],
                    providers: [],
                }]
        }] });
