// @aposin/ng-aquila/esm2020/tree/node.mjs
import { CdkTreeNode, CdkTreeNodeDef } from '@angular/cdk/tree';
import { Component, ContentChild, Directive, Input } from '@angular/core';
import { NxTreeNodeActionItem } from './action-item.directive';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/tree";
import * as i2 from "@angular/cdk/a11y";
/**
 * Wrapper for the CdkTree node with custom design styles.
 */
export class NxTreeNodeComponent extends CdkTreeNode {
    constructor(_elementRef, _tree, _focusMonitor) {
        super(_elementRef, _tree);
        this._elementRef = _elementRef;
        this._tree = _tree;
        this._focusMonitor = _focusMonitor;
        this._focusMonitor.monitor(this._elementRef.nativeElement);
    }
    ngOnDestroy() {
        this._focusMonitor.stopMonitoring(this._elementRef.nativeElement);
        super.ngOnDestroy();
    }
    /** Update the focused data in tree keyboard interaction */
    _focus() {
        this._tree.updateFocusedData(this._data);
    }
    focus() {
        if (this.actionItem) {
            this.actionItem.focus?.();
        }
        else {
            this._elementRef.nativeElement.focus();
        }
    }
}
NxTreeNodeComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTreeNodeComponent, deps: [{ token: i0.ElementRef }, { token: i1.CdkTree }, { token: i2.FocusMonitor }], target: i0.ɵɵFactoryTarget.Component });
NxTreeNodeComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxTreeNodeComponent, selector: "nx-tree-node", inputs: { disabled: "disabled", tabIndex: "tabIndex" }, host: { properties: { "attr.aria-expanded": "isExpanded", "attr.aria-level": "role === \"treeitem\" ? level : null", "attr.role": "role", "class.is-expanded": "isExpanded", "attr.tabindex": "-1" }, classAttribute: "nx-tree__node" }, providers: [{ provide: CdkTreeNode, useExisting: NxTreeNodeComponent }], queries: [{ propertyName: "actionItem", first: true, predicate: NxTreeNodeActionItem, descendants: true }], exportAs: ["nxTreeNode"], usesInheritance: true, ngImport: i0, template: "<div class=\"nx-tree__node-box\">\n    <ng-content></ng-content>\n</div>\n" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTreeNodeComponent, decorators: [{
            type: Component,
            args: [{ selector: 'nx-tree-node', exportAs: 'nxTreeNode', inputs: ['disabled', 'tabIndex'], host: {
                        '[attr.aria-expanded]': 'isExpanded',
                        '[attr.aria-level]': 'role === "treeitem" ? level : null',
                        '[attr.role]': 'role',
                        class: 'nx-tree__node',
                        '[class.is-expanded]': 'isExpanded',
                        '[attr.tabindex]': '-1',
                    }, providers: [{ provide: CdkTreeNode, useExisting: NxTreeNodeComponent }], template: "<div class=\"nx-tree__node-box\">\n    <ng-content></ng-content>\n</div>\n" }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.CdkTree }, { type: i2.FocusMonitor }]; }, propDecorators: { actionItem: [{
                type: ContentChild,
                args: [NxTreeNodeActionItem]
            }] } });
/**
 * Wrapper for the CdkTree node definition with custom design styles.
 */
export class NxTreeNodeDefDirective extends CdkTreeNodeDef {
}
NxTreeNodeDefDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTreeNodeDefDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NxTreeNodeDefDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxTreeNodeDefDirective, selector: "[nxTreeNodeDef]", inputs: { when: ["nxTreeNodeDefWhen", "when"], data: ["nxTreeNode", "data"] }, providers: [{ provide: CdkTreeNodeDef, useExisting: NxTreeNodeDefDirective }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTreeNodeDefDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxTreeNodeDef]',
                    inputs: ['when: nxTreeNodeDefWhen'],
                    providers: [{ provide: CdkTreeNodeDef, useExisting: NxTreeNodeDefDirective }],
                }]
        }], propDecorators: { data: [{
                type: Input,
                args: ['nxTreeNode']
            }] } });
