// @aposin/ng-aquila/esm2020/tree/toggle.mjs
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { CdkTreeNodeToggle } from '@angular/cdk/tree';
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Wrapper for the CdkTree's toggle with custom design styles.
 */
export class NxTreeNodeToggleDirective extends CdkTreeNodeToggle {
    set recursive(value) {
        this._recursive = coerceBooleanProperty(value);
    }
    get recursive() {
        return this._recursive;
    }
}
NxTreeNodeToggleDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTreeNodeToggleDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NxTreeNodeToggleDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxTreeNodeToggleDirective, selector: "[nxTreeNodeToggle]", inputs: { recursive: ["nxTreeNodeToggleRecursive", "recursive"] }, providers: [{ provide: CdkTreeNodeToggle, useExisting: NxTreeNodeToggleDirective }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTreeNodeToggleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxTreeNodeToggle]',
                    providers: [{ provide: CdkTreeNodeToggle, useExisting: NxTreeNodeToggleDirective }],
                }]
        }], propDecorators: { recursive: [{
                type: Input,
                args: ['nxTreeNodeToggleRecursive']
            }] } });
