// @aposin/ng-aquila/esm2020/tree/tree.module.mjs
import { CdkTreeModule } from '@angular/cdk/tree';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxTreeNodeActionItem } from './action-item.directive';
import { NxTreeNodeComponent, NxTreeNodeDefDirective } from './node';
import { NxTreeNodeOutletDirective } from './outlet';
import { NxTreeNodePaddingDirective } from './padding';
import { NxTreeNodeToggleDirective } from './toggle';
import { NxTreeComponent } from './tree.component';
import * as i0 from "@angular/core";
export class NxTreeModule {
}
NxTreeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTreeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxTreeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTreeModule, declarations: [NxTreeComponent,
        NxTreeNodeComponent,
        NxTreeNodeDefDirective,
        NxTreeNodePaddingDirective,
        NxTreeNodeToggleDirective,
        NxTreeNodeOutletDirective,
        NxTreeNodeActionItem], imports: [CdkTreeModule, CommonModule], exports: [NxTreeComponent,
        NxTreeNodeComponent,
        NxTreeNodeDefDirective,
        NxTreeNodePaddingDirective,
        NxTreeNodeToggleDirective,
        NxTreeNodeOutletDirective,
        NxTreeNodeActionItem] });
NxTreeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTreeModule, imports: [[CdkTreeModule, CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTreeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CdkTreeModule, CommonModule],
                    declarations: [
                        NxTreeComponent,
                        NxTreeNodeComponent,
                        NxTreeNodeDefDirective,
                        NxTreeNodePaddingDirective,
                        NxTreeNodeToggleDirective,
                        NxTreeNodeOutletDirective,
                        NxTreeNodeActionItem,
                    ],
                    exports: [
                        NxTreeComponent,
                        NxTreeNodeComponent,
                        NxTreeNodeDefDirective,
                        NxTreeNodePaddingDirective,
                        NxTreeNodeToggleDirective,
                        NxTreeNodeOutletDirective,
                        NxTreeNodeActionItem,
                    ],
                }]
        }] });
