// @aposin/ng-aquila/esm2020/table/expandable/expandable-table.directive.mjs
import { ContentChildren, Directive } from '@angular/core';
import { BehaviorSubject, combineLatest, Subject } from 'rxjs';
import { distinctUntilChanged, flatMap, map, startWith, takeUntil } from 'rxjs/operators';
import { NxExpandableTableRowComponent } from './expandable-table-row.component';
import * as i0 from "@angular/core";
/**
 * The nxExpandableTable adds row expansion functionality to the nxTable.
 * It provides the logic to expand/collapse all expandable rows simultanously.
 */
export class NxExpandableTableDirective {
    constructor() {
        this.expanded = new BehaviorSubject(false);
        this._destroyed = new Subject();
    }
    ngAfterViewInit() {
        this.rows.changes
            .pipe(startWith(this.rows), flatMap((rows) => combineLatest(rows.map(row => row.expanded))), map((values) => values.reduce((a, x) => a && x, true)), distinctUntilChanged(), takeUntil(this._destroyed))
            .subscribe(rowsOpen => {
            this.expanded.next(rowsOpen);
        });
    }
    ngOnDestroy() {
        this._destroyed.next();
        this._destroyed.complete();
    }
    /**
     * Toggles the expandable rows. If all rows are open it will close them, otherwise it will open them.
     */
    toggle() {
        if (this.expanded.value) {
            this.close();
        }
        else {
            this.expand();
        }
    }
    /**
     * Expands all expandable rows.
     */
    expand() {
        this.rows.forEach(row => row.expanded.next(true));
    }
    /**
     * Closes all expandable rows.
     */
    close() {
        this.rows.forEach(row => row.expanded.next(false));
    }
}
NxExpandableTableDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxExpandableTableDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxExpandableTableDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxExpandableTableDirective, selector: "table[nxExpandableTable]", queries: [{ propertyName: "rows", predicate: NxExpandableTableRowComponent, descendants: true }], exportAs: ["nxExpandableTable"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxExpandableTableDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'table[nxExpandableTable]',
                    exportAs: 'nxExpandableTable',
                }]
        }], propDecorators: { rows: [{
                type: ContentChildren,
                args: [NxExpandableTableRowComponent, { descendants: true }]
            }] } });
