// @aposin/ng-aquila/esm2020/table/expandable/expandable-table-cell.component.mjs
import { animate, state, style, transition, trigger } from '@angular/animations';
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "./expandable-table-row.component";
/**
 * This is an expandable table cell.
 * Only works in conjunction with the `[nxExpandableTableRow]` component.
 */
export class NxExpandableTableCellComponent {
    constructor(_row, _cdr) {
        this._cdr = _cdr;
        this._open = false;
        this._destroyed = new Subject();
        _row.expanded.pipe(takeUntil(this._destroyed)).subscribe(open => {
            this._open = open;
            this._cdr.markForCheck();
        });
    }
    ngOnDestroy() {
        this._destroyed.next();
        this._destroyed.complete();
    }
}
NxExpandableTableCellComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxExpandableTableCellComponent, deps: [{ token: i1.NxExpandableTableRowComponent }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
NxExpandableTableCellComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxExpandableTableCellComponent, selector: "td[nxExpandableTableCell]", host: { classAttribute: "nx-expandable-table-cell" }, ngImport: i0, template: "<div [@openClose]=\"_open ? 'open' : 'closed'\" class=\"nx-expandable-table-cell__content\">\n    <div class=\"nx-expandable-table-cell__wrapper\">\n        <ng-content></ng-content>\n    </div>\n</div>\n", animations: [
        trigger('openClose', [
            state('open', style({
                height: '*',
                visibility: 'visible',
            })),
            state('closed, void', style({
                height: '0',
                visibility: 'hidden', // visibility and height to toggle instead of display: none to prevent cell width jumping
            })),
            transition('closed <=> open, void => closed', [animate('225ms cubic-bezier(0.4,0.0,0.2,1)')]),
        ]),
    ], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxExpandableTableCellComponent, decorators: [{
            type: Component,
            args: [{ selector: 'td[nxExpandableTableCell]', host: {
                        class: 'nx-expandable-table-cell',
                    }, animations: [
                        trigger('openClose', [
                            state('open', style({
                                height: '*',
                                visibility: 'visible',
                            })),
                            state('closed, void', style({
                                height: '0',
                                visibility: 'hidden', // visibility and height to toggle instead of display: none to prevent cell width jumping
                            })),
                            transition('closed <=> open, void => closed', [animate('225ms cubic-bezier(0.4,0.0,0.2,1)')]),
                        ]),
                    ], changeDetection: ChangeDetectionStrategy.OnPush, template: "<div [@openClose]=\"_open ? 'open' : 'closed'\" class=\"nx-expandable-table-cell__content\">\n    <div class=\"nx-expandable-table-cell__wrapper\">\n        <ng-content></ng-content>\n    </div>\n</div>\n" }]
        }], ctorParameters: function () { return [{ type: i1.NxExpandableTableRowComponent }, { type: i0.ChangeDetectorRef }]; } });
