// @aposin/ng-aquila/esm2020/table/expandable/toggle-button.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewChild } from '@angular/core';
import { Subscription } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/a11y";
import * as i2 from "@aposin/ng-aquila/icon";
/**
 * This toggle button can be clicked and toggles expandable elements.
 * e.g. it can toggle a expandable table row or anything else that implements the `NxExpandable` interface.
 */
export class NxToggleButtonComponent {
    constructor(_cdr, _focusMonitor) {
        this._cdr = _cdr;
        this._focusMonitor = _focusMonitor;
        this._expanded = false;
        this._ariaLabel = '';
        this._subscription = Subscription.EMPTY;
    }
    /**
     * This is the expandable target that will be toggled when the user clicks the button.
     */
    set target(value) {
        this._target = value;
        this._subscription?.unsubscribe();
        if (this._target) {
            this._subscription = this._target.expanded.subscribe(expanded => {
                this._expanded = expanded;
                this._cdr.markForCheck();
            });
        }
    }
    set ariaLabel(value) {
        this._ariaLabel = value;
        this._cdr.markForCheck();
    }
    ngAfterViewInit() {
        this._focusMonitor.monitor(this._buttonElement);
    }
    _onClick() {
        if (this._target) {
            this._target.toggle();
        }
    }
    ngOnDestroy() {
        this._subscription?.unsubscribe();
        this._focusMonitor.stopMonitoring(this._buttonElement);
    }
}
NxToggleButtonComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxToggleButtonComponent, deps: [{ token: i0.ChangeDetectorRef }, { token: i1.FocusMonitor }], target: i0.ɵɵFactoryTarget.Component });
NxToggleButtonComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxToggleButtonComponent, selector: "nx-toggle-button", inputs: { target: "target", ariaLabel: "ariaLabel" }, viewQueries: [{ propertyName: "_buttonElement", first: true, predicate: ["button"], descendants: true }], ngImport: i0, template: "<button #button (click)=\"_onClick()\" class=\"nx-toggle-button\" [attr.aria-label]=\"_ariaLabel\" [attr.aria-expanded]=\"_expanded\" [class.is-expanded]=\"_expanded\">\n    <nx-icon name=\"chevron-down\" aria-hidden=\"true\" size=\"s\" class=\"nx-toggle-button__icon\"></nx-icon>\n</button>\n", styles: [":host{display:block}.nx-toggle-button{display:block;cursor:pointer;-webkit-appearance:none;appearance:none;border:none;outline:none;background:none;padding:0;line-height:24px;height:24px}.nx-toggle-button .nx-toggle-button__icon{transform:rotate(0);transition:transform .2s ease;color:var(--expandable-table-arrow-icon-color)}.nx-toggle-button.is-expanded .nx-toggle-button__icon{transform:rotate(-180deg)}.nx-toggle-button::-moz-focus-inner{border:0}.nx-toggle-button.cdk-keyboard-focused{border-radius:4px;box-shadow:var(--focus-box-shadow)}@media screen and (-ms-high-contrast: active),(forced-colors: active){.nx-toggle-button.cdk-keyboard-focused{box-shadow:0 0 0 2px background,0 0 0 6px windowText;outline:4px solid CanvasText;outline-offset:2px}}\n"], components: [{ type: i2.NxIconComponent, selector: "nx-icon", inputs: ["name", "outline", "fill", "size", "font"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxToggleButtonComponent, decorators: [{
            type: Component,
            args: [{ selector: 'nx-toggle-button', changeDetection: ChangeDetectionStrategy.OnPush, template: "<button #button (click)=\"_onClick()\" class=\"nx-toggle-button\" [attr.aria-label]=\"_ariaLabel\" [attr.aria-expanded]=\"_expanded\" [class.is-expanded]=\"_expanded\">\n    <nx-icon name=\"chevron-down\" aria-hidden=\"true\" size=\"s\" class=\"nx-toggle-button__icon\"></nx-icon>\n</button>\n", styles: [":host{display:block}.nx-toggle-button{display:block;cursor:pointer;-webkit-appearance:none;appearance:none;border:none;outline:none;background:none;padding:0;line-height:24px;height:24px}.nx-toggle-button .nx-toggle-button__icon{transform:rotate(0);transition:transform .2s ease;color:var(--expandable-table-arrow-icon-color)}.nx-toggle-button.is-expanded .nx-toggle-button__icon{transform:rotate(-180deg)}.nx-toggle-button::-moz-focus-inner{border:0}.nx-toggle-button.cdk-keyboard-focused{border-radius:4px;box-shadow:var(--focus-box-shadow)}@media screen and (-ms-high-contrast: active),(forced-colors: active){.nx-toggle-button.cdk-keyboard-focused{box-shadow:0 0 0 2px background,0 0 0 6px windowText;outline:4px solid CanvasText;outline-offset:2px}}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.ChangeDetectorRef }, { type: i1.FocusMonitor }]; }, propDecorators: { _buttonElement: [{
                type: ViewChild,
                args: ['button']
            }], target: [{
                type: Input
            }], ariaLabel: [{
                type: Input
            }] } });
