// @aposin/ng-aquila/fesm2015/aposin-ng-aquila-toolbar.mjs
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, NgModule } from '@angular/core';

class NxToolbarComponent {
}
NxToolbarComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxToolbarComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
NxToolbarComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxToolbarComponent, selector: "nx-toolbar", ngImport: i0, template: '<ng-content></ng-content>', isInline: true, styles: [":host{height:48px;border-bottom:1px solid;border-bottom-color:var(--toolbar-border-bottom-color);background:var(--toolbar-background);display:flex;box-sizing:border-box;padding:0 32px;width:100%;flex-direction:row;align-items:center;white-space:nowrap}@media screen and (-ms-high-contrast: active){:host{border:1px solid windowText}}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxToolbarComponent, decorators: [{
            type: Component,
            args: [{ selector: 'nx-toolbar', template: '<ng-content></ng-content>', changeDetection: ChangeDetectionStrategy.OnPush, styles: [":host{height:48px;border-bottom:1px solid;border-bottom-color:var(--toolbar-border-bottom-color);background:var(--toolbar-background);display:flex;box-sizing:border-box;padding:0 32px;width:100%;flex-direction:row;align-items:center;white-space:nowrap}@media screen and (-ms-high-contrast: active){:host{border:1px solid windowText}}\n"] }]
        }] });

class NxToolbarModule {
}
NxToolbarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxToolbarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxToolbarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxToolbarModule, declarations: [NxToolbarComponent], exports: [NxToolbarComponent] });
NxToolbarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxToolbarModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxToolbarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxToolbarComponent],
                    exports: [NxToolbarComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NxToolbarComponent, NxToolbarModule };
