// @aposin/ng-aquila/fesm2015/aposin-ng-aquila-badge.mjs
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Input, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

class NxBadgeComponent {
    constructor(_cdr) {
        this._cdr = _cdr;
        this._vibrant = false;
    }
    /** Sets the class name for the badge element.  */
    set type(value) {
        if (value !== this._type) {
            this._type = value; // TODO properly coerce input value
            this._cdr.markForCheck();
        }
    }
    get type() {
        return this._type;
    }
    /** Change badge style to vibrant. */
    set vibrant(value) {
        const newValue = coerceBooleanProperty(value);
        if (value !== this._vibrant) {
            this._vibrant = newValue;
            this._cdr.markForCheck();
        }
    }
    get vibrant() {
        return this._vibrant;
    }
}
NxBadgeComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBadgeComponent, deps: [{ token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
NxBadgeComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxBadgeComponent, selector: "nx-badge", inputs: { type: "type", vibrant: "vibrant" }, host: { properties: { "class.nx-badge--active": "type === \"active\"", "class.nx-badge--positive": "type === \"positive\"", "class.nx-badge--critical": "type === \"critical\"", "class.nx-badge--negative": "type === \"negative\"", "class.nx-badge--vibrant": "vibrant" } }, ngImport: i0, template: '<ng-content></ng-content>', isInline: true, styles: [":host{font-size:var(--badge-font-size);line-height:var(--badge-line-height);font-weight:var(--badge-font-weight);letter-spacing:var(--badge-letter-spacing);display:inline-block;padding:0 8px;white-space:nowrap;border-radius:12px;min-width:24px}@media screen and (-ms-high-contrast: active){:host{border:1px solid windowText}}:host{color:var(--badge-color);background-color:var(--badge-background-color)}:host.nx-badge--active{background-color:var(--badge-active-background-color);color:var(--badge-active-color)}:host.nx-badge--positive{background-color:var(--badge-positive-background-color);color:var(--badge-positive-color)}:host.nx-badge--critical{background-color:var(--badge-critical-background-color);color:var(--badge-critical-color)}:host.nx-badge--negative{background-color:var(--badge-negative-background-color);color:var(--badge-negative-color)}:host.nx-badge--vibrant{color:var(--badge-vibrant-color);background-color:var(--badge-vibrant-background-color)}:host.nx-badge--vibrant.nx-badge--active{background-color:var(--badge-vibrant-active-background-color);color:var(--badge-vibrant-active-color)}:host.nx-badge--vibrant.nx-badge--positive{background-color:var(--badge-vibrant-positive-background-color);color:var(--badge-vibrant-positive-color)}:host.nx-badge--vibrant.nx-badge--critical{background-color:var(--badge-vibrant-critical-background-color);color:var(--badge-vibrant-critical-color)}:host.nx-badge--vibrant.nx-badge--negative{background-color:var(--badge-vibrant-negative-background-color);color:var(--badge-vibrant-negative-color)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBadgeComponent, decorators: [{
            type: Component,
            args: [{ selector: 'nx-badge', template: '<ng-content></ng-content>', changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        '[class.nx-badge--active]': 'type === "active"',
                        '[class.nx-badge--positive]': 'type === "positive"',
                        '[class.nx-badge--critical]': 'type === "critical"',
                        '[class.nx-badge--negative]': 'type === "negative"',
                        '[class.nx-badge--vibrant]': 'vibrant',
                    }, styles: [":host{font-size:var(--badge-font-size);line-height:var(--badge-line-height);font-weight:var(--badge-font-weight);letter-spacing:var(--badge-letter-spacing);display:inline-block;padding:0 8px;white-space:nowrap;border-radius:12px;min-width:24px}@media screen and (-ms-high-contrast: active){:host{border:1px solid windowText}}:host{color:var(--badge-color);background-color:var(--badge-background-color)}:host.nx-badge--active{background-color:var(--badge-active-background-color);color:var(--badge-active-color)}:host.nx-badge--positive{background-color:var(--badge-positive-background-color);color:var(--badge-positive-color)}:host.nx-badge--critical{background-color:var(--badge-critical-background-color);color:var(--badge-critical-color)}:host.nx-badge--negative{background-color:var(--badge-negative-background-color);color:var(--badge-negative-color)}:host.nx-badge--vibrant{color:var(--badge-vibrant-color);background-color:var(--badge-vibrant-background-color)}:host.nx-badge--vibrant.nx-badge--active{background-color:var(--badge-vibrant-active-background-color);color:var(--badge-vibrant-active-color)}:host.nx-badge--vibrant.nx-badge--positive{background-color:var(--badge-vibrant-positive-background-color);color:var(--badge-vibrant-positive-color)}:host.nx-badge--vibrant.nx-badge--critical{background-color:var(--badge-vibrant-critical-background-color);color:var(--badge-vibrant-critical-color)}:host.nx-badge--vibrant.nx-badge--negative{background-color:var(--badge-vibrant-negative-background-color);color:var(--badge-vibrant-negative-color)}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.ChangeDetectorRef }]; }, propDecorators: { type: [{
                type: Input
            }], vibrant: [{
                type: Input
            }] } });

class NxBadgeModule {
}
NxBadgeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBadgeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxBadgeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBadgeModule, declarations: [NxBadgeComponent], imports: [CommonModule], exports: [NxBadgeComponent] });
NxBadgeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBadgeModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBadgeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [NxBadgeComponent],
                    exports: [NxBadgeComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NxBadgeComponent, NxBadgeModule };
