// @aposin/ng-aquila/fesm2015/aposin-ng-aquila-image.mjs
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Input, Directive, NgModule } from '@angular/core';

const DEFAULT_SIZE = 'auto';
class NxFigureComponent {
    constructor() {
        /** @docs-private */
        this.rounded = false;
        /** @docs-private */
        this.size = DEFAULT_SIZE;
    }
    /**
     * Sets the type of the visual appearance of the image.
     * The default value  is 'auto'.
     */
    set classNames(value) {
        var _a, _b;
        if (this._classNames === value) {
            return;
        }
        this._classNames = value; // TODO properly coerce input value
        // TODO kick null safe-guards after setter value or any calling input values are properly coerced as string
        const sizeRegex = /^(auto|1by1|1dot8by1|1dot2by1|1by1dot1|2dot6by1)$/;
        const [size = null] = ((_a = this._classNames) === null || _a === void 0 ? void 0 : _a.match(sizeRegex)) || [DEFAULT_SIZE];
        this.size = size;
        this.rounded = !!((_b = this._classNames) === null || _b === void 0 ? void 0 : _b.match(/rounded/));
    }
    get classNames() {
        return this._classNames;
    }
}
NxFigureComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFigureComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
NxFigureComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxFigureComponent, selector: "figure[nxFigure]", inputs: { classNames: ["nxFigure", "classNames"] }, host: { properties: { "class.nx-image": "true", "class.nx-image--auto": "size === \"auto\" && !rounded", "class.nx-image--1by1": "size === \"1by1\"", "class.nx-image--1dot8by1": "size === \"1dot8by1\"", "class.nx-image--1dot2by1": "size === \"1dot2by1\"", "class.nx-image--1by1dot1": "size === \"1by1dot1\"", "class.nx-image--2dot6by1": "size === \"2dot6by1\"", "class.nx-image--rounded": "rounded" } }, ngImport: i0, template: '<ng-content></ng-content>', isInline: true, styles: [":host{display:block;position:relative;line-height:0}:host ::ng-deep img{display:block;font-family:\"object-fit: cover;\";-o-object-fit:cover;object-fit:cover;height:100%;width:100%}:host(.nx-image--auto) ::ng-deep img{height:auto}:host(:not(.nx-image--auto)) ::ng-deep img{position:absolute;top:0;left:0}[dir=rtl] :host(:not(.nx-image--auto)) ::ng-deep img{right:0;left:auto}:host(.nx-image--rounded){padding-top:100%}:host(.nx-image--rounded) ::ng-deep img{border-radius:50%;bottom:0;left:0;position:absolute;right:0;top:0}:host(.nx-image--1by1){padding-top:100%}:host(.nx-image--1dot8by1){padding-top:55.5555555556%}:host(.nx-image--1dot2by1){padding-top:83.3333333333%}:host(.nx-image--1by1dot1){padding-top:110%}:host(.nx-image--2dot6by1){padding-top:38.4615384615%}:host(.nx-image--1by1dot4){padding-top:140%}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFigureComponent, decorators: [{
            type: Component,
            args: [{ selector: 'figure[nxFigure]', template: '<ng-content></ng-content>', changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        '[class.nx-image]': 'true',
                        '[class.nx-image--auto]': 'size === "auto" && !rounded',
                        '[class.nx-image--1by1]': 'size === "1by1"',
                        '[class.nx-image--1dot8by1]': 'size === "1dot8by1"',
                        '[class.nx-image--1dot2by1]': 'size === "1dot2by1"',
                        '[class.nx-image--1by1dot1]': 'size === "1by1dot1"',
                        '[class.nx-image--2dot6by1]': 'size === "2dot6by1"',
                        '[class.nx-image--rounded]': 'rounded',
                    }, styles: [":host{display:block;position:relative;line-height:0}:host ::ng-deep img{display:block;font-family:\"object-fit: cover;\";-o-object-fit:cover;object-fit:cover;height:100%;width:100%}:host(.nx-image--auto) ::ng-deep img{height:auto}:host(:not(.nx-image--auto)) ::ng-deep img{position:absolute;top:0;left:0}[dir=rtl] :host(:not(.nx-image--auto)) ::ng-deep img{right:0;left:auto}:host(.nx-image--rounded){padding-top:100%}:host(.nx-image--rounded) ::ng-deep img{border-radius:50%;bottom:0;left:0;position:absolute;right:0;top:0}:host(.nx-image--1by1){padding-top:100%}:host(.nx-image--1dot8by1){padding-top:55.5555555556%}:host(.nx-image--1dot2by1){padding-top:83.3333333333%}:host(.nx-image--1by1dot1){padding-top:110%}:host(.nx-image--2dot6by1){padding-top:38.4615384615%}:host(.nx-image--1by1dot4){padding-top:140%}\n"] }]
        }], propDecorators: { classNames: [{
                type: Input,
                args: ['nxFigure']
            }] } });

class NxImageDirective {
    constructor() {
        console.warn('Directive nxImg which was used together with nxFigure ' +
            'is now deprecated and not required anymore. Please remove it from your ' +
            'tag. Check the documentation for more details.');
    }
}
NxImageDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxImageDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxImageDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxImageDirective, selector: "img[nxImg]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxImageDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'img[nxImg]',
                }]
        }], ctorParameters: function () { return []; } });

class NxImageModule {
}
NxImageModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxImageModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxImageModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxImageModule, declarations: [NxFigureComponent, NxImageDirective], exports: [NxFigureComponent, NxImageDirective] });
NxImageModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxImageModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxImageModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxFigureComponent, NxImageDirective],
                    exports: [NxFigureComponent, NxImageDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NxFigureComponent, NxImageDirective, NxImageModule };
