// @aposin/ng-aquila/esm2020/context-menu/context-menu.module.mjs
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxContextMenuComponent } from './context-menu.component';
import { NxContextMenuContentDirective } from './context-menu-content.directive';
import { NxContextMenuHeaderComponent } from './context-menu-header.component';
import { NxContextMenuItemComponent, NxContextMenuItemWrapComponent } from './context-menu-item.component';
import { NX_CONTEXT_MENU_SCROLL_STRATEGY_PROVIDER, NxContextMenuTriggerDirective } from './context-menu-trigger.directive';
import * as i0 from "@angular/core";
const EXPORTED_MODULES = [
    NxContextMenuComponent,
    NxContextMenuContentDirective,
    NxContextMenuItemComponent,
    NxContextMenuItemWrapComponent,
    NxContextMenuTriggerDirective,
    NxContextMenuHeaderComponent,
];
export class NxContextMenuModule {
}
NxContextMenuModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxContextMenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxContextMenuModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxContextMenuModule, declarations: [NxContextMenuComponent,
        NxContextMenuContentDirective,
        NxContextMenuItemComponent,
        NxContextMenuItemWrapComponent,
        NxContextMenuTriggerDirective,
        NxContextMenuHeaderComponent], imports: [CommonModule, OverlayModule, NxIconModule], exports: [NxContextMenuComponent,
        NxContextMenuContentDirective,
        NxContextMenuItemComponent,
        NxContextMenuItemWrapComponent,
        NxContextMenuTriggerDirective,
        NxContextMenuHeaderComponent] });
NxContextMenuModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxContextMenuModule, providers: [NX_CONTEXT_MENU_SCROLL_STRATEGY_PROVIDER], imports: [[CommonModule, OverlayModule, NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxContextMenuModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, OverlayModule, NxIconModule],
                    exports: EXPORTED_MODULES,
                    declarations: EXPORTED_MODULES,
                    providers: [NX_CONTEXT_MENU_SCROLL_STRATEGY_PROVIDER],
                }]
        }] });
