// @aposin/ng-aquila/esm2020/spinner/spinner.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxSpinnerComponent } from './spinner.component';
import * as i0 from "@angular/core";
export class NxSpinnerModule {
}
NxSpinnerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSpinnerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxSpinnerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSpinnerModule, declarations: [NxSpinnerComponent], imports: [CommonModule], exports: [NxSpinnerComponent] });
NxSpinnerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSpinnerModule, providers: [], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSpinnerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxSpinnerComponent],
                    imports: [CommonModule],
                    exports: [NxSpinnerComponent],
                    bootstrap: [],
                    providers: [],
                }]
        }] });
