// @aposin/ng-aquila/esm2020/header/header.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxHeaderActionsDirective, NxHeaderAppTitleDirective, NxHeaderBrandDirective, NxHeaderComponent, NxHeaderLinkComponent, NxHeaderNavigationComponent, NxHeaderNavigationItemDirective, NxHeaderRowDirective, } from './header.component';
import * as i0 from "@angular/core";
const EXPORTS = [
    NxHeaderComponent,
    NxHeaderLinkComponent,
    NxHeaderActionsDirective,
    NxHeaderNavigationComponent,
    NxHeaderBrandDirective,
    NxHeaderRowDirective,
    NxHeaderNavigationItemDirective,
    NxHeaderAppTitleDirective,
];
export class NxHeaderModule {
}
NxHeaderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxHeaderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderModule, declarations: [NxHeaderComponent,
        NxHeaderLinkComponent,
        NxHeaderActionsDirective,
        NxHeaderNavigationComponent,
        NxHeaderBrandDirective,
        NxHeaderRowDirective,
        NxHeaderNavigationItemDirective,
        NxHeaderAppTitleDirective], imports: [CommonModule], exports: [NxHeaderComponent,
        NxHeaderLinkComponent,
        NxHeaderActionsDirective,
        NxHeaderNavigationComponent,
        NxHeaderBrandDirective,
        NxHeaderRowDirective,
        NxHeaderNavigationItemDirective,
        NxHeaderAppTitleDirective] });
NxHeaderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: EXPORTS,
                    exports: EXPORTS,
                    imports: [CommonModule],
                }]
        }] });
