// @aposin/ng-aquila/fesm2020/aposin-ng-aquila-header.mjs
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import * as i0 from '@angular/core';
import { Directive, Component, ChangeDetectionStrategy, ContentChildren, Input, NgModule } from '@angular/core';
import * as i1 from '@angular/cdk/a11y';
import { CommonModule } from '@angular/common';

/** This directive defines a header row within the `<nx-header>` component. */
class NxHeaderRowDirective {
}
NxHeaderRowDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderRowDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxHeaderRowDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxHeaderRowDirective, selector: "nx-header-row", host: { classAttribute: "nx-header__row" }, exportAs: ["NxHeaderRow"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderRowDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'nx-header-row',
                    exportAs: 'NxHeaderRow',
                    host: {
                        class: 'nx-header__row',
                    },
                }]
        }] });
/** The `<nx-header>` component is the main application header. */
class NxHeaderComponent {
}
NxHeaderComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
NxHeaderComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxHeaderComponent, selector: "nx-header, [nx-header]", host: { properties: { "class.nx-header--single-row": "_headerRows.length === 0" }, classAttribute: "nx-header" }, queries: [{ propertyName: "_headerRows", predicate: NxHeaderRowDirective }], ngImport: i0, template: "<ng-content select=\"nx-header-row\"></ng-content>\n<ng-content></ng-content>\n", styles: [":host{font-size:var(--header-font-size);line-height:var(--header-line-height);font-weight:var(--header-font-weight);letter-spacing:var(--header-letter-spacing)}:host(.nx-header:not(.nx-header--single-row)){display:flex;flex-direction:column}:host(.nx-header.nx-header--single-row),::ng-deep .nx-header__row{display:flex;align-items:center;padding:1px 32px 0;border-bottom:1px solid var(--header-border-color);height:60px;justify-content:space-between;background:var(--header-background-color)}@media screen and (-ms-high-contrast: active){:host(.nx-header.nx-header--single-row),::ng-deep .nx-header__row{background-color:window}}::ng-deep .nx-header__link{border-bottom:4px solid transparent;padding-top:4px;text-decoration:none;display:block;margin-right:32px;white-space:nowrap}:host-context([dir=rtl]) ::ng-deep .nx-header__link{margin-right:0;margin-right:initial;margin-left:32px}::ng-deep .nx-header__link:after{display:block;content:attr(text-content);font-weight:var(--header-link-active-font-weight);height:1px;overflow:hidden;visibility:hidden}::ng-deep .nx-header__link.is-active{color:var(--header-link-active-color);border-color:currentColor}::ng-deep .nx-header__link.is-active .nx-header__link-title{font-weight:var(--header-link-active-font-weight)}::ng-deep .nx-header__link:hover,::ng-deep .nx-header__link:active{color:var(--header-link-active-color);border-color:var(--header-link-active-color)}@media screen and (forced-colors: active){::ng-deep .nx-header__link{color:LinkText!important}}@media screen and (-ms-high-contrast: active){::ng-deep .nx-header__link{border-bottom:none}::ng-deep .nx-header__link.is-active{border-bottom:4px solid}}::ng-deep .nx-header__link-title{line-height:0}::ng-deep .nx-header__brand{margin-right:32px;display:flex;align-items:stretch}:host-context([dir=rtl]) ::ng-deep .nx-header__brand{margin-right:0;margin-right:initial;margin-left:32px}::ng-deep .nx-header__actions{display:flex;align-items:center;margin-left:24px}::ng-deep .nx-header__actions.nx-header__actions--show-separator{padding-left:16px;border-left-width:1px;border-left-style:solid;border-left-color:var(--header-border-color)}:host-context([dir=rtl]) ::ng-deep .nx-header__actions{margin-right:24px;margin-left:0;margin-left:initial}:host-context([dir=rtl]) ::ng-deep .nx-header__actions.nx-header__actions--show-separator{padding-right:16px;padding-left:0;padding-left:initial;border-right-width:1px;border-right-style:solid;border-right-color:var(--header-border-color);border-left:none}::ng-deep .nx-header__link.cdk-keyboard-focused{box-shadow:var(--focus-box-shadow);border-radius:4px}@media screen and (-ms-high-contrast: active),(forced-colors: active){::ng-deep .nx-header__link.cdk-keyboard-focused{box-shadow:0 0 0 2px background,0 0 0 6px windowText;outline:4px solid CanvasText;outline-offset:2px}}::ng-deep a:focus{outline:none}::ng-deep .nx-header__app-title{display:flex;font-size:16px;font-weight:700;align-items:center;padding-left:16px;margin-left:16px;white-space:nowrap;border-left-width:1px;border-left-style:solid;border-left-color:var(--header-border-color)}:host-context([dir=rtl]) ::ng-deep .nx-header__app-title{padding-right:16px;padding-left:0;padding-left:initial;margin-right:16px;margin-left:0;margin-left:initial;border-right-width:1px;border-right-style:solid;border-right-color:var(--header-border-color);border-left:none}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderComponent, decorators: [{
            type: Component,
            args: [{ selector: 'nx-header, [nx-header]', changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        class: 'nx-header',
                        '[class.nx-header--single-row]': '_headerRows.length === 0',
                    }, template: "<ng-content select=\"nx-header-row\"></ng-content>\n<ng-content></ng-content>\n", styles: [":host{font-size:var(--header-font-size);line-height:var(--header-line-height);font-weight:var(--header-font-weight);letter-spacing:var(--header-letter-spacing)}:host(.nx-header:not(.nx-header--single-row)){display:flex;flex-direction:column}:host(.nx-header.nx-header--single-row),::ng-deep .nx-header__row{display:flex;align-items:center;padding:1px 32px 0;border-bottom:1px solid var(--header-border-color);height:60px;justify-content:space-between;background:var(--header-background-color)}@media screen and (-ms-high-contrast: active){:host(.nx-header.nx-header--single-row),::ng-deep .nx-header__row{background-color:window}}::ng-deep .nx-header__link{border-bottom:4px solid transparent;padding-top:4px;text-decoration:none;display:block;margin-right:32px;white-space:nowrap}:host-context([dir=rtl]) ::ng-deep .nx-header__link{margin-right:0;margin-right:initial;margin-left:32px}::ng-deep .nx-header__link:after{display:block;content:attr(text-content);font-weight:var(--header-link-active-font-weight);height:1px;overflow:hidden;visibility:hidden}::ng-deep .nx-header__link.is-active{color:var(--header-link-active-color);border-color:currentColor}::ng-deep .nx-header__link.is-active .nx-header__link-title{font-weight:var(--header-link-active-font-weight)}::ng-deep .nx-header__link:hover,::ng-deep .nx-header__link:active{color:var(--header-link-active-color);border-color:var(--header-link-active-color)}@media screen and (forced-colors: active){::ng-deep .nx-header__link{color:LinkText!important}}@media screen and (-ms-high-contrast: active){::ng-deep .nx-header__link{border-bottom:none}::ng-deep .nx-header__link.is-active{border-bottom:4px solid}}::ng-deep .nx-header__link-title{line-height:0}::ng-deep .nx-header__brand{margin-right:32px;display:flex;align-items:stretch}:host-context([dir=rtl]) ::ng-deep .nx-header__brand{margin-right:0;margin-right:initial;margin-left:32px}::ng-deep .nx-header__actions{display:flex;align-items:center;margin-left:24px}::ng-deep .nx-header__actions.nx-header__actions--show-separator{padding-left:16px;border-left-width:1px;border-left-style:solid;border-left-color:var(--header-border-color)}:host-context([dir=rtl]) ::ng-deep .nx-header__actions{margin-right:24px;margin-left:0;margin-left:initial}:host-context([dir=rtl]) ::ng-deep .nx-header__actions.nx-header__actions--show-separator{padding-right:16px;padding-left:0;padding-left:initial;border-right-width:1px;border-right-style:solid;border-right-color:var(--header-border-color);border-left:none}::ng-deep .nx-header__link.cdk-keyboard-focused{box-shadow:var(--focus-box-shadow);border-radius:4px}@media screen and (-ms-high-contrast: active),(forced-colors: active){::ng-deep .nx-header__link.cdk-keyboard-focused{box-shadow:0 0 0 2px background,0 0 0 6px windowText;outline:4px solid CanvasText;outline-offset:2px}}::ng-deep a:focus{outline:none}::ng-deep .nx-header__app-title{display:flex;font-size:16px;font-weight:700;align-items:center;padding-left:16px;margin-left:16px;white-space:nowrap;border-left-width:1px;border-left-style:solid;border-left-color:var(--header-border-color)}:host-context([dir=rtl]) ::ng-deep .nx-header__app-title{padding-right:16px;padding-left:0;padding-left:initial;margin-right:16px;margin-left:0;margin-left:initial;border-right-width:1px;border-right-style:solid;border-right-color:var(--header-border-color);border-left:none}\n"] }]
        }], propDecorators: { _headerRows: [{
                type: ContentChildren,
                args: [NxHeaderRowDirective]
            }] } });
/** This directive defines the header brand section within the `<nx-header>` component. */
class NxHeaderBrandDirective {
}
NxHeaderBrandDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderBrandDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxHeaderBrandDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxHeaderBrandDirective, selector: "nx-header-brand", host: { classAttribute: "nx-header__brand" }, exportAs: ["NxHeaderBrand"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderBrandDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'nx-header-brand',
                    exportAs: 'NxHeaderBrand',
                    host: {
                        class: 'nx-header__brand',
                    },
                }]
        }] });
/** This directive defines the header actions section within the `<nx-header>` component. */
class NxHeaderActionsDirective {
    constructor() {
        this._showSeparator = false;
    }
    /* Whenever to show the left separator*/
    get showSeparator() {
        return this._showSeparator;
    }
    set showSeparator(value) {
        this._showSeparator = coerceBooleanProperty(value);
    }
}
NxHeaderActionsDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderActionsDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxHeaderActionsDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxHeaderActionsDirective, selector: "nx-header-actions", inputs: { showSeparator: "showSeparator" }, host: { properties: { "class.nx-header__actions--show-separator": "showSeparator" }, classAttribute: "nx-header__actions" }, exportAs: ["NxHeaderActions"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderActionsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'nx-header-actions',
                    exportAs: 'NxHeaderActions',
                    host: {
                        class: 'nx-header__actions',
                        '[class.nx-header__actions--show-separator]': 'showSeparator',
                    },
                }]
        }], propDecorators: { showSeparator: [{
                type: Input
            }] } });
/** This component defines the header navigation section within the `<nx-header>` component. */
class NxHeaderNavigationComponent {
}
NxHeaderNavigationComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderNavigationComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
NxHeaderNavigationComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxHeaderNavigationComponent, selector: "nx-header-navigation", host: { attributes: { "role": "navigation" }, classAttribute: "nx-header__navigation" }, exportAs: ["NxHeaderNavigation"], ngImport: i0, template: '<div role="list" class="nx-header__navigation-items"><ng-content></ng-content></div>', isInline: true, styles: [":host(.nx-header__navigation){flex:1 1 auto}.nx-header__navigation-items{display:flex}\n"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderNavigationComponent, decorators: [{
            type: Component,
            args: [{ selector: 'nx-header-navigation', exportAs: 'NxHeaderNavigation', host: {
                        class: 'nx-header__navigation',
                        role: 'navigation',
                    }, template: '<div role="list" class="nx-header__navigation-items"><ng-content></ng-content></div>', styles: [":host(.nx-header__navigation){flex:1 1 auto}.nx-header__navigation-items{display:flex}\n"] }]
        }] });
/** This directive defines a header navigation item within the `<nx-header-navigation>`. */
class NxHeaderNavigationItemDirective {
}
NxHeaderNavigationItemDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderNavigationItemDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxHeaderNavigationItemDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxHeaderNavigationItemDirective, selector: "nx-header-navigation-item", host: { attributes: { "role": "listitem" }, classAttribute: "nx-header__navigation-item" }, exportAs: ["NxHeaderNavigation"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderNavigationItemDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'nx-header-navigation-item',
                    exportAs: 'NxHeaderNavigation',
                    host: {
                        role: 'listitem',
                        class: 'nx-header__navigation-item',
                    },
                }]
        }] });
/** This component defines a header link within the `<nx-header>` component. */
class NxHeaderLinkComponent {
    constructor(_elementRef, _focusMonitor) {
        this._elementRef = _elementRef;
        this._focusMonitor = _focusMonitor;
        this._focusMonitor.monitor(this._elementRef);
    }
    ngOnDestroy() {
        this._focusMonitor.stopMonitoring(this._elementRef);
    }
    get textContent() {
        return this._elementRef.nativeElement.textContent;
    }
}
NxHeaderLinkComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderLinkComponent, deps: [{ token: i0.ElementRef }, { token: i1.FocusMonitor }], target: i0.ɵɵFactoryTarget.Component });
NxHeaderLinkComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxHeaderLinkComponent, selector: "a[nxHeaderLink]", host: { properties: { "attr.text-content": "textContent" }, classAttribute: "nx-header__link" }, exportAs: ["NxHeaderLink"], ngImport: i0, template: '<span class="nx-header__link-title"><ng-content></ng-content></span>', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderLinkComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'a[nxHeaderLink]',
                    exportAs: 'NxHeaderLink',
                    host: {
                        class: 'nx-header__link',
                        '[attr.text-content]': 'textContent',
                    },
                    template: '<span class="nx-header__link-title"><ng-content></ng-content></span>',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.FocusMonitor }]; } });
/** This directive defines the application title within the `<nx-header>` component. */
class NxHeaderAppTitleDirective {
}
NxHeaderAppTitleDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderAppTitleDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxHeaderAppTitleDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxHeaderAppTitleDirective, selector: "nx-header-app-title", host: { classAttribute: "nx-header__app-title" }, exportAs: ["NxHeaderAppTitle"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderAppTitleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'nx-header-app-title',
                    exportAs: 'NxHeaderAppTitle',
                    host: {
                        class: 'nx-header__app-title',
                    },
                }]
        }] });

const EXPORTS = [
    NxHeaderComponent,
    NxHeaderLinkComponent,
    NxHeaderActionsDirective,
    NxHeaderNavigationComponent,
    NxHeaderBrandDirective,
    NxHeaderRowDirective,
    NxHeaderNavigationItemDirective,
    NxHeaderAppTitleDirective,
];
class NxHeaderModule {
}
NxHeaderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxHeaderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderModule, declarations: [NxHeaderComponent,
        NxHeaderLinkComponent,
        NxHeaderActionsDirective,
        NxHeaderNavigationComponent,
        NxHeaderBrandDirective,
        NxHeaderRowDirective,
        NxHeaderNavigationItemDirective,
        NxHeaderAppTitleDirective], imports: [CommonModule], exports: [NxHeaderComponent,
        NxHeaderLinkComponent,
        NxHeaderActionsDirective,
        NxHeaderNavigationComponent,
        NxHeaderBrandDirective,
        NxHeaderRowDirective,
        NxHeaderNavigationItemDirective,
        NxHeaderAppTitleDirective] });
NxHeaderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxHeaderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: EXPORTS,
                    exports: EXPORTS,
                    imports: [CommonModule],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NxHeaderActionsDirective, NxHeaderAppTitleDirective, NxHeaderBrandDirective, NxHeaderComponent, NxHeaderLinkComponent, NxHeaderModule, NxHeaderNavigationComponent, NxHeaderNavigationItemDirective, NxHeaderRowDirective };
