// @aposin/ng-aquila/fesm2020/aposin-ng-aquila-breadcrumb.mjs
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Input, ContentChildren, NgModule } from '@angular/core';
import { Subject } from 'rxjs';
import { startWith, filter, takeUntil } from 'rxjs/operators';
import * as i1 from '@angular/cdk/a11y';
import * as i2 from '@aposin/ng-aquila/icon';
import { NxIconModule } from '@aposin/ng-aquila/icon';

class NxBreadcrumbItemComponent {
    constructor(_renderer, _elemRef, _focusMonitor) {
        this._renderer = _renderer;
        this._elemRef = _elemRef;
        this._focusMonitor = _focusMonitor;
        this._focusMonitor.monitor(this._elemRef);
    }
    ngOnDestroy() {
        this._focusMonitor.stopMonitoring(this._elemRef);
    }
    /** @docs-private */
    setAsLast() {
        this._renderer.setAttribute(this._elemRef.nativeElement, 'aria-current', 'page');
    }
    /** @docs-private */
    resetAriaLabel() {
        this._renderer.removeAttribute(this._elemRef.nativeElement, 'aria-current');
    }
    // prevent location change when the user clicks on the chevron arrow
    _onIconClick(event) {
        event.stopPropagation();
        event.preventDefault();
    }
}
NxBreadcrumbItemComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbItemComponent, deps: [{ token: i0.Renderer2 }, { token: i0.ElementRef }, { token: i1.FocusMonitor }], target: i0.ɵɵFactoryTarget.Component });
NxBreadcrumbItemComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxBreadcrumbItemComponent, selector: "a[nxBreadcrumbItem]", host: { classAttribute: "nx-breadcrumb-item" }, ngImport: i0, template: "<span class=\"nx-breadcrumb-item__text\"><ng-content></ng-content></span>\n<nx-icon (click)=\"_onIconClick($event)\" class=\"nx-breadcrumb-item__chevron\" aria-hidden=\"true\" name=\"chevron-right\"> </nx-icon>\n", styles: [":host{text-decoration:none;display:flex;align-items:center;cursor:auto;color:var(--breadcrumb-color);font-size:var(--breadcrumb-font-size);line-height:var(--breadcrumb-line-height);font-weight:var(--breadcrumb-font-weight);letter-spacing:var(--breadcrumb-letter-spacing)}.nx-breadcrumb-item__chevron{margin-left:4px}:host-context([dir=rtl]) .nx-breadcrumb-item__chevron{margin-left:0;margin-left:initial;margin-right:4px;transform:rotate(180deg)}.nx-breadcrumb-item__text{cursor:pointer}:host.cdk-keyboard-focused .nx-breadcrumb-item__text{box-shadow:var(--focus-box-shadow);border-radius:4px}@media screen and (-ms-high-contrast: active),(forced-colors: active){:host.cdk-keyboard-focused .nx-breadcrumb-item__text{box-shadow:0 0 0 2px background,0 0 0 6px windowText;outline:4px solid CanvasText;outline-offset:2px}}\n"], components: [{ type: i2.NxIconComponent, selector: "nx-icon", inputs: ["name", "outline", "fill", "size", "font"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbItemComponent, decorators: [{
            type: Component,
            args: [{ selector: 'a[nxBreadcrumbItem]', changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        class: 'nx-breadcrumb-item',
                    }, template: "<span class=\"nx-breadcrumb-item__text\"><ng-content></ng-content></span>\n<nx-icon (click)=\"_onIconClick($event)\" class=\"nx-breadcrumb-item__chevron\" aria-hidden=\"true\" name=\"chevron-right\"> </nx-icon>\n", styles: [":host{text-decoration:none;display:flex;align-items:center;cursor:auto;color:var(--breadcrumb-color);font-size:var(--breadcrumb-font-size);line-height:var(--breadcrumb-line-height);font-weight:var(--breadcrumb-font-weight);letter-spacing:var(--breadcrumb-letter-spacing)}.nx-breadcrumb-item__chevron{margin-left:4px}:host-context([dir=rtl]) .nx-breadcrumb-item__chevron{margin-left:0;margin-left:initial;margin-right:4px;transform:rotate(180deg)}.nx-breadcrumb-item__text{cursor:pointer}:host.cdk-keyboard-focused .nx-breadcrumb-item__text{box-shadow:var(--focus-box-shadow);border-radius:4px}@media screen and (-ms-high-contrast: active),(forced-colors: active){:host.cdk-keyboard-focused .nx-breadcrumb-item__text{box-shadow:0 0 0 2px background,0 0 0 6px windowText;outline:4px solid CanvasText;outline-offset:2px}}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.Renderer2 }, { type: i0.ElementRef }, { type: i1.FocusMonitor }]; } });

class NxBreadcrumbComponent {
    constructor(_cdr) {
        this._cdr = _cdr;
        this._negative = false;
        this._appeareance = 'default';
        this._destroyed = new Subject();
    }
    /**
     * Sets the appearance of the breadcrumb. default: 'default'
     */
    set appearance(value) {
        this._appeareance = value;
        this._cdr.markForCheck();
    }
    get appearance() {
        return this._appeareance;
    }
    /** Whether the component uses the negative styling. */
    set negative(value) {
        this._negative = coerceBooleanProperty(value);
        this._cdr.markForCheck();
    }
    get negative() {
        return this._negative;
    }
    ngAfterContentInit() {
        if (this.breadcrumbItems.length === 0) {
            console.warn('A breadcrumb needs NxBreadcrumbItemComponent children wrapped in <li>!');
        }
        this.breadcrumbItems.changes
            .pipe(startWith(this.breadcrumbItems), filter(items => items.length !== 0), takeUntil(this._destroyed))
            .subscribe(items => {
            this.breadcrumbItems.forEach(item => item.resetAriaLabel());
            this.breadcrumbItems.last.setAsLast();
        });
    }
    ngOnDestroy() {
        this._destroyed.next();
        this._destroyed.complete();
    }
}
NxBreadcrumbComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbComponent, deps: [{ token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
NxBreadcrumbComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxBreadcrumbComponent, selector: "ol[nxBreadcrumb]", inputs: { appearance: "appearance", negative: "negative" }, host: { properties: { "class.is-negative": "negative", "class.is-link": "appearance === \"link\"" } }, queries: [{ propertyName: "breadcrumbItems", predicate: NxBreadcrumbItemComponent, descendants: true }], ngImport: i0, template: "<ng-content></ng-content>\n", styles: [":host{display:flex;list-style:none;height:24px;padding:0}:host(.is-link) ::ng-deep .nx-breadcrumb-item{color:var(--breadcrumb-link-color)}:host(.is-link) ::ng-deep .nx-breadcrumb-item__text:hover{color:var(--breadcrumb-link-hover-color)}:host(.is-link) ::ng-deep .nx-breadcrumb-item__text:active{color:var(--breadcrumb-link-active-color)}::ng-deep li:last-child .nx-breadcrumb-item{font-weight:600;cursor:auto}::ng-deep li:last-child .nx-breadcrumb-item__chevron{display:none}::ng-deep li+li .nx-breadcrumb-item{margin-left:4px}:host-context([dir=rtl]) ::ng-deep li+li .nx-breadcrumb-item{margin-left:0;margin-left:initial;margin-right:4px}:host(.is-negative) ::ng-deep .nx-breadcrumb-item{color:var(--negative)}@media screen and (-ms-high-contrast: active){::ng-deep .nx-breadcrumb-item__chevron{color:windowText}}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ol[nxBreadcrumb]', changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        '[class.is-negative]': 'negative',
                        '[class.is-link]': 'appearance === "link"',
                    }, template: "<ng-content></ng-content>\n", styles: [":host{display:flex;list-style:none;height:24px;padding:0}:host(.is-link) ::ng-deep .nx-breadcrumb-item{color:var(--breadcrumb-link-color)}:host(.is-link) ::ng-deep .nx-breadcrumb-item__text:hover{color:var(--breadcrumb-link-hover-color)}:host(.is-link) ::ng-deep .nx-breadcrumb-item__text:active{color:var(--breadcrumb-link-active-color)}::ng-deep li:last-child .nx-breadcrumb-item{font-weight:600;cursor:auto}::ng-deep li:last-child .nx-breadcrumb-item__chevron{display:none}::ng-deep li+li .nx-breadcrumb-item{margin-left:4px}:host-context([dir=rtl]) ::ng-deep li+li .nx-breadcrumb-item{margin-left:0;margin-left:initial;margin-right:4px}:host(.is-negative) ::ng-deep .nx-breadcrumb-item{color:var(--negative)}@media screen and (-ms-high-contrast: active){::ng-deep .nx-breadcrumb-item__chevron{color:windowText}}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.ChangeDetectorRef }]; }, propDecorators: { appearance: [{
                type: Input
            }], negative: [{
                type: Input
            }], breadcrumbItems: [{
                type: ContentChildren,
                args: [NxBreadcrumbItemComponent, { descendants: true }]
            }] } });

class NxBreadcrumbModule {
}
NxBreadcrumbModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxBreadcrumbModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbModule, declarations: [NxBreadcrumbItemComponent, NxBreadcrumbComponent], imports: [NxIconModule], exports: [NxBreadcrumbItemComponent, NxBreadcrumbComponent] });
NxBreadcrumbModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbModule, imports: [[NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxBreadcrumbItemComponent, NxBreadcrumbComponent],
                    exports: [NxBreadcrumbItemComponent, NxBreadcrumbComponent],
                    imports: [NxIconModule],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NxBreadcrumbComponent, NxBreadcrumbItemComponent, NxBreadcrumbModule };
