// @aposin/ng-aquila/fesm2020/aposin-ng-aquila-progressbar.mjs
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Input, NgModule } from '@angular/core';
import { clamp } from '@aposin/ng-aquila/utils';
import * as i1 from '@angular/common';
import { CommonModule } from '@angular/common';

let progressbarId = 0;
class NxProgressbarComponent {
    constructor() {
        /** @docs-private */
        this.progressbarId = `nx-progress-bar-${progressbarId++}`;
        this._value = 0;
    }
    /** Sets the value of the progress bar. Defaults to zero. Mirrored to aria-value now. */
    get value() {
        return this._value;
    }
    set value(value) {
        this._value = clamp(value || 0); // TODO properly coerce input value
    }
    _primaryTransform() {
        const scale = this.value;
        return { transform: `scaleX(${scale})` };
    }
}
NxProgressbarComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressbarComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
NxProgressbarComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxProgressbarComponent, selector: "nx-progressbar", inputs: { value: "value" }, host: { properties: { "attr.aria-valuenow": "value" } }, ngImport: i0, template: "<div class=\"nx-progress__track\">\n    <div class=\"nx-progress__indicator\" [ngStyle]=\"_primaryTransform()\"></div>\n</div>\n", styles: [":host{display:block;height:4px;overflow:hidden;position:relative;width:100%}.nx-progress__track{background-color:var(--progressbar-background-color);height:100%}.nx-progress__indicator{transform-origin:top left;height:100%;position:absolute;width:100%}:host-context([dir=rtl]) .nx-progress__indicator{transform-origin:top right}.nx-progress__indicator:after{height:100%;position:absolute;width:100%;animation:none;content:\"\";display:inline-block;left:0;background-color:var(--progressbar-indicator-color)}:host-context([dir=rtl]) .nx-progress__indicator:after{right:0;left:auto}@media screen and (-ms-high-contrast: active){:host{height:8px}.nx-progress__track{border:1px solid windowFrame;border-radius:2px}.nx-progress__indicator{border-right:1px solid windowFrame}[dir=rtl] .nx-progress__indicator{border-right:none;border-left:1px solid windowFrame}.nx-progress__indicator:after{background:highlight}}\n"], directives: [{ type: i1.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressbarComponent, decorators: [{
            type: Component,
            args: [{ selector: 'nx-progressbar', changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        '[attr.aria-valuenow]': 'value',
                    }, template: "<div class=\"nx-progress__track\">\n    <div class=\"nx-progress__indicator\" [ngStyle]=\"_primaryTransform()\"></div>\n</div>\n", styles: [":host{display:block;height:4px;overflow:hidden;position:relative;width:100%}.nx-progress__track{background-color:var(--progressbar-background-color);height:100%}.nx-progress__indicator{transform-origin:top left;height:100%;position:absolute;width:100%}:host-context([dir=rtl]) .nx-progress__indicator{transform-origin:top right}.nx-progress__indicator:after{height:100%;position:absolute;width:100%;animation:none;content:\"\";display:inline-block;left:0;background-color:var(--progressbar-indicator-color)}:host-context([dir=rtl]) .nx-progress__indicator:after{right:0;left:auto}@media screen and (-ms-high-contrast: active){:host{height:8px}.nx-progress__track{border:1px solid windowFrame;border-radius:2px}.nx-progress__indicator{border-right:1px solid windowFrame}[dir=rtl] .nx-progress__indicator{border-right:none;border-left:1px solid windowFrame}.nx-progress__indicator:after{background:highlight}}\n"] }]
        }], propDecorators: { value: [{
                type: Input
            }] } });

class NxProgressbarModule {
}
NxProgressbarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressbarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxProgressbarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressbarModule, declarations: [NxProgressbarComponent], imports: [CommonModule], exports: [NxProgressbarComponent] });
NxProgressbarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressbarModule, providers: [], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressbarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [NxProgressbarComponent],
                    declarations: [NxProgressbarComponent],
                    providers: [],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NxProgressbarComponent, NxProgressbarModule };
