// @aposin/ng-aquila/esm2020/rating/rating.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxRatingComponent } from './rating.component';
import * as i0 from "@angular/core";
export class NxRatingModule {
}
NxRatingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRatingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxRatingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRatingModule, declarations: [NxRatingComponent], imports: [CommonModule, NxIconModule], exports: [NxRatingComponent] });
NxRatingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRatingModule, imports: [[CommonModule, NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRatingModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, NxIconModule],
                    declarations: [NxRatingComponent],
                    exports: [NxRatingComponent],
                }]
        }] });
