// @aposin/ng-aquila/esm2020/natural-language-form/natural-language-form.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxInputModule } from '@aposin/ng-aquila/input';
import { NxPopoverModule } from '@aposin/ng-aquila/popover';
import { NxNaturalLanguageFormComponent } from './natural-language-form.component';
import { NxWordComponent } from './word.component';
import * as i0 from "@angular/core";
export class NxNaturalLanguageFormModule {
}
NxNaturalLanguageFormModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNaturalLanguageFormModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxNaturalLanguageFormModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNaturalLanguageFormModule, declarations: [NxNaturalLanguageFormComponent, NxWordComponent], imports: [CommonModule, NxInputModule, NxPopoverModule], exports: [NxNaturalLanguageFormComponent, NxWordComponent] });
NxNaturalLanguageFormModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNaturalLanguageFormModule, imports: [[CommonModule, NxInputModule, NxPopoverModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNaturalLanguageFormModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, NxInputModule, NxPopoverModule],
                    declarations: [NxNaturalLanguageFormComponent, NxWordComponent],
                    exports: [NxNaturalLanguageFormComponent, NxWordComponent],
                }]
        }] });
