// @aposin/ng-aquila/esm2020/circle-toggle/circle-toggle.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxCircleToggleComponent } from './circle-toggle/circle-toggle.component';
import { NxCircleToggleGroupComponent } from './circle-toggle-group/circle-toggle-group.component';
import { NxIconToggleButtonComponent } from './icon-toggle-button/icon-toggle-button.component';
import { NxMobileToggleButtonComponent } from './mobile-toggle-button/mobile-toggle-button.component';
import * as i0 from "@angular/core";
export class NxCircleToggleModule {
}
NxCircleToggleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxCircleToggleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxCircleToggleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxCircleToggleModule, declarations: [NxCircleToggleGroupComponent, NxCircleToggleComponent, NxIconToggleButtonComponent, NxMobileToggleButtonComponent], imports: [CommonModule, NxIconModule], exports: [NxCircleToggleGroupComponent, NxIconToggleButtonComponent, NxCircleToggleComponent, NxMobileToggleButtonComponent] });
NxCircleToggleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxCircleToggleModule, imports: [[CommonModule, NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxCircleToggleModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxCircleToggleGroupComponent, NxCircleToggleComponent, NxIconToggleButtonComponent, NxMobileToggleButtonComponent],
                    exports: [NxCircleToggleGroupComponent, NxIconToggleButtonComponent, NxCircleToggleComponent, NxMobileToggleButtonComponent],
                    imports: [CommonModule, NxIconModule],
                }]
        }] });
