// @aposin/ng-aquila/esm2020/dropdown/multi-select/multi-select-option.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, HostListener, Input, Output } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@aposin/ng-aquila/icon";
import * as i2 from "@angular/common";
let optionId = 0;
/**
 * A single option witin the multi select.
 *
 * @docs-private
 */
export class NxMultiSelectOptionComponent {
    constructor(_cdr, elementRef) {
        this._cdr = _cdr;
        this.elementRef = elementRef;
        this._active = false;
        this.id = `nx-multi-select-option-${optionId++}`;
        this.appearance = 'auto';
        /**
         * Label of this option.
         */
        this.label = '';
        /**
         * Whether this option is selected.
         */
        this.selected = false;
        /**
         * Whether thisoption is disabled.
         */
        this.disabled = false;
        /**
         * Emits an event when this option is selected or unselected by the user.
         */
        this.selectedChange = new EventEmitter();
    }
    /**
     * Sets this option active highlighting it to the user.
     */
    set active(value) {
        this._active = value;
        this._cdr.markForCheck();
    }
    get active() {
        return this._active;
    }
    setActiveStyles() {
        this.active = true;
    }
    setInactiveStyles() {
        this.active = false;
    }
    _onClick() {
        if (!this.disabled) {
            this.selected = !this.selected;
            this.selectedChange.emit(this.selected);
        }
    }
    /**
     * Selects this option as if the user clicked on it.
     */
    selectViaInteraction() {
        this._onClick();
        this._cdr.markForCheck();
    }
}
NxMultiSelectOptionComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMultiSelectOptionComponent, deps: [{ token: i0.ChangeDetectorRef }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
NxMultiSelectOptionComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxMultiSelectOptionComponent, selector: "nx-multi-select-option", inputs: { appearance: "appearance", value: "value", label: "label", selected: "selected", disabled: "disabled" }, outputs: { selectedChange: "selectedChange" }, host: { attributes: { "role": "option" }, listeners: { "click": "_onClick()" }, properties: { "id": "id", "attr.aria-selected": "selected || null", "attr.aria-disabled": "disabled || null", "class.is-outline": "appearance === \"outline\"" } }, ngImport: i0, template: "<div class=\"nx-checkbox\" [class.is-active]=\"active\" [class.is-selected]=\"selected\" [class.is-disabled]=\"disabled\">\n    <span class=\"nx-checkbox__control\">\n        <nx-icon *ngIf=\"selected\" name=\"check\" aria-hidden=\"true\"></nx-icon>\n    </span>\n\n    <span class=\"nx-checkbox__label\">\n        {{ label }}\n    </span>\n</div>\n", styles: [":host{display:flex;width:100%;padding:8px 0}:host.is-outline{padding:8px 0}:host.is-outline .nx-checkbox{font-size:var(--checkbox-small-label-font-size);line-height:var(--checkbox-small-label-line-height);font-weight:var(--checkbox-small-label-font-weight);letter-spacing:var(--checkbox-small-label-letter-spacing)}.nx-checkbox{font-size:var(--checkbox-large-label-font-size);line-height:var(--checkbox-large-label-line-height);font-weight:var(--checkbox-large-label-font-weight);letter-spacing:var(--checkbox-large-label-letter-spacing);color:var(--checkbox-label-color);cursor:pointer;display:flex;align-items:flex-start}.nx-checkbox.is-active{box-shadow:var(--focus-box-shadow);border-radius:4px}@media screen and (-ms-high-contrast: active),(forced-colors: active){.nx-checkbox.is-active{box-shadow:0 0 0 2px background,0 0 0 6px windowText;outline:4px solid CanvasText;outline-offset:2px}}.nx-checkbox.is-disabled{cursor:not-allowed}.nx-checkbox.is-disabled .nx-checkbox__control{background-color:var(--checkbox-disabled-background-color);border-color:var(--checkbox-disabled-border-color);color:var(--checkbox-disabled-icon-color)}.nx-checkbox.is-disabled .nx-checkbox__label{color:var(--checkbox-label-disabled-color)}.nx-checkbox.is-selected .nx-checkbox__control{background-color:var(--checkbox-selected-background-color);border-color:var(--checkbox-selected-border-color)}.nx-checkbox__control{background-color:var(--checkbox-background-color);border-style:solid;border-width:var(--checkbox-border-width);border-color:var(--checkbox-border-color);border-radius:var(--checkbox-border-radius);color:var(--checkbox-icon-color);font-size:var(--checkbox-icon-size);height:var(--checkbox-control-size);width:var(--checkbox-control-size);min-width:var(--checkbox-control-size);display:flex;justify-content:center;align-items:flex-start;z-index:0}.nx-checkbox__label{padding:0;margin-left:var(--checkbox-small-label-text-margin-left)}:host-context([dir=rtl]) .nx-checkbox__label{margin-left:0;margin-left:initial;margin-right:var(--checkbox-small-label-text-margin-left)}\n"], components: [{ type: i1.NxIconComponent, selector: "nx-icon", inputs: ["name", "outline", "fill", "size", "font"] }], directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMultiSelectOptionComponent, decorators: [{
            type: Component,
            args: [{ selector: 'nx-multi-select-option', host: {
                        role: 'option',
                        '[id]': 'id',
                        '[attr.aria-selected]': 'selected || null',
                        '[attr.aria-disabled]': 'disabled || null',
                        '[class.is-outline]': 'appearance === "outline"',
                    }, changeDetection: ChangeDetectionStrategy.OnPush, template: "<div class=\"nx-checkbox\" [class.is-active]=\"active\" [class.is-selected]=\"selected\" [class.is-disabled]=\"disabled\">\n    <span class=\"nx-checkbox__control\">\n        <nx-icon *ngIf=\"selected\" name=\"check\" aria-hidden=\"true\"></nx-icon>\n    </span>\n\n    <span class=\"nx-checkbox__label\">\n        {{ label }}\n    </span>\n</div>\n", styles: [":host{display:flex;width:100%;padding:8px 0}:host.is-outline{padding:8px 0}:host.is-outline .nx-checkbox{font-size:var(--checkbox-small-label-font-size);line-height:var(--checkbox-small-label-line-height);font-weight:var(--checkbox-small-label-font-weight);letter-spacing:var(--checkbox-small-label-letter-spacing)}.nx-checkbox{font-size:var(--checkbox-large-label-font-size);line-height:var(--checkbox-large-label-line-height);font-weight:var(--checkbox-large-label-font-weight);letter-spacing:var(--checkbox-large-label-letter-spacing);color:var(--checkbox-label-color);cursor:pointer;display:flex;align-items:flex-start}.nx-checkbox.is-active{box-shadow:var(--focus-box-shadow);border-radius:4px}@media screen and (-ms-high-contrast: active),(forced-colors: active){.nx-checkbox.is-active{box-shadow:0 0 0 2px background,0 0 0 6px windowText;outline:4px solid CanvasText;outline-offset:2px}}.nx-checkbox.is-disabled{cursor:not-allowed}.nx-checkbox.is-disabled .nx-checkbox__control{background-color:var(--checkbox-disabled-background-color);border-color:var(--checkbox-disabled-border-color);color:var(--checkbox-disabled-icon-color)}.nx-checkbox.is-disabled .nx-checkbox__label{color:var(--checkbox-label-disabled-color)}.nx-checkbox.is-selected .nx-checkbox__control{background-color:var(--checkbox-selected-background-color);border-color:var(--checkbox-selected-border-color)}.nx-checkbox__control{background-color:var(--checkbox-background-color);border-style:solid;border-width:var(--checkbox-border-width);border-color:var(--checkbox-border-color);border-radius:var(--checkbox-border-radius);color:var(--checkbox-icon-color);font-size:var(--checkbox-icon-size);height:var(--checkbox-control-size);width:var(--checkbox-control-size);min-width:var(--checkbox-control-size);display:flex;justify-content:center;align-items:flex-start;z-index:0}.nx-checkbox__label{padding:0;margin-left:var(--checkbox-small-label-text-margin-left)}:host-context([dir=rtl]) .nx-checkbox__label{margin-left:0;margin-left:initial;margin-right:var(--checkbox-small-label-text-margin-left)}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.ChangeDetectorRef }, { type: i0.ElementRef }]; }, propDecorators: { appearance: [{
                type: Input
            }], value: [{
                type: Input
            }], label: [{
                type: Input
            }], selected: [{
                type: Input
            }], disabled: [{
                type: Input
            }], selectedChange: [{
                type: Output
            }], _onClick: [{
                type: HostListener,
                args: ['click']
            }] } });
