// @aposin/ng-aquila/esm2020/dropdown/multi-select/multi-select.component.mjs
var _NxMultiSelectComponent_disableSelectAll, _NxMultiSelectComponent_filterFn;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { ActiveDescendantKeyManager } from '@angular/cdk/a11y';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { CdkConnectedOverlay } from '@angular/cdk/overlay';
import { ChangeDetectionStrategy, Component, EventEmitter, HostBinding, Input, Optional, Output, Self, ViewChild, ViewChildren, } from '@angular/core';
import { NxFormfieldControl } from '@aposin/ng-aquila/formfield';
import { Subject } from 'rxjs';
import { take, takeUntil } from 'rxjs/operators';
import { getPositionOffset, getPositions } from '../dropdown-position';
import { NxMultiSelectOptionComponent } from './multi-select-option.component';
import * as i0 from "@angular/core";
import * as i1 from "../dropdown";
import * as i2 from "@aposin/ng-aquila/utils";
import * as i3 from "@aposin/ng-aquila/formfield";
import * as i4 from "@angular/forms";
import * as i5 from "@aposin/ng-aquila/icon";
import * as i6 from "@aposin/ng-aquila/checkbox";
import * as i7 from "@aposin/ng-aquila/button";
import * as i8 from "./multi-select-option.component";
import * as i9 from "@angular/cdk/overlay";
import * as i10 from "@aposin/ng-aquila/tooltip";
import * as i11 from "@angular/common";
import * as i12 from "@aposin/ng-aquila/input";
let id = 0;
const OVERLAY_MIN_WIDTH = 260;
const _defaultFilterFn = (query, label) => label.toLowerCase().includes(query.toLowerCase());
/**
 * Multi Select component.
 * @typeParam S Type of the items in the options array
 * @typeParam T Type of the ngModel value
 */
export class NxMultiSelectComponent {
    constructor(_intl, _elementRef, _errorStateMatcher, _cdr, _formFieldComponent, ngControl, _parentForm, _parentFormGroup) {
        this._intl = _intl;
        this._elementRef = _elementRef;
        this._errorStateMatcher = _errorStateMatcher;
        this._cdr = _cdr;
        this._formFieldComponent = _formFieldComponent;
        this.ngControl = ngControl;
        this._parentForm = _parentForm;
        this._parentFormGroup = _parentFormGroup;
        _NxMultiSelectComponent_disableSelectAll.set(this, false);
        _NxMultiSelectComponent_filterFn.set(this, void 0);
        this._required = false;
        this._disabled = false;
        this._readonly = false;
        this._placeholder = '';
        this._filter = false;
        this._openedBy = 'mouse';
        /** @docs-private */
        this._divider = 0;
        /** Event emitted when the selected value has been changed. */
        this.selectionChange = new EventEmitter();
        /** @docs-private */
        this.controlType = 'nx-multi-select';
        /** @docs-private */
        this.errorState = false;
        this._positions = getPositions('auto', 0);
        this._inputFocused = false;
        this._ariaDescribedby = '';
        this._width = 0;
        this._filterValue = '';
        this._tooltipText = '';
        this.listItems = [];
        this.selectedItems = new Set();
        this.id = `nx-multi-select-${id++}`;
        this._comboboxId = `${this.id}-combobox`;
        this.stateChanges = new Subject();
        /**
         * List of options to choose from.
         */
        this.options = [];
        /**
         * Placeholder for the filter input.
         */
        this.filterPlaceholder = 'Type to filter';
        /**
         * Selector to get the value of an option.
         * Can be either a property name or a selector function.
         * When empty the whole option is treated as the value.
         */
        this.selectValue = '';
        /**
         * Selector to get the label of an option.
         * Can be either a property name or a selector function.
         * When empty the whole option is treated as the label.
         */
        this.selectLabel = '';
        this._isOpen = false;
        this._destroyed = new Subject();
        this._onChange = () => { };
        this._onTouched = () => { };
        if (this.ngControl) {
            // Note: we provide the value accessor through here, instead of
            // the `providers` to avoid running into a circular import.
            this.ngControl.valueAccessor = this;
        }
        this._isDisabled = this._isDisabled.bind(this);
        _intl.changes.pipe(takeUntil(this._destroyed)).subscribe(() => _cdr.markForCheck());
    }
    get value() {
        return this.options.filter(option => this.selectedItems.has(option)).map(option => this._selectValue(option));
    }
    get _isOutline() {
        return this._appearance === 'outline';
    }
    get _appearance() {
        return this._formFieldComponent?.appearance;
    }
    get _overlayLabel() {
        return this._formFieldComponent?.label;
    }
    /**
     * Whether the value is empty.
     */
    get empty() {
        return this.selectedItems.size === 0;
    }
    get focused() {
        return this._isOpen || this._inputFocused;
    }
    // @Input()
    // selectValue(value: string | () )
    get required() {
        return this._required;
    }
    set required(value) {
        this._required = value;
    }
    /** Whether the multi select is disabled. */
    get disabled() {
        return this._disabled;
    }
    set disabled(value) {
        this._disabled = value;
    }
    /** Whether the multi select should be read only. */
    get readonly() {
        return this._readonly;
    }
    set readonly(value) {
        this._readonly = coerceBooleanProperty(value);
        this.stateChanges.next();
    }
    /** The placeholder shown in the multi select field. */
    get placeholder() {
        return this._placeholder;
    }
    set placeholder(value) {
        this._placeholder = value;
        this.stateChanges.next();
    }
    /**
     * Whether this multi select can be filtered.
     */
    set filter(value) {
        this._filter = coerceBooleanProperty(value);
    }
    get filter() {
        return this._filter;
    }
    /** Whether the (select all / clear all) should be disabled and hidden. */
    set disableSelectAll(value) {
        const coercedValue = coerceBooleanProperty(value);
        if (__classPrivateFieldGet(this, _NxMultiSelectComponent_disableSelectAll, "f") !== coercedValue) {
            __classPrivateFieldSet(this, _NxMultiSelectComponent_disableSelectAll, coercedValue, "f");
            this._cdr.markForCheck();
        }
    }
    get disableSelectAll() {
        return __classPrivateFieldGet(this, _NxMultiSelectComponent_disableSelectAll, "f");
    }
    /**
     * Function to be used when the user types into the search filter. The first argument is the user input,
     * the second argument is the dropdown item value as displayed. The dropdown items will use this function
     * to set their visibility state. A boolean should be returned.
     *
     * Defaults to lower case inclusion.
     */
    set filterFn(value) {
        __classPrivateFieldSet(this, _NxMultiSelectComponent_filterFn, value, "f");
    }
    get filterFn() {
        return __classPrivateFieldGet(this, _NxMultiSelectComponent_filterFn, "f") ?? _defaultFilterFn;
    }
    /** @docs-private */
    get shouldLabelFloat() {
        return this.focused || !this.empty || !!(this.placeholder && this.placeholder.length > 0);
    }
    /** @docs-private */
    get elementRef() {
        return this._elementRef;
    }
    get _allSelected() {
        return this.selectedItems.size === this.options.filter(option => !this._isDisabled(option)).length;
    }
    get _someSelected() {
        return this.selectedItems.size > 0 && !this._allSelected;
    }
    get _isActiveItemFiltered() {
        return (typeof this._keyManager.activeItem !== 'undefined' &&
            !this.listItems.some(option => this._selectValue(option) === this._keyManager.activeItem?.value));
    }
    ngOnDestroy() {
        this._destroyed.next();
        this._destroyed.complete();
    }
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
        this._cdr.markForCheck();
        this.stateChanges.next();
    }
    _selectValue(option) {
        if (!this.selectValue) {
            return option;
        }
        if (typeof this.selectValue === 'string') {
            return option[this.selectValue];
        }
        return this.selectValue(option);
    }
    _selectLabel(option) {
        if (!this.selectLabel) {
            return String(this._selectValue(option));
        }
        if (typeof this.selectLabel === 'string') {
            return String(option[this.selectLabel]);
        }
        return this.selectLabel(option);
    }
    _updatePositions() {
        if (this._formFieldComponent) {
            const offset = getPositionOffset(this._elementRef.nativeElement, this._formFieldComponent.elementRef.nativeElement, this._panelHeader?.nativeElement);
            this._positions = getPositions(this._appearance, offset);
        }
    }
    _isDisabled(option) {
        if (typeof this.selectDisabled === 'string') {
            return option[this.selectDisabled];
        }
        if (typeof this.selectDisabled === 'function') {
            return this.selectDisabled(option);
        }
        return false;
    }
    /**
     * Returns html ids of multi select rendered value and label (if available),
     * separated by space.
     */
    _getAriaLabelledBy() {
        const valueId = this.id;
        const labelId = this._formFieldComponent?.labelId;
        if (labelId) {
            return `${valueId} ${labelId}`;
        }
        return valueId;
    }
    /** @docs-private */
    setDescribedByIds(ids) {
        this._ariaDescribedby = ids.join(' ');
    }
    ngAfterViewInit() {
        this._initKeyManager();
    }
    open($event, origin) {
        this._open($event, origin);
    }
    _open($event, origin) {
        if (this._isOpen || this.disabled) {
            return;
        }
        const sortSelectedToTop = (a, b) => {
            const aSelected = this.selectedItems.has(a);
            const bSelected = this.selectedItems.has(b);
            if (aSelected && !bSelected) {
                return -1;
            }
            else if (!aSelected && bSelected) {
                return 1;
            }
            return 0;
        };
        $event.preventDefault();
        this._filterValue = '';
        this._width = Math.max(OVERLAY_MIN_WIDTH, this._trigger?.nativeElement.getBoundingClientRect().width);
        this._isOpen = true;
        this.listItems = this.options.slice().sort(sortSelectedToTop);
        this._divider = this.selectedItems.size - 1;
        this._openedBy = origin;
        this._cdr.markForCheck();
    }
    _close() {
        this._isOpen = false;
        this._updateTooltipText();
        this._trigger?.nativeElement.focus();
    }
    _onSelect(item, selected) {
        if (selected) {
            this.selectedItems.add(item);
        }
        else {
            this.selectedItems.delete(item);
        }
        this._onChange(this.value);
        this.selectionChange.emit(this.value);
    }
    _onKeydown($event) {
        if (!this._isOpen || this.disabled) {
            return;
        }
        const key = $event.key;
        if (key === 'ArrowUp' || key === 'ArrowDown' || key === 'Home' || key === 'End') {
            this._keyManager.onKeydown($event);
            this._scrollActiveOptionIntoView();
            $event.preventDefault();
        }
        if ((window.navigator.platform.match('Mac') ? $event.metaKey : $event.ctrlKey) && key === 'a') {
            this._onSelectAll();
            $event.preventDefault();
        }
        if (key === 'Tab') {
            this._keyManager.onKeydown($event);
        }
        if (key === 'Enter' || (this._filterValue.trim() === '' && key === ' ')) {
            this._keyManager.activeItem?.selectViaInteraction();
            $event.preventDefault();
        }
    }
    _onTriggerBlur() {
        this._onTouched();
    }
    _onFocusWithinOverlay($event) {
        // Keeps focus on the filter when clicking around in the overlay
        if (this._filterInput && $event.target !== this._filterInput.nativeElement) {
            this._filterInput.nativeElement.focus();
        }
    }
    _getAriaActiveDescendant() {
        if (this._keyManager.activeItem) {
            return this._keyManager.activeItem.id;
        }
        return null;
    }
    _onFilterChange(query) {
        if (query) {
            this.listItems = this.options.filter(item => this.filterFn(query, this._selectLabel(item)));
        }
        else {
            this.listItems = this.options.slice();
        }
        if (this._isActiveItemFiltered) {
            setTimeout(() => {
                this._keyManager.setFirstItemActive();
            });
        }
        this._optionsList?.nativeElement.scrollTo(0, 0);
    }
    _getValueText() {
        return this.options
            .filter(option => this.selectedItems.has(option))
            .map(option => this._selectLabel(option))
            .join(', ');
    }
    _onAttach() {
        this._overlayDir?.positionChange.pipe(take(1)).subscribe(() => {
            const overlayRef = this._overlayDir.overlayRef;
            const positionStrategy = overlayRef.getConfig().positionStrategy;
            this._updatePositions();
            positionStrategy.withPositions(this._positions);
            overlayRef.updatePosition();
            this._filterInput?.nativeElement.focus();
            this._panelContent?.nativeElement.focus();
            if (this._openedBy === 'keyboard') {
                this._keyManager.setFirstItemActive();
                this._scrollActiveOptionIntoView();
            }
            this._cdr.markForCheck();
        });
    }
    _onDetach() {
        this._close();
    }
    _onSelectAll() {
        if (this._allSelected) {
            this.selectedItems.clear();
        }
        else {
            this.listItems.filter(option => !this._isDisabled(option)).forEach(option => this.selectedItems.add(option));
        }
        this._onChange(this.value);
        this.selectionChange.emit(this.value);
    }
    _clear() {
        this.selectedItems.clear();
        this._onChange(this.value);
        this.selectionChange.emit(this.value);
    }
    _clearFilter() {
        this._filterValue = '';
        this._onFilterChange(this._filterValue);
    }
    writeValue(value) {
        this.selectedItems.clear();
        if (Array.isArray(value)) {
            for (const item of value) {
                const selectedItem = this.options.find(option => this._selectValue(option) === item);
                if (selectedItem) {
                    this.selectedItems.add(selectedItem);
                }
                else {
                    console.warn('NxMultiSelect: Model contains value that does not exist in given options', item);
                }
            }
        }
        this._cdr.markForCheck();
    }
    registerOnChange(fn) {
        this._onChange = fn;
    }
    registerOnTouched(fn) {
        this._onTouched = fn;
    }
    ngDoCheck() {
        if (this.ngControl) {
            this.updateErrorState();
        }
    }
    updateErrorState() {
        const oldState = this.errorState;
        const parent = this._parentFormGroup || this._parentForm;
        const control = this.ngControl ? this.ngControl.control : null;
        const newState = this._errorStateMatcher.isErrorState(control, parent);
        if (newState !== oldState) {
            this.errorState = newState;
            this.stateChanges.next();
        }
    }
    _updateTooltipText() {
        if (!this._trigger) {
            return;
        }
        const [label, suffix, icon] = this._trigger.nativeElement.children;
        const { paddingLeft, paddingRight } = getComputedStyle(this._trigger.nativeElement);
        const triggerContentWidth = this._trigger.nativeElement.clientWidth - parseInt(paddingLeft, 10) - parseInt(paddingRight, 10);
        if (triggerContentWidth - suffix.offsetWidth - icon.offsetWidth <= label.offsetWidth) {
            this._tooltipText = this._getValueText();
        }
        else {
            this._tooltipText = '';
        }
    }
    _scrollActiveOptionIntoView() {
        if (!this._keyManager.activeItem) {
            return;
        }
        const activeItem = this._keyManager.activeItem.elementRef.nativeElement;
        const list = this._optionsList?.nativeElement;
        const listTopScrollPosition = list.scrollTop;
        const listHeight = list.clientHeight;
        const itemTop = activeItem.offsetTop;
        const itemBottom = activeItem.offsetTop + activeItem.offsetHeight;
        // item half or less visible on top
        if (itemTop < listTopScrollPosition) {
            list.scrollTop = itemTop;
            // item half or less visible on bottom
        }
        else if (itemBottom > listTopScrollPosition + listHeight) {
            list.scrollTop = itemBottom - listHeight;
        }
    }
    _initKeyManager() {
        this._keyManager = new ActiveDescendantKeyManager(this._options)
            .withHomeAndEnd()
            .withVerticalOrientation()
            .withHorizontalOrientation('ltr')
            .skipPredicate(item => item.disabled);
        this._keyManager.tabOut.pipe(takeUntil(this._destroyed)).subscribe(() => this._close());
    }
}
_NxMultiSelectComponent_disableSelectAll = new WeakMap(), _NxMultiSelectComponent_filterFn = new WeakMap();
NxMultiSelectComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMultiSelectComponent, deps: [{ token: i1.NxDropdownIntl }, { token: i0.ElementRef }, { token: i2.ErrorStateMatcher }, { token: i0.ChangeDetectorRef }, { token: i3.NxFormfieldComponent, optional: true }, { token: i4.NgControl, optional: true, self: true }, { token: i4.NgForm, optional: true }, { token: i4.FormGroupDirective, optional: true }], target: i0.ɵɵFactoryTarget.Component });
NxMultiSelectComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxMultiSelectComponent, selector: "nx-multi-select", inputs: { required: "required", disabled: "disabled", readonly: "readonly", placeholder: "placeholder", filter: "filter", disableSelectAll: "disableSelectAll", filterFn: "filterFn", options: "options", filterPlaceholder: "filterPlaceholder", selectValue: "selectValue", selectLabel: "selectLabel", selectDisabled: "selectDisabled" }, outputs: { selectionChange: "selectionChange" }, host: { properties: { "class.is-open": "this._isOpen" } }, providers: [{ provide: NxFormfieldControl, useExisting: NxMultiSelectComponent }], viewQueries: [{ propertyName: "_panelContent", first: true, predicate: ["panelContent"], descendants: true }, { propertyName: "_filterInput", first: true, predicate: ["filterInput"], descendants: true }, { propertyName: "_optionsList", first: true, predicate: ["itemsList"], descendants: true }, { propertyName: "_trigger", first: true, predicate: ["trigger"], descendants: true }, { propertyName: "_panelHeader", first: true, predicate: ["panelHeader"], descendants: true }, { propertyName: "_overlayDir", first: true, predicate: CdkConnectedOverlay, descendants: true, static: true }, { propertyName: "_options", predicate: NxMultiSelectOptionComponent, descendants: true }], ngImport: i0, template: "<div\n    [nxTooltip]=\"_tooltipText\"\n    class=\"value\"\n    (click)=\"_open($event, 'mouse')\"\n    (keydown.enter)=\"_open($event, 'keyboard')\"\n    (keydown.space)=\"_open($event, 'keyboard')\"\n    cdkOverlayOrigin\n    #trigger\n    #origin=\"cdkOverlayOrigin\"\n    role=\"combobox\"\n    tabindex=\"0\"\n    aria-haspopup=\"listbox\"\n    [attr.aria-controls]=\"_comboboxId\"\n    (blur)=\"_onTriggerBlur(); _inputFocused = false\"\n    [attr.aria-expanded]=\"_isOpen\"\n    [attr.aria-labelledby]=\"_getAriaLabelledBy()\"\n    [id]=\"id\"\n    (focus)=\"disabled ? null : (_inputFocused = true)\"\n    (keydown)=\"_onKeydown($event)\"\n>\n    <span class=\"value-text\">{{ empty ? placeholder : _getValueText() }}</span>\n    <span>{{ selectedItems.size ? '(' + selectedItems.size + ')' : '' }}</span>\n    <nx-icon size=\"s\" aria-hidden=\"true\" class=\"indicator\" name=\"chevron-down\"></nx-icon>\n</div>\n<ng-template\n    cdkConnectedOverlay\n    cdkConnectedOverlayHasBackdrop\n    cdkConnectedOverlayPush\n    cdkConnectedOverlayBackdropClass=\"cdk-overlay-transparent-backdrop\"\n    [cdkConnectedOverlayPositions]=\"_positions\"\n    [cdkConnectedOverlayWidth]=\"_width\"\n    [cdkConnectedOverlayOrigin]=\"origin\"\n    [cdkConnectedOverlayOpen]=\"_isOpen\"\n    (overlayOutsideClick)=\"_close()\"\n    (attach)=\"_onAttach()\"\n    (detach)=\"_onDetach()\"\n>\n    <div\n        class=\"panel\"\n        [class.is-outline]=\"_isOutline\"\n        (keydown)=\"_onKeydown($event)\"\n        [attr.aria-owns]=\"_comboboxId\"\n        aria-expanded=\"true\"\n        (focusin)=\"_onFocusWithinOverlay($event)\"\n        role=\"combobox\"\n        [attr.aria-activedescendant]=\"_getAriaActiveDescendant()\"\n        tabindex=\"-1\"\n        [attr.aria-labelledby]=\"_getAriaLabelledBy()\"\n        #panelContent\n    >\n        <div *ngIf=\"!_isOutline && _overlayLabel\" class=\"panel-header\" #panelHeader>\n            {{ _overlayLabel }}\n        </div>\n        <div *ngIf=\"filter\" class=\"filter-wrapper\">\n            <nx-formfield class=\"filter\" appearance=\"auto\">\n                <input\n                    nxInput\n                    #filterInput\n                    [(ngModel)]=\"_filterValue\"\n                    (ngModelChange)=\"_onFilterChange($event)\"\n                    [placeholder]=\"filterPlaceholder\"\n                    autocomplete=\"off\"\n                    [attr.aria-controls]=\"_comboboxId\"\n                    [attr.aria-labelledby]=\"_getAriaLabelledBy()\"\n                    [attr.aria-activedescendant]=\"_getAriaActiveDescendant()\"\n                />\n                <nx-icon\n                    aria-hidden=\"true\"\n                    nxFormfieldSuffix\n                    [class.hidden]=\"!_filterValue.length\"\n                    class=\"clear\"\n                    (click)=\"_clearFilter()\"\n                    name=\"close\"\n                ></nx-icon>\n            </nx-formfield>\n        </div>\n        <div class=\"item actions\" *ngIf=\"_isOutline && !disableSelectAll\">\n            <nx-checkbox [ngModel]=\"_allSelected\" (ngModelChange)=\"_onSelectAll()\" [indeterminate]=\"_someSelected\"> {{ _intl.selectAll }} </nx-checkbox>\n            <button type=\"button\" nxPlainButton (click)=\"_clear()\" [disabled]=\"selectedItems.size === 0\"> {{ _intl.clearAll }} </button>\n        </div>\n        <div class=\"items\" #itemsList [id]=\"_comboboxId\" role=\"listbox\" aria-multiselectable=\"true\">\n            <nx-multi-select-option\n                *ngFor=\"let option of listItems; let i = index\"\n                class=\"item\"\n                [class.divider]=\"_divider === i\"\n                (selectedChange)=\"_onSelect(option, $event)\"\n                [value]=\"_selectValue(option)\"\n                [label]=\"_selectLabel(option)\"\n                [selected]=\"this.selectedItems.has(option)\"\n                [disabled]=\"this._isDisabled(option)\"\n                [appearance]=\"_appearance\"\n            ></nx-multi-select-option>\n        </div>\n    </div>\n</ng-template>\n", styles: [":host{display:block;width:100%;height:100%}.panel{width:100%;box-shadow:var(--shadow-small);border-radius:4px;background:var(--ui-01)}.panel.is-outline .filter{--formfield-label-height: 8px;--formfield-bottom-padding: 8px}.panel.is-outline .items{padding-bottom:8px;max-height:200px}.value{display:grid;align-items:center;width:100%;height:100%;max-width:100%;grid-template-columns:auto -webkit-min-content -webkit-min-content;grid-template-columns:auto min-content min-content;cursor:pointer}.value:focus{outline:none}.value::-moz-focus-inner{border:0}:host-context(.nx-formfield.is-disabled) .value{cursor:not-allowed}.value-text{overflow:hidden;white-space:nowrap;text-overflow:ellipsis;padding-right:8px}.panel-header{font-size:var(--formfield-label-font-size);line-height:var(--formfield-label-line-height);font-weight:var(--formfield-label-font-weight);letter-spacing:var(--formfield-label-letter-spacing);color:var(--dropdown-panel-header-text-color);background-color:var(--dropdown-panel-header-background-color);padding:8px 32px;border-top-left-radius:4px;border-top-right-radius:4px}.label span:nth-child(2){font-weight:600;margin-left:auto;padding-right:12px;flex:0 0 auto}.indicator{transform:rotate(0)}:host(.is-open) .indicator{transform:rotate(180deg)}.items{list-style:none;max-height:272px;overflow-y:auto;position:relative;padding:12px 32px}.filter-wrapper{padding:0 32px}.filter{width:100%}.border{padding-top:8px;margin-top:8px;border-top:1px solid lightgrey}.toggle{width:100%;display:block}.actions{display:flex;justify-content:space-between;white-space:nowrap;padding:8px 32px}.clear{cursor:pointer}.hidden{display:none}.divider{border-bottom:1px solid var(--dropdown-item-divider);margin-bottom:8px;padding-bottom:16px}\n"], components: [{ type: i5.NxIconComponent, selector: "nx-icon", inputs: ["name", "outline", "fill", "size", "font"] }, { type: i3.NxFormfieldComponent, selector: "nx-formfield", inputs: ["nxLabel", "nxFloatLabel", "nxStyle", "appearance"] }, { type: i6.NxCheckboxComponent, selector: "nx-checkbox", inputs: ["id", "name", "disabled", "labelSize", "negative", "checked", "indeterminate", "required", "value"], outputs: ["indeterminateChange", "checkedChange", "checkboxChange"] }, { type: i7.NxPlainButtonComponent, selector: "button[nxPlainButton]", inputs: ["nxPlainButton", "disabled"] }, { type: i8.NxMultiSelectOptionComponent, selector: "nx-multi-select-option", inputs: ["appearance", "value", "label", "selected", "disabled"], outputs: ["selectedChange"] }], directives: [{ type: i9.CdkOverlayOrigin, selector: "[cdk-overlay-origin], [overlay-origin], [cdkOverlayOrigin]", exportAs: ["cdkOverlayOrigin"] }, { type: i10.NxTooltipDirective, selector: "[nxTooltip]", inputs: ["nxTooltipPosition", "nxTooltipDisabled", "nxTooltipSelectable", "nxTooltipShowDelay", "nxTooltipHideDelay", "nxTooltip"], exportAs: ["nxTooltip"] }, { type: i9.CdkConnectedOverlay, selector: "[cdk-connected-overlay], [connected-overlay], [cdkConnectedOverlay]", inputs: ["cdkConnectedOverlayOrigin", "cdkConnectedOverlayPositions", "cdkConnectedOverlayPositionStrategy", "cdkConnectedOverlayOffsetX", "cdkConnectedOverlayOffsetY", "cdkConnectedOverlayWidth", "cdkConnectedOverlayHeight", "cdkConnectedOverlayMinWidth", "cdkConnectedOverlayMinHeight", "cdkConnectedOverlayBackdropClass", "cdkConnectedOverlayPanelClass", "cdkConnectedOverlayViewportMargin", "cdkConnectedOverlayScrollStrategy", "cdkConnectedOverlayOpen", "cdkConnectedOverlayDisableClose", "cdkConnectedOverlayTransformOriginOn", "cdkConnectedOverlayHasBackdrop", "cdkConnectedOverlayLockPosition", "cdkConnectedOverlayFlexibleDimensions", "cdkConnectedOverlayGrowAfterOpen", "cdkConnectedOverlayPush"], outputs: ["backdropClick", "positionChange", "attach", "detach", "overlayKeydown", "overlayOutsideClick"], exportAs: ["cdkConnectedOverlay"] }, { type: i11.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i12.NxInputDirective, selector: "input[nxInput], textarea[nxInput], select[nxInput]", inputs: ["nxAriaLabel", "id", "value", "readonly", "disabled", "required", "type", "placeholder"], exportAs: ["nxInput"] }, { type: i4.DefaultValueAccessor, selector: "input:not([type=checkbox])[formControlName],textarea[formControlName],input:not([type=checkbox])[formControl],textarea[formControl],input:not([type=checkbox])[ngModel],textarea[ngModel],[ngDefaultControl]" }, { type: i4.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { type: i4.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }, { type: i3.NxFormfieldSuffixDirective, selector: "[nxFormfieldSuffix]" }, { type: i11.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMultiSelectComponent, decorators: [{
            type: Component,
            args: [{ selector: 'nx-multi-select', providers: [{ provide: NxFormfieldControl, useExisting: NxMultiSelectComponent }], changeDetection: ChangeDetectionStrategy.OnPush, template: "<div\n    [nxTooltip]=\"_tooltipText\"\n    class=\"value\"\n    (click)=\"_open($event, 'mouse')\"\n    (keydown.enter)=\"_open($event, 'keyboard')\"\n    (keydown.space)=\"_open($event, 'keyboard')\"\n    cdkOverlayOrigin\n    #trigger\n    #origin=\"cdkOverlayOrigin\"\n    role=\"combobox\"\n    tabindex=\"0\"\n    aria-haspopup=\"listbox\"\n    [attr.aria-controls]=\"_comboboxId\"\n    (blur)=\"_onTriggerBlur(); _inputFocused = false\"\n    [attr.aria-expanded]=\"_isOpen\"\n    [attr.aria-labelledby]=\"_getAriaLabelledBy()\"\n    [id]=\"id\"\n    (focus)=\"disabled ? null : (_inputFocused = true)\"\n    (keydown)=\"_onKeydown($event)\"\n>\n    <span class=\"value-text\">{{ empty ? placeholder : _getValueText() }}</span>\n    <span>{{ selectedItems.size ? '(' + selectedItems.size + ')' : '' }}</span>\n    <nx-icon size=\"s\" aria-hidden=\"true\" class=\"indicator\" name=\"chevron-down\"></nx-icon>\n</div>\n<ng-template\n    cdkConnectedOverlay\n    cdkConnectedOverlayHasBackdrop\n    cdkConnectedOverlayPush\n    cdkConnectedOverlayBackdropClass=\"cdk-overlay-transparent-backdrop\"\n    [cdkConnectedOverlayPositions]=\"_positions\"\n    [cdkConnectedOverlayWidth]=\"_width\"\n    [cdkConnectedOverlayOrigin]=\"origin\"\n    [cdkConnectedOverlayOpen]=\"_isOpen\"\n    (overlayOutsideClick)=\"_close()\"\n    (attach)=\"_onAttach()\"\n    (detach)=\"_onDetach()\"\n>\n    <div\n        class=\"panel\"\n        [class.is-outline]=\"_isOutline\"\n        (keydown)=\"_onKeydown($event)\"\n        [attr.aria-owns]=\"_comboboxId\"\n        aria-expanded=\"true\"\n        (focusin)=\"_onFocusWithinOverlay($event)\"\n        role=\"combobox\"\n        [attr.aria-activedescendant]=\"_getAriaActiveDescendant()\"\n        tabindex=\"-1\"\n        [attr.aria-labelledby]=\"_getAriaLabelledBy()\"\n        #panelContent\n    >\n        <div *ngIf=\"!_isOutline && _overlayLabel\" class=\"panel-header\" #panelHeader>\n            {{ _overlayLabel }}\n        </div>\n        <div *ngIf=\"filter\" class=\"filter-wrapper\">\n            <nx-formfield class=\"filter\" appearance=\"auto\">\n                <input\n                    nxInput\n                    #filterInput\n                    [(ngModel)]=\"_filterValue\"\n                    (ngModelChange)=\"_onFilterChange($event)\"\n                    [placeholder]=\"filterPlaceholder\"\n                    autocomplete=\"off\"\n                    [attr.aria-controls]=\"_comboboxId\"\n                    [attr.aria-labelledby]=\"_getAriaLabelledBy()\"\n                    [attr.aria-activedescendant]=\"_getAriaActiveDescendant()\"\n                />\n                <nx-icon\n                    aria-hidden=\"true\"\n                    nxFormfieldSuffix\n                    [class.hidden]=\"!_filterValue.length\"\n                    class=\"clear\"\n                    (click)=\"_clearFilter()\"\n                    name=\"close\"\n                ></nx-icon>\n            </nx-formfield>\n        </div>\n        <div class=\"item actions\" *ngIf=\"_isOutline && !disableSelectAll\">\n            <nx-checkbox [ngModel]=\"_allSelected\" (ngModelChange)=\"_onSelectAll()\" [indeterminate]=\"_someSelected\"> {{ _intl.selectAll }} </nx-checkbox>\n            <button type=\"button\" nxPlainButton (click)=\"_clear()\" [disabled]=\"selectedItems.size === 0\"> {{ _intl.clearAll }} </button>\n        </div>\n        <div class=\"items\" #itemsList [id]=\"_comboboxId\" role=\"listbox\" aria-multiselectable=\"true\">\n            <nx-multi-select-option\n                *ngFor=\"let option of listItems; let i = index\"\n                class=\"item\"\n                [class.divider]=\"_divider === i\"\n                (selectedChange)=\"_onSelect(option, $event)\"\n                [value]=\"_selectValue(option)\"\n                [label]=\"_selectLabel(option)\"\n                [selected]=\"this.selectedItems.has(option)\"\n                [disabled]=\"this._isDisabled(option)\"\n                [appearance]=\"_appearance\"\n            ></nx-multi-select-option>\n        </div>\n    </div>\n</ng-template>\n", styles: [":host{display:block;width:100%;height:100%}.panel{width:100%;box-shadow:var(--shadow-small);border-radius:4px;background:var(--ui-01)}.panel.is-outline .filter{--formfield-label-height: 8px;--formfield-bottom-padding: 8px}.panel.is-outline .items{padding-bottom:8px;max-height:200px}.value{display:grid;align-items:center;width:100%;height:100%;max-width:100%;grid-template-columns:auto -webkit-min-content -webkit-min-content;grid-template-columns:auto min-content min-content;cursor:pointer}.value:focus{outline:none}.value::-moz-focus-inner{border:0}:host-context(.nx-formfield.is-disabled) .value{cursor:not-allowed}.value-text{overflow:hidden;white-space:nowrap;text-overflow:ellipsis;padding-right:8px}.panel-header{font-size:var(--formfield-label-font-size);line-height:var(--formfield-label-line-height);font-weight:var(--formfield-label-font-weight);letter-spacing:var(--formfield-label-letter-spacing);color:var(--dropdown-panel-header-text-color);background-color:var(--dropdown-panel-header-background-color);padding:8px 32px;border-top-left-radius:4px;border-top-right-radius:4px}.label span:nth-child(2){font-weight:600;margin-left:auto;padding-right:12px;flex:0 0 auto}.indicator{transform:rotate(0)}:host(.is-open) .indicator{transform:rotate(180deg)}.items{list-style:none;max-height:272px;overflow-y:auto;position:relative;padding:12px 32px}.filter-wrapper{padding:0 32px}.filter{width:100%}.border{padding-top:8px;margin-top:8px;border-top:1px solid lightgrey}.toggle{width:100%;display:block}.actions{display:flex;justify-content:space-between;white-space:nowrap;padding:8px 32px}.clear{cursor:pointer}.hidden{display:none}.divider{border-bottom:1px solid var(--dropdown-item-divider);margin-bottom:8px;padding-bottom:16px}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.NxDropdownIntl }, { type: i0.ElementRef }, { type: i2.ErrorStateMatcher }, { type: i0.ChangeDetectorRef }, { type: i3.NxFormfieldComponent, decorators: [{
                    type: Optional
                }] }, { type: i4.NgControl, decorators: [{
                    type: Optional
                }, {
                    type: Self
                }] }, { type: i4.NgForm, decorators: [{
                    type: Optional
                }] }, { type: i4.FormGroupDirective, decorators: [{
                    type: Optional
                }] }]; }, propDecorators: { required: [{
                type: Input
            }], disabled: [{
                type: Input
            }], readonly: [{
                type: Input
            }], placeholder: [{
                type: Input
            }], filter: [{
                type: Input
            }], disableSelectAll: [{
                type: Input
            }], filterFn: [{
                type: Input
            }], _options: [{
                type: ViewChildren,
                args: [NxMultiSelectOptionComponent]
            }], _panelContent: [{
                type: ViewChild,
                args: ['panelContent']
            }], _filterInput: [{
                type: ViewChild,
                args: ['filterInput']
            }], _optionsList: [{
                type: ViewChild,
                args: ['itemsList']
            }], _trigger: [{
                type: ViewChild,
                args: ['trigger']
            }], _panelHeader: [{
                type: ViewChild,
                args: ['panelHeader']
            }], _overlayDir: [{
                type: ViewChild,
                args: [CdkConnectedOverlay, { static: true }]
            }], selectionChange: [{
                type: Output
            }], options: [{
                type: Input
            }], filterPlaceholder: [{
                type: Input
            }], selectValue: [{
                type: Input
            }], selectLabel: [{
                type: Input
            }], selectDisabled: [{
                type: Input
            }], _isOpen: [{
                type: HostBinding,
                args: ['class.is-open']
            }] } });
