// @aposin/ng-aquila/esm2020/taglist/taglist.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxTagComponent } from './tag.component';
import { NxTaglistComponent } from './taglist.component';
import * as i0 from "@angular/core";
export class NxTaglistModule {
}
NxTaglistModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTaglistModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxTaglistModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTaglistModule, declarations: [NxTaglistComponent, NxTagComponent], imports: [CommonModule, NxIconModule], exports: [NxTaglistComponent, NxTagComponent] });
NxTaglistModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTaglistModule, imports: [[CommonModule, NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTaglistModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxTaglistComponent, NxTagComponent],
                    exports: [NxTaglistComponent, NxTagComponent],
                    imports: [CommonModule, NxIconModule],
                }]
        }] });
