// @aposin/ng-aquila/esm2020/list/list.module.mjs
import { NgModule } from '@angular/core';
import { NxIconComponent, NxIconModule } from '@aposin/ng-aquila/icon';
import { NxListComponent } from './list.component';
import { NxListIconComponent } from './list-icon.component';
import * as i0 from "@angular/core";
export class NxListModule {
}
NxListModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxListModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxListModule, bootstrap: [NxIconComponent], declarations: [NxListComponent, NxListIconComponent], imports: [NxIconModule], exports: [NxListComponent, NxListIconComponent] });
NxListModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxListModule, imports: [[NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxListModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxListComponent, NxListIconComponent],
                    imports: [NxIconModule],
                    exports: [NxListComponent, NxListIconComponent],
                    bootstrap: [NxIconComponent],
                }]
        }] });
