// @aposin/ng-aquila/esm2020/breadcrumb/breadcrumb.component.mjs
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { ChangeDetectionStrategy, Component, ContentChildren, Input } from '@angular/core';
import { Subject } from 'rxjs';
import { filter, startWith, takeUntil } from 'rxjs/operators';
import { NxBreadcrumbItemComponent } from './breadcrumb-item.component';
import * as i0 from "@angular/core";
export class NxBreadcrumbComponent {
    constructor(_cdr) {
        this._cdr = _cdr;
        this._negative = false;
        this._appeareance = 'default';
        this._destroyed = new Subject();
    }
    /**
     * Sets the appearance of the breadcrumb. default: 'default'
     */
    set appearance(value) {
        this._appeareance = value;
        this._cdr.markForCheck();
    }
    get appearance() {
        return this._appeareance;
    }
    /** Whether the component uses the negative styling. */
    set negative(value) {
        this._negative = coerceBooleanProperty(value);
        this._cdr.markForCheck();
    }
    get negative() {
        return this._negative;
    }
    ngAfterContentInit() {
        if (this.breadcrumbItems.length === 0) {
            console.warn('A breadcrumb needs NxBreadcrumbItemComponent children wrapped in <li>!');
        }
        this.breadcrumbItems.changes
            .pipe(startWith(this.breadcrumbItems), filter(items => items.length !== 0), takeUntil(this._destroyed))
            .subscribe(items => {
            this.breadcrumbItems.forEach(item => item.resetAriaLabel());
            this.breadcrumbItems.last.setAsLast();
        });
    }
    ngOnDestroy() {
        this._destroyed.next();
        this._destroyed.complete();
    }
}
NxBreadcrumbComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbComponent, deps: [{ token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
NxBreadcrumbComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxBreadcrumbComponent, selector: "ol[nxBreadcrumb]", inputs: { appearance: "appearance", negative: "negative" }, host: { properties: { "class.is-negative": "negative", "class.is-link": "appearance === \"link\"" } }, queries: [{ propertyName: "breadcrumbItems", predicate: NxBreadcrumbItemComponent, descendants: true }], ngImport: i0, template: "<ng-content></ng-content>\n", styles: [":host{display:flex;list-style:none;height:24px;padding:0}:host(.is-link) ::ng-deep .nx-breadcrumb-item{color:var(--breadcrumb-link-color)}:host(.is-link) ::ng-deep .nx-breadcrumb-item__text:hover{color:var(--breadcrumb-link-hover-color)}:host(.is-link) ::ng-deep .nx-breadcrumb-item__text:active{color:var(--breadcrumb-link-active-color)}::ng-deep li:last-child .nx-breadcrumb-item{font-weight:600;cursor:auto}::ng-deep li:last-child .nx-breadcrumb-item__chevron{display:none}::ng-deep li+li .nx-breadcrumb-item{margin-left:4px}:host-context([dir=rtl]) ::ng-deep li+li .nx-breadcrumb-item{margin-left:0;margin-left:initial;margin-right:4px}:host(.is-negative) ::ng-deep .nx-breadcrumb-item{color:var(--negative)}@media screen and (-ms-high-contrast: active){::ng-deep .nx-breadcrumb-item__chevron{color:windowText}}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ol[nxBreadcrumb]', changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        '[class.is-negative]': 'negative',
                        '[class.is-link]': 'appearance === "link"',
                    }, template: "<ng-content></ng-content>\n", styles: [":host{display:flex;list-style:none;height:24px;padding:0}:host(.is-link) ::ng-deep .nx-breadcrumb-item{color:var(--breadcrumb-link-color)}:host(.is-link) ::ng-deep .nx-breadcrumb-item__text:hover{color:var(--breadcrumb-link-hover-color)}:host(.is-link) ::ng-deep .nx-breadcrumb-item__text:active{color:var(--breadcrumb-link-active-color)}::ng-deep li:last-child .nx-breadcrumb-item{font-weight:600;cursor:auto}::ng-deep li:last-child .nx-breadcrumb-item__chevron{display:none}::ng-deep li+li .nx-breadcrumb-item{margin-left:4px}:host-context([dir=rtl]) ::ng-deep li+li .nx-breadcrumb-item{margin-left:0;margin-left:initial;margin-right:4px}:host(.is-negative) ::ng-deep .nx-breadcrumb-item{color:var(--negative)}@media screen and (-ms-high-contrast: active){::ng-deep .nx-breadcrumb-item__chevron{color:windowText}}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.ChangeDetectorRef }]; }, propDecorators: { appearance: [{
                type: Input
            }], negative: [{
                type: Input
            }], breadcrumbItems: [{
                type: ContentChildren,
                args: [NxBreadcrumbItemComponent, { descendants: true }]
            }] } });
