// @aposin/ng-aquila/esm2020/breadcrumb/breadcrumb.module.mjs
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxBreadcrumbComponent } from './breadcrumb.component';
import { NxBreadcrumbItemComponent } from './breadcrumb-item.component';
import * as i0 from "@angular/core";
export class NxBreadcrumbModule {
}
NxBreadcrumbModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxBreadcrumbModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbModule, declarations: [NxBreadcrumbItemComponent, NxBreadcrumbComponent], imports: [NxIconModule], exports: [NxBreadcrumbItemComponent, NxBreadcrumbComponent] });
NxBreadcrumbModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbModule, imports: [[NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxBreadcrumbModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxBreadcrumbItemComponent, NxBreadcrumbComponent],
                    exports: [NxBreadcrumbItemComponent, NxBreadcrumbComponent],
                    imports: [NxIconModule],
                }]
        }] });
