// @aposin/ng-aquila/esm2020/message/message.module.mjs
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxMessageComponent } from './message/message.component';
import { NxMessageBannerComponent } from './message-banner/message-banner.component';
import { NxMessageToastComponent } from './message-toast/message-toast.component';
import * as i0 from "@angular/core";
export class NxMessageModule {
}
NxMessageModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMessageModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxMessageModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMessageModule, declarations: [NxMessageComponent, NxMessageToastComponent, NxMessageBannerComponent], imports: [CommonModule, NxIconModule, OverlayModule, PortalModule], exports: [NxMessageComponent, NxMessageBannerComponent] });
NxMessageModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMessageModule, imports: [[CommonModule, NxIconModule, OverlayModule, PortalModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMessageModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, NxIconModule, OverlayModule, PortalModule],
                    declarations: [NxMessageComponent, NxMessageToastComponent, NxMessageBannerComponent],
                    exports: [NxMessageComponent, NxMessageBannerComponent],
                }]
        }] });
