// @aposin/ng-aquila/esm2020/avatar/avatar.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxAvatarButtonDirective, NxAvatarComponent } from './avatar';
import * as i0 from "@angular/core";
export class NxAvatarModule {
}
NxAvatarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAvatarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxAvatarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAvatarModule, declarations: [NxAvatarComponent, NxAvatarButtonDirective], imports: [CommonModule], exports: [NxAvatarComponent, NxAvatarButtonDirective] });
NxAvatarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAvatarModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAvatarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [NxAvatarComponent, NxAvatarButtonDirective],
                    exports: [NxAvatarComponent, NxAvatarButtonDirective],
                }]
        }] });
