// @aposin/ng-aquila/esm2020/dynamic-table/dynamic-table.module.mjs
import { ScrollingModule } from '@angular/cdk/scrolling';
import { CdkTableModule } from '@angular/cdk/table';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxDynamicTableComponent } from './dynamic-table.component';
import * as i0 from "@angular/core";
export class NxDynamicTableModule {
}
NxDynamicTableModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxDynamicTableModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxDynamicTableModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxDynamicTableModule, declarations: [NxDynamicTableComponent], imports: [CommonModule, CdkTableModule, ScrollingModule], exports: [NxDynamicTableComponent] });
NxDynamicTableModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxDynamicTableModule, imports: [[CommonModule, CdkTableModule, ScrollingModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxDynamicTableModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxDynamicTableComponent],
                    exports: [NxDynamicTableComponent],
                    imports: [CommonModule, CdkTableModule, ScrollingModule],
                }]
        }] });
