// @aposin/ng-aquila/esm2020/comparison-table/toggle-section/toggle-section.directive.mjs
import { ContentChild, ContentChildren, Directive, Input } from '@angular/core';
import { NxComparisonTableRowDirective } from '../comparison-table-row.directive';
import { NxTableContentElement } from '../table-content-element.directive';
import { NxToggleSectionBase } from './toggle-section-base';
import { NxToggleSectionHeaderComponent } from './toggle-section-header.component';
import * as i0 from "@angular/core";
export class NxToggleSectionDirective extends NxToggleSectionBase {
    constructor() {
        super(...arguments);
        this._isExpanded = true;
    }
    /** Whether the toggle section is expanded. Default: true. */
    set isExpanded(value) {
        if (this._isExpanded !== value) {
            this._isExpanded = value;
        }
    }
    get isExpanded() {
        return this._isExpanded;
    }
    _numberOfRows() {
        return this.rows.filter(row => row instanceof NxComparisonTableRowDirective).length;
    }
}
NxToggleSectionDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxToggleSectionDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NxToggleSectionDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxToggleSectionDirective, selector: "[nxComparisonTableToggleSection]", inputs: { isExpanded: "isExpanded" }, providers: [
        { provide: NxTableContentElement, useExisting: NxToggleSectionDirective },
        { provide: NxToggleSectionBase, useExisting: NxToggleSectionDirective },
    ], queries: [{ propertyName: "toggleSectionHeader", first: true, predicate: NxToggleSectionHeaderComponent, descendants: true }, { propertyName: "rows", predicate: NxTableContentElement }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxToggleSectionDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxComparisonTableToggleSection]',
                    providers: [
                        { provide: NxTableContentElement, useExisting: NxToggleSectionDirective },
                        { provide: NxToggleSectionBase, useExisting: NxToggleSectionDirective },
                    ],
                }]
        }], propDecorators: { toggleSectionHeader: [{
                type: ContentChild,
                args: [NxToggleSectionHeaderComponent, { static: false }]
            }], rows: [{
                type: ContentChildren,
                args: [NxTableContentElement]
            }], isExpanded: [{
                type: Input
            }] } });
