// @aposin/ng-aquila/esm2020/number-stepper/auto-resize.directive.mjs
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { Directive, HostBinding, Input } from '@angular/core';
import { getFontShorthand } from '@aposin/ng-aquila/utils';
import * as i0 from "@angular/core";
/** @docs-private */
export class NxAutoResizeDirective {
    constructor(_element, _renderer, _cdr) {
        this._element = _element;
        this._renderer = _renderer;
        this._cdr = _cdr;
        this._resize = true;
        this.updateInputWidth = this.updateInputWidth.bind(this);
    }
    set resize(value) {
        this._resize = coerceBooleanProperty(value);
        if (this._resize) {
            this._addEventListener();
            this.updateInputWidth();
        }
        else {
            this._removeEventListener();
        }
    }
    get resize() {
        return this._resize;
    }
    ngAfterViewInit() {
        if (this.resize) {
            this._addEventListener();
        }
    }
    ngOnDestroy() {
        this._removeEventListener();
    }
    updateInputWidth() {
        const measureCanvas = this._renderer.createElement('canvas');
        const ctx = measureCanvas.getContext('2d');
        const styles = window.getComputedStyle(this._element.nativeElement);
        ctx.font = getFontShorthand(styles);
        const metrics = ctx.measureText(this._element.nativeElement.value);
        const padding = this.sumStyles(styles.paddingLeft, styles.paddingRight);
        const border = this.sumStyles(styles.borderLeftWidth, styles.borderRightWidth);
        // the pixels are needed, because despite the correct calculation the last pixels of a number are always cut
        const newWidth = metrics.width + padding + border + 16;
        // Limit to own given minimal width
        const parsed = parseFloat(styles.minWidth);
        this.width = Math.max(Number.isNaN(parsed) ? 0 : parsed, newWidth);
        // needed when the outer component is onPush
        this._cdr.markForCheck();
    }
    _addEventListener() {
        this._element.nativeElement.addEventListener('input', this.updateInputWidth, true);
        this._element.nativeElement.addEventListener('change', this.updateInputWidth, true);
    }
    _removeEventListener() {
        this._element.nativeElement.removeEventListener('input', this.updateInputWidth, true);
        this._element.nativeElement.removeEventListener('change', this.updateInputWidth, true);
    }
    sumStyles(left, right) {
        return (parseInt(left, 10) || 0) + (parseInt(right, 10) || 0);
    }
}
NxAutoResizeDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAutoResizeDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Directive });
NxAutoResizeDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxAutoResizeDirective, selector: "input[nxAutoResize]", inputs: { resize: ["nxAutoResize", "resize"] }, host: { properties: { "style.width.px": "this.width" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAutoResizeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'input[nxAutoResize]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }, { type: i0.ChangeDetectorRef }]; }, propDecorators: { width: [{
                type: HostBinding,
                args: ['style.width.px']
            }], resize: [{
                type: Input,
                args: ['nxAutoResize']
            }] } });
