// @aposin/ng-aquila/esm2020/number-stepper/number-stepper-prefix.directive.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
/** Directive used to pass a suffix to the number stepper. */
export class NxNumberStepperPrefixDirective {
}
NxNumberStepperPrefixDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNumberStepperPrefixDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxNumberStepperPrefixDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxNumberStepperPrefixDirective, selector: "nx-number-stepper-prefix", host: { classAttribute: "nx-stepper__prefix" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNumberStepperPrefixDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'nx-number-stepper-prefix',
                    host: {
                        class: 'nx-stepper__prefix',
                    },
                }]
        }] });
