// @aposin/ng-aquila/esm2020/number-stepper/number-stepper-suffix.directive.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
/** Directive used to pass a suffix to the number stepper. */
export class NxNumberStepperSuffixDirective {
}
NxNumberStepperSuffixDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNumberStepperSuffixDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxNumberStepperSuffixDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxNumberStepperSuffixDirective, selector: "nx-number-stepper-suffix", host: { classAttribute: "nx-stepper__suffix" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNumberStepperSuffixDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'nx-number-stepper-suffix',
                    host: {
                        class: 'nx-stepper__suffix',
                    },
                }]
        }] });
