// @aposin/ng-aquila/esm2020/sidepanel/sidepanel.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxButtonModule } from '@aposin/ng-aquila/button';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxSidepanelComponent, NxSidepanelOuterContainerComponent } from './sidepanel';
import { NxSidepanelCloseButtonComponent } from './sidepanel-close-button';
import { NxSidepanelContentComponent } from './sidepanel-content';
import { NxSidepanelHeaderComponent } from './sidepanel-header';
import * as i0 from "@angular/core";
export class NxSidepanelModule {
}
NxSidepanelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSidepanelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxSidepanelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSidepanelModule, declarations: [NxSidepanelComponent,
        NxSidepanelHeaderComponent,
        NxSidepanelContentComponent,
        NxSidepanelCloseButtonComponent,
        NxSidepanelOuterContainerComponent], imports: [CommonModule, NxIconModule, NxButtonModule], exports: [NxSidepanelComponent,
        NxSidepanelHeaderComponent,
        NxSidepanelContentComponent,
        NxSidepanelCloseButtonComponent,
        NxSidepanelOuterContainerComponent] });
NxSidepanelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSidepanelModule, imports: [[CommonModule, NxIconModule, NxButtonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSidepanelModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, NxIconModule, NxButtonModule],
                    declarations: [
                        NxSidepanelComponent,
                        NxSidepanelHeaderComponent,
                        NxSidepanelContentComponent,
                        NxSidepanelCloseButtonComponent,
                        NxSidepanelOuterContainerComponent,
                    ],
                    exports: [
                        NxSidepanelComponent,
                        NxSidepanelHeaderComponent,
                        NxSidepanelContentComponent,
                        NxSidepanelCloseButtonComponent,
                        NxSidepanelOuterContainerComponent,
                    ],
                }]
        }] });
