// @aposin/ng-aquila/esm2020/menu/menu-item.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
/**
  This is a menu item is used to structure content whithin a menu.
 */
export class NxMenuItemDirective {
    constructor() {
        this._size = 's';
    }
    set size(value) {
        this._size = value === 'l' ? 'l' : 's';
    }
    get size() {
        return this._size;
    }
}
NxMenuItemDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMenuItemDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxMenuItemDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxMenuItemDirective, selector: "[nxMenuItem]", inputs: { size: ["nxMenuItem", "size"] }, host: { properties: { "class.nx-menu__item--small": "size === \"s\"", "class.nx-menu__item--large": "size === \"l\"" }, classAttribute: "nx-menu__item" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMenuItemDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxMenuItem]',
                    host: {
                        class: 'nx-menu__item',
                        '[class.nx-menu__item--small]': 'size === "s"',
                        '[class.nx-menu__item--large]': 'size === "l"',
                    },
                }]
        }], propDecorators: { size: [{
                type: Input,
                args: ['nxMenuItem']
            }] } });
