// @aposin/ng-aquila/esm2020/menu/menu.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxMenuComponent } from './menu.component';
import { NxMenuButtonComponent, NxMenuButtonIconDirective } from './menu-button.component';
import { NxMenuItemDirective } from './menu-item.directive';
import { NxMenuLinkDirective } from './menu-link.directive';
import * as i0 from "@angular/core";
const EXPORTED_MODULES = [NxMenuComponent, NxMenuLinkDirective, NxMenuItemDirective, NxMenuButtonComponent, NxMenuButtonIconDirective];
export class NxMenuModule {
}
NxMenuModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxMenuModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMenuModule, declarations: [NxMenuComponent, NxMenuLinkDirective, NxMenuItemDirective, NxMenuButtonComponent, NxMenuButtonIconDirective], imports: [CommonModule, NxIconModule], exports: [NxMenuComponent, NxMenuLinkDirective, NxMenuItemDirective, NxMenuButtonComponent, NxMenuButtonIconDirective] });
NxMenuModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMenuModule, imports: [[CommonModule, NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxMenuModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, NxIconModule],
                    declarations: EXPORTED_MODULES,
                    exports: EXPORTED_MODULES,
                }]
        }] });
