// @aposin/ng-aquila/esm2020/licence-plate/licence-plate.validator.mjs
import { Directive, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import * as i0 from "@angular/core";
const DE_SPECIAL_PATTERN = /^[\dA-Za-zÄÖÜäöü]{1,3}[ -]?[\dA-Za-z]{1,5}[ -]?[\dA-Za-z]{1,5}$/;
const DE_STANDARD_PATTERN = /^[A-Za-zÄÖÜäöü]{1,3}-[A-Za-z]{1,2} \d{1,4}[EHeh]?$/;
const OTHER_PATTERN = /^.+$/;
const LICENCE_PLATE_PATTERNS = {
    de_standard: DE_STANDARD_PATTERN,
    de_special: DE_SPECIAL_PATTERN,
    de_season: DE_STANDARD_PATTERN,
    other: OTHER_PATTERN,
    euro: OTHER_PATTERN,
};
/**
 * Licence plate validator.
 * Supports validation for german standard and special plate numbers.
 * Choose `other` to allow all values.
 */
export function nxLicensePlateValidator(type) {
    return (control) => {
        const valid = LICENCE_PLATE_PATTERNS[type].test(control.value);
        if (valid) {
            return null;
        }
        return { nxLicensePlateError: true };
    };
}
export class NxLicencePlateValidatorDirective {
    constructor() {
        /**
         * Licence plate type.
         * Supports validation for german standard and special plate numbers.
         * Choose `other` to allow all values.
         */
        this.type = 'other';
    }
    validate(control) {
        return nxLicensePlateValidator(this.type)(control);
    }
}
NxLicencePlateValidatorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLicencePlateValidatorDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxLicencePlateValidatorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxLicencePlateValidatorDirective, selector: "[nxLicensePlate]", inputs: { type: ["nxLicensePlate", "type"] }, providers: [
        {
            provide: NG_VALIDATORS,
            useExisting: NxLicencePlateValidatorDirective,
            multi: true,
        },
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLicencePlateValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxLicensePlate]',
                    providers: [
                        {
                            provide: NG_VALIDATORS,
                            useExisting: NxLicencePlateValidatorDirective,
                            multi: true,
                        },
                    ],
                }]
        }], propDecorators: { type: [{
                type: Input,
                args: ['nxLicensePlate']
            }] } });
