// @aposin/ng-aquila/esm2020/licence-plate/licence-plate.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxLicencePlateValidatorDirective } from './licence-plate.validator';
import { NxLicencePlateEuroPrefixComponent } from './licence-plate-euro-prefix.component';
import { NxLicencePlateSeasonSuffixComponent } from './licence-plate-season-suffix.component';
import * as i0 from "@angular/core";
export class NxLicencePlateModule {
}
NxLicencePlateModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLicencePlateModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxLicencePlateModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLicencePlateModule, declarations: [NxLicencePlateValidatorDirective, NxLicencePlateEuroPrefixComponent, NxLicencePlateSeasonSuffixComponent], imports: [CommonModule], exports: [NxLicencePlateValidatorDirective, NxLicencePlateEuroPrefixComponent, NxLicencePlateSeasonSuffixComponent] });
NxLicencePlateModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLicencePlateModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLicencePlateModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxLicencePlateValidatorDirective, NxLicencePlateEuroPrefixComponent, NxLicencePlateSeasonSuffixComponent],
                    imports: [CommonModule],
                    exports: [NxLicencePlateValidatorDirective, NxLicencePlateEuroPrefixComponent, NxLicencePlateSeasonSuffixComponent],
                }]
        }] });
