// @aposin/ng-aquila/esm2020/base/error/error.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxMessageModule } from '@aposin/ng-aquila/message';
import { NxErrorComponent } from './error.component';
import * as i0 from "@angular/core";
export class NxErrorModule {
}
NxErrorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxErrorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxErrorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxErrorModule, declarations: [NxErrorComponent], imports: [CommonModule, NxIconModule, NxMessageModule], exports: [NxErrorComponent] });
NxErrorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxErrorModule, imports: [[CommonModule, NxIconModule, NxMessageModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxErrorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, NxIconModule, NxMessageModule],
                    declarations: [NxErrorComponent],
                    exports: [NxErrorComponent],
                }]
        }] });
