// @aposin/ng-aquila/esm2020/comparison-table/comparison-table-base.mjs
import { Directive, EventEmitter } from '@angular/core';
import { NxBreakpoints } from '@aposin/ng-aquila/utils';
import { merge, Subject } from 'rxjs';
import { filter, mapTo, takeUntil } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@aposin/ng-aquila/utils";
/** @docs-private */
export class NxComparisonTableBase {
    constructor(viewportService, _cdr) {
        this.viewportService = viewportService;
        this._cdr = _cdr;
        this._disabledIndexes = [];
        this._hiddenIndexes = [];
        this.viewTypeChange = new EventEmitter();
        this._viewType = 'desktop';
        this._destroyed = new Subject();
        const mobile$ = this.viewportService.max(NxBreakpoints.BREAKPOINT_MEDIUM);
        const tablet$ = this.viewportService.between(NxBreakpoints.BREAKPOINT_MEDIUM, NxBreakpoints.BREAKPOINT_LARGE);
        const desktop$ = this.viewportService.min(NxBreakpoints.BREAKPOINT_LARGE);
        merge(mobile$.pipe(filter(value => value), mapTo('mobile')), tablet$.pipe(filter(value => value), mapTo('tablet')), desktop$.pipe(filter(value => value), mapTo('desktop')))
            .pipe(takeUntil(this._destroyed))
            .subscribe(value => {
            if (this._viewType !== value) {
                this.viewTypeChange.emit(value);
            }
            this._viewType = value;
            // We need to run change detection here or the view doesn't get updated
            // if the component is wrapped inside a parent with onPush change detection
            // we need mark for check as parts like the description cell look at the table's
            // viewType but because the description cell is part of ng-content it is not checked
            // when the comparison table view is checked but only when the component where the ng-content
            // gets declared is checked
            this._cdr.markForCheck();
        });
    }
    /** @docs-private */
    get viewType() {
        return this._viewType;
    }
    ngOnDestroy() {
        this._destroyed.next();
        this._destroyed.complete();
    }
}
NxComparisonTableBase.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxComparisonTableBase, deps: [{ token: i1.NxViewportService }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Directive });
NxComparisonTableBase.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxComparisonTableBase, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxComparisonTableBase, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i1.NxViewportService }, { type: i0.ChangeDetectorRef }]; } });
