// @aposin/ng-aquila/esm2020/comparison-table/comparison-table-row-group.directive.mjs
import { coerceBooleanProperty, coerceNumberProperty } from '@angular/cdk/coercion';
import { ContentChildren, Directive, EventEmitter, Inject, Input, Optional, Output } from '@angular/core';
import { COMPARISON_TABLE_DEFAULT_OPTIONS } from './comparison-table.models';
import { NxComparisonTableRowDirective } from './comparison-table-row.directive';
import { NxComparisonTableRowGroupBase } from './comparison-table-row-group-base';
import { NxTableContentElement } from './table-content-element.directive';
import * as i0 from "@angular/core";
import * as i1 from "./toggle-section/toggle-section-base";
export class NxComparisonTableRowGroupDirective extends NxComparisonTableRowGroupBase {
    constructor(_toggleSection, _defaultOptions) {
        super();
        this._toggleSection = _toggleSection;
        this._defaultOptions = _defaultOptions;
        this._labelCollapsed = 'More services';
        this._labelExpanded = 'Less services';
        this._visibleRows = 5;
        this._isExpanded = false;
        /** An event emitted every time the expanded state of the group changes */
        this.isExpandedChange = new EventEmitter();
    }
    /**
     * Sets the label of the expandable area that is shown when the row group is collapsed.
     */
    set labelCollapsed(newValue) {
        if (newValue !== this._labelCollapsed) {
            this._labelCollapsed = newValue;
        }
    }
    get labelCollapsed() {
        return this._labelCollapsed;
    }
    /** Sets the label of the expandable area that is shown when the row group is expanded. */
    set labelExpanded(newValue) {
        if (newValue !== this._labelExpanded) {
            this._labelExpanded = newValue;
        }
    }
    get labelExpanded() {
        return this._labelExpanded;
    }
    /** Sets the number of rows that are visible when loading the component. Default: 5. */
    set visibleRows(value) {
        const newValue = coerceNumberProperty(value);
        if (newValue !== this._visibleRows) {
            this._visibleRows = newValue;
        }
    }
    get visibleRows() {
        return this._visibleRows;
    }
    /** Sets the expanded state of the row group */
    set isExpanded(value) {
        const newValue = coerceBooleanProperty(value);
        if (newValue !== this._isExpanded) {
            this._isExpanded = newValue;
        }
    }
    get isExpanded() {
        return this._isExpanded;
    }
    /**
     * **Expert Option**
     *
     * Sets if the expandable area uses the full width of the row or leaves out the first column. Default: false.
     */
    set useFullRowForExpandableArea(value) {
        this._useFullRowForExpandableArea = coerceBooleanProperty(value);
    }
    get useFullRowForExpandableArea() {
        if (this._useFullRowForExpandableArea !== undefined) {
            return this._useFullRowForExpandableArea;
        }
        if (this._defaultOptions?.useFullRowForExpandableArea !== undefined) {
            return this._defaultOptions.useFullRowForExpandableArea;
        }
        return false;
    }
    _isPartOfToggleSection() {
        return !!this._toggleSection;
    }
    _handleIsExpandedChange(value) {
        this.isExpanded = value;
        this.isExpandedChange.emit(this.isExpanded);
    }
}
NxComparisonTableRowGroupDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxComparisonTableRowGroupDirective, deps: [{ token: i1.NxToggleSectionBase, optional: true }, { token: COMPARISON_TABLE_DEFAULT_OPTIONS, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
NxComparisonTableRowGroupDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxComparisonTableRowGroupDirective, selector: "[nxComparisonTableRowGroup]", inputs: { labelCollapsed: "labelCollapsed", labelExpanded: "labelExpanded", visibleRows: "visibleRows", isExpanded: "isExpanded", useFullRowForExpandableArea: "useFullRowForExpandableArea" }, outputs: { isExpandedChange: "isExpandedChange" }, providers: [
        { provide: NxTableContentElement, useExisting: NxComparisonTableRowGroupDirective },
        { provide: NxComparisonTableRowGroupBase, useExisting: NxComparisonTableRowGroupDirective },
    ], queries: [{ propertyName: "rows", predicate: NxComparisonTableRowDirective }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxComparisonTableRowGroupDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxComparisonTableRowGroup]',
                    providers: [
                        { provide: NxTableContentElement, useExisting: NxComparisonTableRowGroupDirective },
                        { provide: NxComparisonTableRowGroupBase, useExisting: NxComparisonTableRowGroupDirective },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i1.NxToggleSectionBase, decorators: [{
                    type: Optional
                }] }, { type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [COMPARISON_TABLE_DEFAULT_OPTIONS]
                }] }]; }, propDecorators: { rows: [{
                type: ContentChildren,
                args: [NxComparisonTableRowDirective]
            }], labelCollapsed: [{
                type: Input
            }], labelExpanded: [{
                type: Input
            }], visibleRows: [{
                type: Input
            }], isExpanded: [{
                type: Input
            }], useFullRowForExpandableArea: [{
                type: Input
            }], isExpandedChange: [{
                type: Output
            }] } });
