// @aposin/ng-aquila/esm2020/comparison-table/comparison-table.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxComparisonTableCell } from './cell/cell.component';
import { NxComparisonTableComponent } from './comparison-table.component';
import { NxComparisonTableRowDirective } from './comparison-table-row.directive';
import { NxComparisonTableRowGroupDirective } from './comparison-table-row-group.directive';
import { NxComparisonTableDescriptionCell } from './description-cell/description-cell.component';
import { NxComparisonTableDesktopGroup } from './desktop-group/desktop-group.component';
import { NxComparisonTableFlexRow } from './flex-row/flex-row.component';
import { NxComparisonTableIntersectionCell } from './intersection-cell/intersection-cell.component';
import { NxComparisonTablePopularCell } from './popular-cell/popular-cell.component';
import { NxComparisonTableSelectButton } from './select-button/select-button.component';
import { NxToggleSectionDirective } from './toggle-section/toggle-section.directive';
import { NxToggleSectionHeaderComponent } from './toggle-section/toggle-section-header.component';
import * as i0 from "@angular/core";
export class NxComparisonTableModule {
}
NxComparisonTableModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxComparisonTableModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxComparisonTableModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxComparisonTableModule, declarations: [NxComparisonTableComponent,
        NxComparisonTableCell,
        NxComparisonTableDescriptionCell,
        NxComparisonTableIntersectionCell,
        NxToggleSectionDirective,
        NxToggleSectionHeaderComponent,
        NxComparisonTableRowDirective,
        NxComparisonTableSelectButton,
        NxComparisonTableFlexRow,
        NxComparisonTableRowGroupDirective,
        NxComparisonTableDesktopGroup,
        NxComparisonTablePopularCell], imports: [CommonModule, NxIconModule], exports: [NxComparisonTableComponent,
        NxComparisonTableCell,
        NxComparisonTableDescriptionCell,
        NxComparisonTableIntersectionCell,
        NxToggleSectionDirective,
        NxToggleSectionHeaderComponent,
        NxComparisonTableRowDirective,
        NxComparisonTableSelectButton,
        NxComparisonTableFlexRow,
        NxComparisonTableRowGroupDirective,
        NxComparisonTableDesktopGroup,
        NxComparisonTablePopularCell] });
NxComparisonTableModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxComparisonTableModule, imports: [[CommonModule, NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxComparisonTableModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, NxIconModule],
                    declarations: [
                        NxComparisonTableComponent,
                        NxComparisonTableCell,
                        NxComparisonTableDescriptionCell,
                        NxComparisonTableIntersectionCell,
                        NxToggleSectionDirective,
                        NxToggleSectionHeaderComponent,
                        NxComparisonTableRowDirective,
                        NxComparisonTableSelectButton,
                        NxComparisonTableFlexRow,
                        NxComparisonTableRowGroupDirective,
                        NxComparisonTableDesktopGroup,
                        NxComparisonTablePopularCell,
                    ],
                    exports: [
                        NxComparisonTableComponent,
                        NxComparisonTableCell,
                        NxComparisonTableDescriptionCell,
                        NxComparisonTableIntersectionCell,
                        NxToggleSectionDirective,
                        NxToggleSectionHeaderComponent,
                        NxComparisonTableRowDirective,
                        NxComparisonTableSelectButton,
                        NxComparisonTableFlexRow,
                        NxComparisonTableRowGroupDirective,
                        NxComparisonTableDesktopGroup,
                        NxComparisonTablePopularCell,
                    ],
                }]
        }] });
