// @aposin/ng-aquila/esm2020/tooltip/tooltip.module.mjs
import { A11yModule } from '@angular/cdk/a11y';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxTooltipComponent } from './tooltip.component';
import { NX_TOOLTIP_SCROLL_STRATEGY_PROVIDER, NxTooltipDirective } from './tooltip.directive';
import * as i0 from "@angular/core";
export class NxTooltipModule {
}
NxTooltipModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTooltipModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxTooltipModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTooltipModule, declarations: [NxTooltipDirective, NxTooltipComponent], imports: [A11yModule, CommonModule, OverlayModule], exports: [NxTooltipDirective, NxTooltipComponent] });
NxTooltipModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTooltipModule, providers: [NX_TOOLTIP_SCROLL_STRATEGY_PROVIDER], imports: [[A11yModule, CommonModule, OverlayModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTooltipModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [A11yModule, CommonModule, OverlayModule],
                    exports: [NxTooltipDirective, NxTooltipComponent],
                    declarations: [NxTooltipDirective, NxTooltipComponent],
                    providers: [NX_TOOLTIP_SCROLL_STRATEGY_PROVIDER],
                }]
        }] });
