// @aposin/ng-aquila/esm2020/pagination/pagination.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxPaginationComponent } from './pagination.component';
import { NxPaginationUtils } from './pagination-utils';
import * as i0 from "@angular/core";
export class NxPaginationModule {
}
NxPaginationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPaginationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxPaginationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPaginationModule, declarations: [NxPaginationComponent], imports: [CommonModule, NxIconModule], exports: [NxPaginationComponent] });
NxPaginationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPaginationModule, providers: [NxPaginationUtils], imports: [[CommonModule, NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPaginationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, NxIconModule],
                    declarations: [NxPaginationComponent],
                    exports: [NxPaginationComponent],
                    providers: [NxPaginationUtils],
                }]
        }] });
