// @aposin/ng-aquila/esm2020/grid/grid.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxColComponent } from './col.component';
import { NxLayoutComponent } from './layout.component';
import { NxRowComponent } from './row.component';
import * as i0 from "@angular/core";
export class NxGridModule {
}
NxGridModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxGridModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxGridModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxGridModule, declarations: [NxLayoutComponent, NxRowComponent, NxColComponent], imports: [CommonModule], exports: [NxLayoutComponent, NxRowComponent, NxColComponent] });
NxGridModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxGridModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxGridModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxLayoutComponent, NxRowComponent, NxColComponent],
                    exports: [NxLayoutComponent, NxRowComponent, NxColComponent],
                    imports: [CommonModule],
                }]
        }] });
