// @aposin/ng-aquila/esm2020/grid/layout.component.mjs
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class NxLayoutComponent {
    constructor() {
        this._classNames = '';
        /** @docs-private */
        this.grid = true;
    }
    /**
     * Type of layout.
     *
     * Values: grid | grid nogutters | grid maxwidth | grid nopadding. Default value: grid.
     */
    set classNames(value) {
        if (this._classNames === value) {
            return;
        }
        this._classNames = value;
        // TODO kick null safe-guards after setter value or any calling input values are properly coerced as string
        this.grid = !!this._classNames?.includes('grid');
        this.noGutters = !!this._classNames?.includes('nogutters');
        this.maxWidth = !!this._classNames?.includes('maxwidth');
        this.noPadding = !!this._classNames?.includes('nopadding');
    }
    get classNames() {
        return this._classNames;
    }
}
NxLayoutComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLayoutComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
NxLayoutComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxLayoutComponent, selector: "[nxLayout]", inputs: { classNames: ["nxLayout", "classNames"] }, host: { properties: { "class.nx-grid": "grid", "class.nx-grid--no-gutters": "noGutters", "class.nx-grid--max-width": "maxWidth", "class.nx-grid--no-padding": "noPadding" } }, ngImport: i0, template: '<ng-content></ng-content>', isInline: true, styles: [":host(.nx-grid){position:relative;margin-left:auto;margin-right:auto;padding-right:16px;padding-left:16px}@media (min-width: 704px){:host(.nx-grid){padding-right:32px;padding-left:32px}}:host(.nx-grid--no-gutters) ::ng-deep{padding-left:0;padding-right:0}:host(.nx-grid--no-gutters) ::ng-deep .nx-grid__row{margin-left:0;margin-right:0}:host(.nx-grid--no-gutters) ::ng-deep .nx-grid__row>[class*=nx-grid__column]{padding-left:0;padding-right:0}:host(.nx-grid--no-padding) ::ng-deep{padding-left:0;padding-right:0}:host(.nx-grid--no-padding) ::ng-deep .nx-grid__row{margin-left:-16px;margin-right:-16px}@media (max-width: 703px){:host(.nx-grid--no-padding) ::ng-deep .nx-grid__row{margin-left:-8px;margin-right:-8px}}:host(.nx-grid--max-width){max-width:100%;width:var(--grid-max-width)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLayoutComponent, decorators: [{
            type: Component,
            args: [{ selector: '[nxLayout]', template: '<ng-content></ng-content>', changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        '[class.nx-grid]': 'grid',
                        '[class.nx-grid--no-gutters]': 'noGutters',
                        '[class.nx-grid--max-width]': 'maxWidth',
                        '[class.nx-grid--no-padding]': 'noPadding',
                    }, styles: [":host(.nx-grid){position:relative;margin-left:auto;margin-right:auto;padding-right:16px;padding-left:16px}@media (min-width: 704px){:host(.nx-grid){padding-right:32px;padding-left:32px}}:host(.nx-grid--no-gutters) ::ng-deep{padding-left:0;padding-right:0}:host(.nx-grid--no-gutters) ::ng-deep .nx-grid__row{margin-left:0;margin-right:0}:host(.nx-grid--no-gutters) ::ng-deep .nx-grid__row>[class*=nx-grid__column]{padding-left:0;padding-right:0}:host(.nx-grid--no-padding) ::ng-deep{padding-left:0;padding-right:0}:host(.nx-grid--no-padding) ::ng-deep .nx-grid__row{margin-left:-16px;margin-right:-16px}@media (max-width: 703px){:host(.nx-grid--no-padding) ::ng-deep .nx-grid__row{margin-left:-8px;margin-right:-8px}}:host(.nx-grid--max-width){max-width:100%;width:var(--grid-max-width)}\n"] }]
        }], propDecorators: { classNames: [{
                type: Input,
                args: ['nxLayout']
            }] } });
