// @aposin/ng-aquila/esm2020/table/sort-header/sort-header-intl.mjs
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class NxSortHeaderIntl {
    constructor() {
        /**
         * Stream that emits whenever the labels here are changed. Use this to notify
         * components if the labels have changed after initialization.
         */
        this.changes = new Subject();
        /** The aria label for a header that can be clicked to sort ascending. */
        this.sortAscendingAriaLabel = 'click to sort ascending';
        /** The aria label for a header that can be clicked to sort descending. */
        this.sortDescendingAriaLabel = 'click to sort descending';
        /** The aria label for a column that is sorted ascending. */
        this.sortedAscendingAriaLabel = 'sorted ascending by';
        /** The aria label for a column that is sorted descending. */
        this.sortedDescendingAriaLabel = 'sorted descending by';
    }
}
NxSortHeaderIntl.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSortHeaderIntl, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
NxSortHeaderIntl.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSortHeaderIntl });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSortHeaderIntl, decorators: [{
            type: Injectable
        }] });
