// @aposin/ng-aquila/esm2020/table/sort-header/sort.directive.mjs
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class SortEvent {
    constructor(active, direction) {
        this.active = active;
        this.direction = direction;
    }
}
export class NxSortDirective {
    constructor() {
        this._stateChanges = new Subject();
        this._direction = 'asc';
        /**
         * @docs-private
         * An event emitted when the active value has changed.
         */
        this.activeChange = new EventEmitter();
        /**
         * @docs-private
         * An event emitted when the direction value has changed.
         */
        this.directionChange = new EventEmitter();
        /** An event emitted when either the active sort or sort direction changes. */
        this.sortChange = new EventEmitter();
    }
    /** Sets the key of the most recently sorted column. */
    set active(value) {
        if (this._active !== value) {
            this._active = value;
            this.sortChange.emit(new SortEvent(this.active, this.direction));
            this._stateChanges.next();
        }
    }
    get active() {
        return this._active;
    }
    /** Sets the direction of the currently active sorted column. Default: 'asc'. */
    set direction(value) {
        if (this._direction !== value) {
            this._direction = value;
            this.sortChange.emit(new SortEvent(this.active, this.direction));
            this._stateChanges.next();
        }
    }
    get direction() {
        return this._direction;
    }
    ngOnDestroy() {
        this._stateChanges.complete();
    }
    /**
     * Sets the active sort key and determines the new sort direction.
     * Afterwards the output event `sortChange` is called.
     *
     * If `sortable` is not the key of the active sort header, the initial direction to sort is 'asc'.
     * Otherwise the direction of the sorted column changes.
     */
    sort(sortable) {
        if (this._active !== sortable) {
            this._active = sortable;
            this.activeChange.emit(this.active);
            this._direction = 'asc';
            this.directionChange.emit(this.direction);
        }
        else if (this._direction === 'asc') {
            this._direction = 'desc';
            this.directionChange.emit(this.direction);
        }
        else {
            this._direction = 'asc';
            this.directionChange.emit(this.direction);
        }
        this.sortChange.emit(new SortEvent(this.active, this.direction));
        this._stateChanges.next();
    }
}
NxSortDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSortDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxSortDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxSortDirective, selector: "table[nxSort]", inputs: { active: "active", direction: "direction" }, outputs: { activeChange: "activeChange", directionChange: "directionChange", sortChange: "sortChange" }, exportAs: ["NxSortDirective"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSortDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'table[nxSort]',
                    exportAs: 'NxSortDirective',
                }]
        }], propDecorators: { active: [{
                type: Input
            }], direction: [{
                type: Input
            }], activeChange: [{
                type: Output
            }], directionChange: [{
                type: Output
            }], sortChange: [{
                type: Output
            }] } });
