// @aposin/ng-aquila/esm2020/accordion/accordion.module.mjs
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxAccordionDirective } from './accordion';
import { NxExpansionPanelComponent } from './expansion-panel';
import { NxExpansionPanelBodyDirective } from './expansion-panel-body';
import { NxExpansionPanelDescriptionDirective, NxExpansionPanelHeaderComponent, NxExpansionPanelTitleDirective } from './expansion-panel-header';
import * as i0 from "@angular/core";
export class NxAccordionModule {
}
NxAccordionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAccordionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxAccordionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAccordionModule, declarations: [NxAccordionDirective,
        NxExpansionPanelComponent,
        NxExpansionPanelHeaderComponent,
        NxExpansionPanelDescriptionDirective,
        NxExpansionPanelTitleDirective,
        NxExpansionPanelBodyDirective], imports: [CommonModule, PortalModule, NxIconModule], exports: [NxAccordionDirective,
        NxExpansionPanelComponent,
        NxExpansionPanelHeaderComponent,
        NxExpansionPanelDescriptionDirective,
        NxExpansionPanelTitleDirective,
        NxExpansionPanelBodyDirective] });
NxAccordionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAccordionModule, providers: [], imports: [[CommonModule, PortalModule, NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAccordionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PortalModule, NxIconModule],
                    exports: [
                        NxAccordionDirective,
                        NxExpansionPanelComponent,
                        NxExpansionPanelHeaderComponent,
                        NxExpansionPanelDescriptionDirective,
                        NxExpansionPanelTitleDirective,
                        NxExpansionPanelBodyDirective,
                    ],
                    declarations: [
                        NxAccordionDirective,
                        NxExpansionPanelComponent,
                        NxExpansionPanelHeaderComponent,
                        NxExpansionPanelDescriptionDirective,
                        NxExpansionPanelTitleDirective,
                        NxExpansionPanelBodyDirective,
                    ],
                    providers: [],
                }]
        }] });
