// @aposin/ng-aquila/esm2020/accordion/accordion.mjs
import { CdkAccordion } from '@angular/cdk/accordion';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
const DEFAULT_TYPE = 'regular';
export class NxAccordionDirective extends CdkAccordion {
    constructor() {
        super(...arguments);
        this._style = 'regular';
    }
    /**
     * Value for the styling that should be chosen.
     * Default value: 'regular'
     */
    set style(value) {
        value = value ? value : DEFAULT_TYPE;
        const [newValue] = value.match(/regular|light|extra-light/) || [DEFAULT_TYPE];
        this._style = newValue;
    }
    get style() {
        return this._style;
    }
    /** Whether the negative set of styles should be used. */
    set negative(value) {
        this._negative = coerceBooleanProperty(value);
    }
    get negative() {
        return !!this._negative;
    }
}
NxAccordionDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAccordionDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NxAccordionDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxAccordionDirective, selector: "nx-accordion", inputs: { style: ["nxStyle", "style"], negative: "negative" }, host: { attributes: { "role": "presentation" }, properties: { "class.nx-accordion": "true" } }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAccordionDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'nx-accordion',
                    host: {
                        '[class.nx-accordion]': 'true',
                        role: 'presentation',
                    },
                }]
        }], propDecorators: { style: [{
                type: Input,
                args: ['nxStyle']
            }], negative: [{
                type: Input
            }] } });
