// @aposin/ng-aquila/esm2020/base/label/label.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxLabelComponent } from './label.component';
import * as i0 from "@angular/core";
export class NxLabelModule {
}
NxLabelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLabelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxLabelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLabelModule, declarations: [NxLabelComponent], imports: [CommonModule], exports: [NxLabelComponent] });
NxLabelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLabelModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLabelModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [NxLabelComponent],
                    exports: [NxLabelComponent],
                }]
        }] });
