// @aposin/ng-aquila/esm2020/base/label/label.component.mjs
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { ChangeDetectionStrategy, Component, Inject, InjectionToken, Input, Optional } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
let nextId = 0;
const DEFAULT_SIZE = 'large';
export const LABEL_DEFAULT_OPTIONS = new InjectionToken('LABEL_DEFAULT_OPTIONS');
export class NxLabelComponent {
    constructor(_defaultOptions, _cdr) {
        this._defaultOptions = _defaultOptions;
        this._cdr = _cdr;
        this._stateChanges = new Subject();
        this._disabled = false;
        this._negative = false;
        this._for = null;
        this._id = `nx-label-${nextId++}`;
    }
    /** Sets the label to disabled */
    set disabled(value) {
        this._disabled = coerceBooleanProperty(value);
        this._stateChanges.next();
    }
    get disabled() {
        return this._disabled;
    }
    /** Sets the label to disabled */
    set negative(value) {
        this._negative = coerceBooleanProperty(value);
        this._stateChanges.next();
    }
    get negative() {
        return this._negative;
    }
    /** Sets the Id of the label */
    set id(value) {
        this._id = value;
    }
    get id() {
        return this._id;
    }
    /**
     * **Expert option**
     * Sets the appearance of the label */
    set size(value) {
        this._size = value;
        this._stateChanges.next();
    }
    get size() {
        return this._size || this._defaultOptions?.size || DEFAULT_SIZE;
    }
    /**
     * Sets the html `for` attribute on the label.
     */
    set for(value) {
        this._for = value;
        this._cdr.markForCheck();
        this._stateChanges.next();
    }
    get for() {
        return this._for;
    }
}
NxLabelComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLabelComponent, deps: [{ token: LABEL_DEFAULT_OPTIONS, optional: true }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
NxLabelComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxLabelComponent, selector: "nx-label", inputs: { disabled: "disabled", negative: "negative", id: "id", size: "size", for: "for" }, host: { properties: { "attr.disabled": "disabled", "class.nx-label--negative": "negative", "class.nx-label--large": "size === \"large\"", "class.nx-label--small": "size === \"small\"" } }, ngImport: i0, template: "<label class=\"nx-label__content\" [for]=\"for\" [id]=\"id\">\n    <ng-content></ng-content>\n</label>\n", styles: [":host{display:block;color:var(--text-01)}:host.nx-label--small{font-size:var(--base-label-small-font-size);line-height:var(--base-label-small-line-height);font-weight:var(--base-label-small-font-weight);letter-spacing:var(--base-label-small-letter-spacing)}:host.nx-label--large{font-size:var(--base-label-large-font-size);line-height:var(--base-label-large-line-height);font-weight:var(--base-label-large-font-weight);letter-spacing:var(--base-label-large-letter-spacing)}:host[disabled=true] label{cursor:not-allowed}:host.nx-label--negative{color:var(--negative)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxLabelComponent, decorators: [{
            type: Component,
            args: [{ selector: 'nx-label', changeDetection: ChangeDetectionStrategy.OnPush, host: {
                        '[attr.disabled]': 'disabled',
                        '[class.nx-label--negative]': 'negative',
                        '[class.nx-label--large]': 'size === "large"',
                        '[class.nx-label--small]': 'size === "small"',
                    }, template: "<label class=\"nx-label__content\" [for]=\"for\" [id]=\"id\">\n    <ng-content></ng-content>\n</label>\n", styles: [":host{display:block;color:var(--text-01)}:host.nx-label--small{font-size:var(--base-label-small-font-size);line-height:var(--base-label-small-line-height);font-weight:var(--base-label-small-font-weight);letter-spacing:var(--base-label-small-letter-spacing)}:host.nx-label--large{font-size:var(--base-label-large-font-size);line-height:var(--base-label-large-line-height);font-weight:var(--base-label-large-font-weight);letter-spacing:var(--base-label-large-letter-spacing)}:host[disabled=true] label{cursor:not-allowed}:host.nx-label--negative{color:var(--negative)}\n"] }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [LABEL_DEFAULT_OPTIONS]
                }] }, { type: i0.ChangeDetectorRef }]; }, propDecorators: { disabled: [{
                type: Input
            }], negative: [{
                type: Input
            }], id: [{
                type: Input
            }], size: [{
                type: Input
            }], for: [{
                type: Input
            }] } });
