// @aposin/ng-aquila/esm2020/data-display/data-display.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxDataDisplayComponent } from './data-display/data-display.component';
import { NxDataDisplayLabelComponent } from './data-display-label/data-display-label.component';
import * as i0 from "@angular/core";
export class NxDataDisplayModule {
}
NxDataDisplayModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxDataDisplayModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxDataDisplayModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxDataDisplayModule, declarations: [NxDataDisplayComponent, NxDataDisplayLabelComponent], imports: [CommonModule], exports: [NxDataDisplayComponent, NxDataDisplayLabelComponent] });
NxDataDisplayModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxDataDisplayModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxDataDisplayModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [NxDataDisplayComponent, NxDataDisplayLabelComponent],
                    exports: [NxDataDisplayComponent, NxDataDisplayLabelComponent],
                }]
        }] });
