// @aposin/ng-aquila/esm2020/formfield/appendix.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
let uniqueId = 0;
// This Directive solely purpose is to mark given ng-content and project it into the required destination.
export class NxFormfieldAppendixDirective {
    constructor() {
        /** Sets the id of the formfield appendix. */
        this.id = `nx-formfield-appendix-${uniqueId++}`;
    }
}
NxFormfieldAppendixDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFormfieldAppendixDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxFormfieldAppendixDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxFormfieldAppendixDirective, selector: "[nxFormfieldAppendix]", inputs: { id: "id" }, host: { properties: { "attr.id": "id" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFormfieldAppendixDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxFormfieldAppendix]',
                    host: {
                        '[attr.id]': 'id',
                    },
                }]
        }], propDecorators: { id: [{
                type: Input
            }] } });
