// @aposin/ng-aquila/esm2020/formfield/error.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
let uniqueId = 0;
export class NxFormfieldErrorDirective {
    constructor() {
        // create a unique id to be used by aria-described-by
        /** Sets the id of the formfield error. */
        this.id = `nx-formfield-error-${uniqueId++}`;
    }
}
NxFormfieldErrorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFormfieldErrorDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxFormfieldErrorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxFormfieldErrorDirective, selector: "[nxFormfieldError], [nxError]", inputs: { id: "id" }, host: { attributes: { "role": "alert" }, properties: { "attr.id": "id" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFormfieldErrorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxFormfieldError], [nxError]',
                    host: {
                        role: 'alert',
                        '[attr.id]': 'id',
                    },
                }]
        }], propDecorators: { id: [{
                type: Input
            }] } });
