// @aposin/ng-aquila/esm2020/formfield/formfield.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxErrorModule, NxLabelModule } from '@aposin/ng-aquila/base';
import { NxFormfieldAppendixDirective } from './appendix.directive';
import { NxFormfieldErrorDirective } from './error.directive';
import { NxFormfieldComponent } from './formfield.component';
import { NxFormfieldHintDirective } from './hint.directive';
import { NxFormfieldLabelDirective } from './label.directive';
import { NxFormfieldNoteDirective } from './note.directive';
import { NxFormfieldPrefixDirective } from './prefix.directive';
import { NxFormfieldSuffixDirective } from './suffix.directive';
import * as i0 from "@angular/core";
export class NxFormfieldModule {
}
NxFormfieldModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFormfieldModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxFormfieldModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFormfieldModule, declarations: [NxFormfieldComponent,
        NxFormfieldHintDirective,
        NxFormfieldErrorDirective,
        NxFormfieldNoteDirective,
        NxFormfieldPrefixDirective,
        NxFormfieldSuffixDirective,
        NxFormfieldAppendixDirective,
        NxFormfieldLabelDirective], imports: [CommonModule], exports: [NxFormfieldComponent,
        NxFormfieldHintDirective,
        NxFormfieldErrorDirective,
        NxFormfieldNoteDirective,
        NxFormfieldPrefixDirective,
        NxFormfieldSuffixDirective,
        NxFormfieldAppendixDirective,
        NxFormfieldLabelDirective,
        NxErrorModule,
        NxLabelModule] });
NxFormfieldModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFormfieldModule, imports: [[CommonModule], NxErrorModule,
        NxLabelModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFormfieldModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [
                        NxFormfieldComponent,
                        NxFormfieldHintDirective,
                        NxFormfieldErrorDirective,
                        NxFormfieldNoteDirective,
                        NxFormfieldPrefixDirective,
                        NxFormfieldSuffixDirective,
                        NxFormfieldAppendixDirective,
                        NxFormfieldLabelDirective,
                    ],
                    exports: [
                        NxFormfieldComponent,
                        NxFormfieldHintDirective,
                        NxFormfieldErrorDirective,
                        NxFormfieldNoteDirective,
                        NxFormfieldPrefixDirective,
                        NxFormfieldSuffixDirective,
                        NxFormfieldAppendixDirective,
                        NxFormfieldLabelDirective,
                        NxErrorModule,
                        NxLabelModule,
                    ],
                }]
        }] });
