// @aposin/ng-aquila/esm2020/formfield/hint.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
let uniqueId = 0;
// This Directive solely purpose is to mark given ng-content and project it into the required destination.
export class NxFormfieldHintDirective {
    constructor() {
        /** Sets the id of the formfield hint. */
        this.id = `nx-formfield-hint-${uniqueId++}`;
    }
}
NxFormfieldHintDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFormfieldHintDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxFormfieldHintDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxFormfieldHintDirective, selector: "[nxFormfieldHint]", inputs: { id: "id" }, host: { properties: { "attr.id": "id", "class.nx-formfield-hint": "true" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFormfieldHintDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxFormfieldHint]',
                    host: {
                        '[attr.id]': 'id',
                        '[class.nx-formfield-hint]': 'true',
                    },
                }]
        }], propDecorators: { id: [{
                type: Input
            }] } });
