// @aposin/ng-aquila/esm2020/formfield/note.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
let uniqueId = 0;
// This Directive solely purpose is to mark given ng-content and project it into the required destination.
export class NxFormfieldNoteDirective {
    constructor() {
        /** Sets the id of the formfield note. */
        this.id = `nx-formfield-note-${uniqueId++}`;
    }
}
NxFormfieldNoteDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFormfieldNoteDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxFormfieldNoteDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxFormfieldNoteDirective, selector: "[nxFormfieldNote]", inputs: { id: "id" }, host: { properties: { "attr.id": "id" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxFormfieldNoteDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[nxFormfieldNote]',
                    host: {
                        '[attr.id]': 'id',
                    },
                }]
        }], propDecorators: { id: [{
                type: Input
            }] } });
