// @aposin/ng-aquila/esm2020/indicator/indicator.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIndicatorComponent } from './indicator.component';
import * as i0 from "@angular/core";
export class NxIndicatorModule {
}
NxIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxIndicatorModule, declarations: [NxIndicatorComponent], imports: [CommonModule], exports: [NxIndicatorComponent] });
NxIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxIndicatorModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [NxIndicatorComponent],
                    exports: [NxIndicatorComponent],
                }]
        }] });
