// @aposin/ng-aquila/esm2020/progress-stepper/buttons.mjs
import { CdkStepper, CdkStepperNext, CdkStepperPrevious } from '@angular/cdk/stepper';
import { Directive } from '@angular/core';
import { NxProgressStepperDirective } from './progress-stepper.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/stepper";
export class NxStepperNextDirective extends CdkStepperNext {
    constructor(_stepper, _cdr) {
        super(_stepper);
        this._cdr = _cdr;
        /** @docs-private */
        this.disabled = false;
    }
    /**
     * Implemented to prevent changed after checked error after stepper init.
     * When the stepper initializes the directive is checked first before
     * the stepper can check its content thus it doesn't know about it's children yet
     * so _stepper.hasNext returns false and disables the buttons. In the same CD cycle
     * after the stepper has checked its contents the button needs to be enabled resulting
     * in the error.
     */
    ngDoCheck() {
        const stepper = this._stepper;
        if (this.disabled !== !stepper.hasNext) {
            this.disabled = !stepper.hasNext;
            this._cdr.markForCheck();
        }
    }
}
NxStepperNextDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxStepperNextDirective, deps: [{ token: i1.CdkStepper }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Directive });
NxStepperNextDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxStepperNextDirective, selector: "button[nxStepperNext]", inputs: { type: "type" }, host: { properties: { "type": "type", "disabled": "disabled" } }, providers: [{ provide: CdkStepper, useExisting: NxProgressStepperDirective }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxStepperNextDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button[nxStepperNext]',
                    host: {
                        '[type]': 'type',
                        '[disabled]': 'disabled',
                    },
                    inputs: ['type'],
                    providers: [{ provide: CdkStepper, useExisting: NxProgressStepperDirective }],
                }]
        }], ctorParameters: function () { return [{ type: i1.CdkStepper }, { type: i0.ChangeDetectorRef }]; } });
export class NxStepperPreviousDirective extends CdkStepperPrevious {
    constructor(_stepper) {
        super(_stepper);
    }
    get stepper() {
        return this._stepper;
    }
}
NxStepperPreviousDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxStepperPreviousDirective, deps: [{ token: i1.CdkStepper }], target: i0.ɵɵFactoryTarget.Directive });
NxStepperPreviousDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxStepperPreviousDirective, selector: "button[nxStepperPrevious]", inputs: { type: "type" }, host: { properties: { "type": "type", "disabled": "!stepper.hasPrevious" } }, providers: [{ provide: CdkStepper, useExisting: NxProgressStepperDirective }], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxStepperPreviousDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button[nxStepperPrevious]',
                    host: {
                        '[type]': 'type',
                        '[disabled]': '!stepper.hasPrevious',
                    },
                    inputs: ['type'],
                    providers: [{ provide: CdkStepper, useExisting: NxProgressStepperDirective }],
                }]
        }], ctorParameters: function () { return [{ type: i1.CdkStepper }]; } });
