// @aposin/ng-aquila/esm2020/progress-stepper/progress-stepper.module.mjs
import { CdkStepperModule } from '@angular/cdk/stepper';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxLabelModule } from '@aposin/ng-aquila/base';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxProgressbarModule } from '@aposin/ng-aquila/progressbar';
import { NxStepperNextDirective, NxStepperPreviousDirective } from './buttons';
import { NxMultiStepperComponent } from './multi/multi-step.component';
import { NxMultiStepperGroupComponent } from './multi/multi-step-group.component';
import { NxMultiStepItemComponent } from './multi/multi-step-item.component';
import { NxProgressStepperComponent } from './progress/progress.component';
import { NxProgressStepperDirective, NxStepComponent } from './progress-stepper.component';
import { NxSingleStepperComponent } from './single/single-step.component';
import * as i0 from "@angular/core";
export class NxProgressStepperModule {
}
NxProgressStepperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressStepperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxProgressStepperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressStepperModule, declarations: [NxProgressStepperDirective,
        NxMultiStepperComponent,
        NxStepComponent,
        NxStepperNextDirective,
        NxStepperPreviousDirective,
        NxProgressStepperComponent,
        NxSingleStepperComponent,
        NxMultiStepItemComponent,
        NxMultiStepperGroupComponent], imports: [CommonModule, CdkStepperModule, NxIconModule, NxProgressbarModule, NxLabelModule], exports: [NxProgressStepperDirective,
        NxSingleStepperComponent,
        NxProgressStepperComponent,
        NxMultiStepperComponent,
        NxStepComponent,
        NxStepperNextDirective,
        NxStepperPreviousDirective,
        NxMultiStepperGroupComponent] });
NxProgressStepperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressStepperModule, imports: [[CommonModule, CdkStepperModule, NxIconModule, NxProgressbarModule, NxLabelModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxProgressStepperModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, CdkStepperModule, NxIconModule, NxProgressbarModule, NxLabelModule],
                    declarations: [
                        NxProgressStepperDirective,
                        NxMultiStepperComponent,
                        NxStepComponent,
                        NxStepperNextDirective,
                        NxStepperPreviousDirective,
                        NxProgressStepperComponent,
                        NxSingleStepperComponent,
                        NxMultiStepItemComponent,
                        NxMultiStepperGroupComponent,
                    ],
                    exports: [
                        NxProgressStepperDirective,
                        NxSingleStepperComponent,
                        NxProgressStepperComponent,
                        NxMultiStepperComponent,
                        NxStepComponent,
                        NxStepperNextDirective,
                        NxStepperPreviousDirective,
                        NxMultiStepperGroupComponent,
                    ],
                }]
        }] });
