// @aposin/ng-aquila/esm2020/input/input.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxFormfieldModule } from '@aposin/ng-aquila/formfield';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxInputDirective } from './input.directive';
import { NxPasswordToggleComponent } from './password-toggle.component';
import * as i0 from "@angular/core";
export class NxInputModule {
}
NxInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxInputModule, declarations: [NxInputDirective, NxPasswordToggleComponent], imports: [CommonModule, NxFormfieldModule, NxIconModule], exports: [NxFormfieldModule, NxInputDirective, NxPasswordToggleComponent] });
NxInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxInputModule, imports: [[CommonModule, NxFormfieldModule, NxIconModule], NxFormfieldModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxInputModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, NxFormfieldModule, NxIconModule],
                    declarations: [NxInputDirective, NxPasswordToggleComponent],
                    exports: [NxFormfieldModule, NxInputDirective, NxPasswordToggleComponent],
                }]
        }] });
