// @aposin/ng-aquila/esm2020/input/password-toggle.component.mjs
import { ENTER, SPACE } from '@angular/cdk/keycodes';
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/a11y";
import * as i2 from "@aposin/ng-aquila/icon";
const visibilityIcons = {
    show: 'password-show-o',
    hide: 'password-hide-o',
};
export class NxPasswordToggleComponent {
    constructor(_cdr, _elementRef, _focusMonitor) {
        this._cdr = _cdr;
        this._elementRef = _elementRef;
        this._focusMonitor = _focusMonitor;
        /** @docs-private */
        this._currentIcon = visibilityIcons['show'];
        /** @docs-private */
        this._pressed = false;
        this._ariaLabel = 'Show password';
        this._focusMonitor.monitor(this._elementRef);
    }
    /** Sets the aria-label needed for accessibility.
     * Notice that this `aria-label` should describe the initial action according to the status of the visibility.
     * E.g if you have an `input[type=password]` at the beginning then the
     * password will be hidden and the correct aria-label would be 'Show password.'
     */
    set ariaLabel(value) {
        if (value !== this._ariaLabel) {
            this._ariaLabel = value;
            this._cdr.markForCheck();
        }
    }
    get ariaLabel() {
        return this._ariaLabel;
    }
    ngAfterViewInit() {
        if (!this.control) {
            console.warn('You need to pass an input as a control to the password toggle.');
            return;
        }
        // show the right icon according to the initial type of the input
        this._currentIcon = this.control.type === 'password' ? visibilityIcons['show'] : visibilityIcons['hide'];
    }
    ngOnDestroy() {
        this._focusMonitor.stopMonitoring(this._elementRef);
    }
    /** Toggles the type of the input. */
    toggleInputType() {
        if (this.control) {
            this.control.type = this.control.type === 'password' ? 'text' : 'password';
            this._pressed = !this._pressed;
            this.toggleIcon();
            this._cdr.markForCheck();
        }
    }
    /** @docs-private */
    toggleIcon() {
        this._currentIcon = this._currentIcon === visibilityIcons['show'] ? visibilityIcons['hide'] : visibilityIcons['show'];
    }
    /** @docs-private */
    _onKeydown($event) {
        if ($event && ($event.keyCode === ENTER || $event.keyCode === SPACE)) {
            this.toggleInputType();
        }
    }
    /** @docs-private */
    get tabindex() {
        if (this.control) {
            return this.control.disabled ? -1 : 0;
        }
        return null;
    }
}
NxPasswordToggleComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPasswordToggleComponent, deps: [{ token: i0.ChangeDetectorRef }, { token: i0.ElementRef }, { token: i1.FocusMonitor }], target: i0.ɵɵFactoryTarget.Component });
NxPasswordToggleComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxPasswordToggleComponent, selector: "nx-password-toggle", inputs: { control: "control", ariaLabel: "ariaLabel" }, host: { attributes: { "role": "button" }, listeners: { "click": "toggleInputType()", "keydown": "_onKeydown($event)" }, properties: { "attr.aria-label": "ariaLabel", "attr.tabindex": "tabindex", "attr.aria-pressed": "_pressed" } }, ngImport: i0, template: `<nx-icon aria-hidden="true" [name]="_currentIcon"> </nx-icon>`, isInline: true, styles: [":host{position:relative;display:inline-block;height:24px;width:24px;cursor:pointer}@media screen and (-ms-high-contrast: active){:host{color:buttonText!important}}:host:focus{outline:none}:host:hover{color:var(--hover-primary)}:host.cdk-keyboard-focused ::ng-deep nx-icon{box-shadow:var(--focus-box-shadow);border-radius:4px}@media screen and (-ms-high-contrast: active),(forced-colors: active){:host.cdk-keyboard-focused ::ng-deep nx-icon{box-shadow:0 0 0 2px background,0 0 0 6px windowText;outline:4px solid CanvasText;outline-offset:2px}}:host-context(.is-disabled){pointer-events:none}@media screen and (-ms-high-contrast: active){:host-context(.is-disabled){color:GrayText!important}}\n"], components: [{ type: i2.NxIconComponent, selector: "nx-icon", inputs: ["name", "outline", "fill", "size", "font"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPasswordToggleComponent, decorators: [{
            type: Component,
            args: [{ selector: 'nx-password-toggle', template: `<nx-icon aria-hidden="true" [name]="_currentIcon"> </nx-icon>`, host: {
                        '[attr.aria-label]': 'ariaLabel',
                        '[attr.tabindex]': 'tabindex',
                        '[attr.aria-pressed]': '_pressed',
                        role: 'button',
                        '(click)': 'toggleInputType()',
                        '(keydown)': '_onKeydown($event)',
                    }, changeDetection: ChangeDetectionStrategy.OnPush, styles: [":host{position:relative;display:inline-block;height:24px;width:24px;cursor:pointer}@media screen and (-ms-high-contrast: active){:host{color:buttonText!important}}:host:focus{outline:none}:host:hover{color:var(--hover-primary)}:host.cdk-keyboard-focused ::ng-deep nx-icon{box-shadow:var(--focus-box-shadow);border-radius:4px}@media screen and (-ms-high-contrast: active),(forced-colors: active){:host.cdk-keyboard-focused ::ng-deep nx-icon{box-shadow:0 0 0 2px background,0 0 0 6px windowText;outline:4px solid CanvasText;outline-offset:2px}}:host-context(.is-disabled){pointer-events:none}@media screen and (-ms-high-contrast: active){:host-context(.is-disabled){color:GrayText!important}}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.ChangeDetectorRef }, { type: i0.ElementRef }, { type: i1.FocusMonitor }]; }, propDecorators: { control: [{
                type: Input
            }], ariaLabel: [{
                type: Input
            }] } });
