// @aposin/ng-aquila/esm2020/sidebar/sidebar.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxButtonModule } from '@aposin/ng-aquila/button';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxSidebarComponent } from './sidebar.component';
import { NxSidebarFooterComponent } from './sidebar-footer';
import { NxSidebarToggleComponent } from './sidebar-toggle';
import * as i0 from "@angular/core";
export class NxSidebarModule {
}
NxSidebarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSidebarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxSidebarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSidebarModule, declarations: [NxSidebarComponent, NxSidebarFooterComponent, NxSidebarToggleComponent], imports: [CommonModule, NxIconModule, NxButtonModule], exports: [NxSidebarComponent, NxSidebarFooterComponent, NxSidebarToggleComponent] });
NxSidebarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSidebarModule, imports: [[CommonModule, NxIconModule, NxButtonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSidebarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, NxIconModule, NxButtonModule],
                    declarations: [NxSidebarComponent, NxSidebarFooterComponent, NxSidebarToggleComponent],
                    exports: [NxSidebarComponent, NxSidebarFooterComponent, NxSidebarToggleComponent],
                }]
        }] });
