// @aposin/ng-aquila/esm2020/table/table-cell.component.mjs
import { ChangeDetectionStrategy, Component } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * This is a table cell.
 */
export class NxTableCellComponent {
}
NxTableCellComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTableCellComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
NxTableCellComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.1", type: NxTableCellComponent, selector: "td[nxTableCell], td[nxCell]", host: { properties: { "class.nx-table-cell": "true" } }, ngImport: i0, template: '<ng-content></ng-content>', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxTableCellComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'td[nxTableCell], td[nxCell]',
                    template: '<ng-content></ng-content>',
                    host: {
                        '[class.nx-table-cell]': 'true',
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }] });
