// @aposin/ng-aquila/esm2020/datefield/datefield.module.mjs
import { A11yModule } from '@angular/cdk/a11y';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NX_DATE_LOCALE_PROVIDER } from './adapter/index';
import { NxDatefieldDirective } from './datefield.directive';
import { NxCalendarComponent } from './datepicker/calendar';
import { NxCalendarBodyComponent } from './datepicker/calendar-body';
import { NX_DATEPICKER_SCROLL_STRATEGY_PROVIDER, NxDatepickerComponent, NxDatepickerContentComponent } from './datepicker/datepicker.component';
import { NxDatepickerIntl } from './datepicker/datepicker-intl';
import { NxDatepickerToggleComponent, NxDatepickerToggleIconComponent } from './datepicker/datepicker-toggle';
import { NxMonthViewComponent } from './datepicker/month-view';
import { NxMultiYearViewComponent } from './datepicker/multi-year-view';
import { NxYearViewComponent } from './datepicker/year-view';
import * as i0 from "@angular/core";
export class NxDatefieldModule {
}
NxDatefieldModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxDatefieldModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxDatefieldModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxDatefieldModule, declarations: [NxDatefieldDirective,
        NxDatepickerComponent,
        NxDatepickerToggleIconComponent,
        NxDatepickerToggleComponent,
        NxDatepickerContentComponent,
        NxCalendarComponent,
        NxCalendarBodyComponent,
        NxMonthViewComponent,
        NxMultiYearViewComponent,
        NxYearViewComponent], imports: [CommonModule, OverlayModule, A11yModule, NxIconModule], exports: [NxDatefieldDirective,
        NxDatepickerComponent,
        NxDatepickerToggleIconComponent,
        NxDatepickerToggleComponent,
        NxCalendarComponent,
        NxCalendarBodyComponent,
        NxMonthViewComponent,
        NxMultiYearViewComponent,
        NxYearViewComponent] });
NxDatefieldModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxDatefieldModule, providers: [
        // Provide our NX_DATE_LOCALE to determine the current language
        NX_DATE_LOCALE_PROVIDER,
        NxDatepickerIntl,
        NX_DATEPICKER_SCROLL_STRATEGY_PROVIDER,
    ], imports: [[CommonModule, OverlayModule, A11yModule, NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxDatefieldModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, OverlayModule, A11yModule, NxIconModule],
                    providers: [
                        // Provide our NX_DATE_LOCALE to determine the current language
                        NX_DATE_LOCALE_PROVIDER,
                        NxDatepickerIntl,
                        NX_DATEPICKER_SCROLL_STRATEGY_PROVIDER,
                    ],
                    declarations: [
                        NxDatefieldDirective,
                        NxDatepickerComponent,
                        NxDatepickerToggleIconComponent,
                        NxDatepickerToggleComponent,
                        NxDatepickerContentComponent,
                        NxCalendarComponent,
                        NxCalendarBodyComponent,
                        NxMonthViewComponent,
                        NxMultiYearViewComponent,
                        NxYearViewComponent,
                    ],
                    exports: [
                        NxDatefieldDirective,
                        NxDatepickerComponent,
                        NxDatepickerToggleIconComponent,
                        NxDatepickerToggleComponent,
                        NxCalendarComponent,
                        NxCalendarBodyComponent,
                        NxMonthViewComponent,
                        NxMultiYearViewComponent,
                        NxYearViewComponent,
                    ],
                }]
        }] });
