// @aposin/ng-aquila/esm2020/datefield/date-validators.mjs
/**
 * Provides a set of date validators that can be used by form controls.
 * @dynamic
 */
export class NxDateValidators {
    /** The form control validator for whether the input parses. */
    static parse(dateAdapter, dateFormats, input, strict, customParseFormat) {
        return () => {
            if (!input.value) {
                return null;
            }
            const parsedValue = dateAdapter.parse(input.value, customParseFormat || dateFormats.parse.dateInput, strict);
            const valid = parsedValue && dateAdapter.isValid(parsedValue);
            return valid ? null : { nxDatefieldParse: { text: input } };
        };
    }
    /** The form control validator for the min date. */
    static min(dateAdapter, min) {
        return (control) => {
            const controlValue = getValidDateOrNull(dateAdapter, dateAdapter.deserialize(control.value));
            return !min || !controlValue || dateAdapter.compareDate(min, controlValue) <= 0 ? null : { nxDatefieldMin: { min, actual: controlValue } };
        };
    }
    /** The form control validator for the max date. */
    static max(dateAdapter, max) {
        return (control) => {
            const controlValue = getValidDateOrNull(dateAdapter, dateAdapter.deserialize(control.value));
            return !max || !controlValue || dateAdapter.compareDate(max, controlValue) >= 0 ? null : { nxDatefieldMax: { max, actual: controlValue } };
        };
    }
    /** The form control validator for the date filter. */
    static filter(dateAdapter, dateFilter) {
        return (control) => {
            const controlValue = getValidDateOrNull(dateAdapter, dateAdapter.deserialize(control.value));
            return !dateFilter || !controlValue || dateFilter(controlValue) ? null : { nxDatefieldFilter: true };
        };
    }
}
function getValidDateOrNull(dateAdapter, obj) {
    return dateAdapter.isDateInstance(obj) && dateAdapter.isValid(obj) ? obj : null;
}
