// @aposin/ng-aquila/esm2020/autocomplete/autocomplete.module.mjs
import { A11yModule } from '@angular/cdk/a11y';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxAutocompleteComponent } from './autocomplete.component';
import { NxAutocompleteOptionComponent } from './autocomplete-option.component';
import { NX_AUTOCOMPLETE_SCROLL_STRATEGY_PROVIDER, NxAutocompleteTriggerDirective } from './autocomplete-trigger.directive';
import * as i0 from "@angular/core";
export class NxAutocompleteModule {
}
NxAutocompleteModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAutocompleteModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxAutocompleteModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAutocompleteModule, declarations: [NxAutocompleteComponent, NxAutocompleteOptionComponent, NxAutocompleteTriggerDirective], imports: [CommonModule, OverlayModule, A11yModule], exports: [NxAutocompleteComponent, NxAutocompleteOptionComponent, NxAutocompleteTriggerDirective] });
NxAutocompleteModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAutocompleteModule, providers: [NX_AUTOCOMPLETE_SCROLL_STRATEGY_PROVIDER], imports: [[CommonModule, OverlayModule, A11yModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxAutocompleteModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxAutocompleteComponent, NxAutocompleteOptionComponent, NxAutocompleteTriggerDirective],
                    exports: [NxAutocompleteComponent, NxAutocompleteOptionComponent, NxAutocompleteTriggerDirective],
                    imports: [CommonModule, OverlayModule, A11yModule],
                    providers: [NX_AUTOCOMPLETE_SCROLL_STRATEGY_PROVIDER],
                }]
        }] });
