// @aposin/ng-aquila/esm2020/radio-toggle/radio-toggle-button-base.component.mjs
import { Directive, Input, ViewChild } from '@angular/core';
import { removeClasses } from '@aposin/ng-aquila/utils';
import * as i0 from "@angular/core";
let nextId = 0;
/** @docs-private */
export class NxRadioToggleButtonBaseComponent {
    constructor(/** @docs-private */ renderer) {
        this.renderer = renderer;
        /** @docs-private */
        this.toggleButtonId = 'nx-radio-toggle-button-' + nextId++;
    }
    /** @docs-private */
    triggerInputClick(event) {
        event.stopPropagation();
        this.toggleInput.nativeElement.click();
    }
    /** @docs-private */
    setFirstButton() {
        this.renderer.addClass(this.labelContainer.nativeElement, 'nx-radio-toggle__label-container--first');
    }
    /** @docs-private */
    setLastButton() {
        this.renderer.addClass(this.labelContainer.nativeElement, 'nx-radio-toggle__label-container--last');
    }
    /** @docs-private */
    resetClasses() {
        removeClasses(this.renderer, this.labelContainer, 'nx-radio-toggle__label-container--first nx-radio-toggle__label-container--last');
    }
}
NxRadioToggleButtonBaseComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRadioToggleButtonBaseComponent, deps: [{ token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Directive });
NxRadioToggleButtonBaseComponent.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxRadioToggleButtonBaseComponent, inputs: { value: ["nxValue", "value"] }, viewQueries: [{ propertyName: "labelContainer", first: true, predicate: ["toggleButtonLabelContainer"], descendants: true, static: true }, { propertyName: "toggleInput", first: true, predicate: ["toggleInput"], descendants: true, static: true }], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRadioToggleButtonBaseComponent, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i0.Renderer2 }]; }, propDecorators: { value: [{
                type: Input,
                args: ['nxValue']
            }], labelContainer: [{
                type: ViewChild,
                args: ['toggleButtonLabelContainer', { static: true }]
            }], toggleInput: [{
                type: ViewChild,
                args: ['toggleInput', { static: true }]
            }] } });
