// @aposin/ng-aquila/esm2020/radio-toggle/radio-toggle.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxRadioToggleComponent } from './radio-toggle.component';
import { NxRadioToggleButtonComponent } from './radio-toggle-button.component';
import * as i0 from "@angular/core";
export class NxRadioToggleModule {
}
NxRadioToggleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRadioToggleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxRadioToggleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRadioToggleModule, declarations: [NxRadioToggleComponent, NxRadioToggleButtonComponent], imports: [CommonModule, FormsModule, NxIconModule], exports: [NxRadioToggleComponent, NxRadioToggleButtonComponent] });
NxRadioToggleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRadioToggleModule, imports: [[CommonModule, FormsModule, NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxRadioToggleModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxRadioToggleComponent, NxRadioToggleButtonComponent],
                    exports: [NxRadioToggleComponent, NxRadioToggleButtonComponent],
                    imports: [CommonModule, FormsModule, NxIconModule],
                }]
        }] });
