// @aposin/ng-aquila/esm2020/iso-date-adapter/adapter/iso-date-adapter.module.mjs
import { NgModule } from '@angular/core';
import { NX_DATE_FORMATS, NX_DATE_LOCALE, NX_DATE_LOCALE_PROVIDER, NxDateAdapter } from '@aposin/ng-aquila/datefield';
import { NxIsoDateAdapter } from './iso-date-adapter';
import { NX_ISO_DATE_FORMATS } from './iso-date-formats';
import * as i0 from "@angular/core";
export class IsoDateModule {
}
IsoDateModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: IsoDateModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
IsoDateModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: IsoDateModule });
IsoDateModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: IsoDateModule, providers: [NX_DATE_LOCALE_PROVIDER, { provide: NxDateAdapter, useClass: NxIsoDateAdapter, deps: [NX_DATE_LOCALE] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: IsoDateModule, decorators: [{
            type: NgModule,
            args: [{
                    providers: [NX_DATE_LOCALE_PROVIDER, { provide: NxDateAdapter, useClass: NxIsoDateAdapter, deps: [NX_DATE_LOCALE] }],
                }]
        }] });
export class NxIsoDateModule {
}
NxIsoDateModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxIsoDateModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxIsoDateModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxIsoDateModule, imports: [IsoDateModule] });
NxIsoDateModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxIsoDateModule, providers: [{ provide: NX_DATE_FORMATS, useValue: NX_ISO_DATE_FORMATS }], imports: [[IsoDateModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxIsoDateModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [IsoDateModule],
                    providers: [{ provide: NX_DATE_FORMATS, useValue: NX_ISO_DATE_FORMATS }],
                }]
        }] });
