// @aposin/ng-aquila/esm2020/swipebar/swipebar.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxCopytextModule } from '@aposin/ng-aquila/copytext';
import { NxSwipebarComponent } from './swipebar.component';
import * as i0 from "@angular/core";
export class NxSwipebarModule {
}
NxSwipebarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSwipebarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxSwipebarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSwipebarModule, declarations: [NxSwipebarComponent], imports: [CommonModule, NxCopytextModule], exports: [NxSwipebarComponent] });
NxSwipebarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSwipebarModule, imports: [[CommonModule, NxCopytextModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSwipebarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NxSwipebarComponent],
                    exports: [NxSwipebarComponent],
                    imports: [CommonModule, NxCopytextModule],
                }]
        }] });
