// @aposin/ng-aquila/esm2020/phone-input/phone-input-intl.mjs
import { Injectable } from '@angular/core';
import countries from 'i18n-iso-countries';
import en from 'i18n-iso-countries/langs/en.json';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
countries.registerLocale(en);
export class NxPhoneInputIntl {
    constructor() {
        /**
         * Stream that emits whenever the labels here are changed. Use this to notify
         * components if the properties have changed after initialization.
         */
        this.changes = new Subject();
        /** The label that is shown at the top of the opened overlay. */
        this.areaCodeLabel = 'Country code';
        /** The object providing the country name translations. */
        this.countryNames = countries.getNames('en');
    }
}
NxPhoneInputIntl.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPhoneInputIntl, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
NxPhoneInputIntl.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPhoneInputIntl, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPhoneInputIntl, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
