// @aposin/ng-aquila/esm2020/phone-input/phone-input.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { NxDropdownModule } from '@aposin/ng-aquila/dropdown';
import { NxFormfieldModule } from '@aposin/ng-aquila/formfield';
import { NxInputModule } from '@aposin/ng-aquila/input';
import { NxPhoneInputComponent } from './phone-input.component';
import * as i0 from "@angular/core";
export class NxPhoneInputModule {
}
NxPhoneInputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPhoneInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxPhoneInputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPhoneInputModule, declarations: [NxPhoneInputComponent], imports: [NxDropdownModule, CommonModule, FormsModule, NxInputModule], exports: [NxPhoneInputComponent, NxFormfieldModule] });
NxPhoneInputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPhoneInputModule, providers: [], imports: [[NxDropdownModule, CommonModule, FormsModule, NxInputModule], NxFormfieldModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPhoneInputModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [NxDropdownModule, CommonModule, FormsModule, NxInputModule],
                    exports: [NxPhoneInputComponent, NxFormfieldModule],
                    declarations: [NxPhoneInputComponent],
                    providers: [],
                }]
        }] });
