// @aposin/ng-aquila/esm2020/notification-panel/notification-item/notification-item-slots.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class NxNotificationItemMetadataDirective {
}
NxNotificationItemMetadataDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationItemMetadataDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxNotificationItemMetadataDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxNotificationItemMetadataDirective, selector: "nx-notification-item-metadata", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationItemMetadataDirective, decorators: [{
            type: Directive,
            args: [{ selector: 'nx-notification-item-metadata' }]
        }] });
export class NxNotificationItemContentDirective {
}
NxNotificationItemContentDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationItemContentDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxNotificationItemContentDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxNotificationItemContentDirective, selector: "nx-notification-item-content", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationItemContentDirective, decorators: [{
            type: Directive,
            args: [{ selector: 'nx-notification-item-content' }]
        }] });
export class NxNotificationItemActionsDirective {
}
NxNotificationItemActionsDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationItemActionsDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxNotificationItemActionsDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxNotificationItemActionsDirective, selector: "nx-notification-item-actions", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationItemActionsDirective, decorators: [{
            type: Directive,
            args: [{ selector: 'nx-notification-item-actions' }]
        }] });
export class NxNotificationItemHeaderDirective {
}
NxNotificationItemHeaderDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationItemHeaderDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
NxNotificationItemHeaderDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.1", type: NxNotificationItemHeaderDirective, selector: "nx-notification-header", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxNotificationItemHeaderDirective, decorators: [{
            type: Directive,
            args: [{ selector: 'nx-notification-header' }]
        }] });
