// @aposin/ng-aquila/esm2020/slider/slider.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxSliderComponent } from './slider.component';
import * as i0 from "@angular/core";
export class NxSliderModule {
}
NxSliderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSliderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxSliderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSliderModule, declarations: [NxSliderComponent], imports: [CommonModule], exports: [NxSliderComponent] });
NxSliderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSliderModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxSliderModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [NxSliderComponent],
                    exports: [NxSliderComponent],
                }]
        }] });
