// @aposin/ng-aquila/esm2020/popover/popover.module.mjs
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { NxPopoverComponent } from './popover.component';
import { NxPopoverContentDirective } from './popover-content';
import { NxPopoverIntl } from './popover-intl';
import { NX_POPOVER_SCROLL_STRATEGY_PROVIDER, NxPopoverTriggerDirective } from './popover-trigger.directive';
import * as i0 from "@angular/core";
export class NxPopoverModule {
}
NxPopoverModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPopoverModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NxPopoverModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPopoverModule, declarations: [NxPopoverTriggerDirective, NxPopoverComponent, NxPopoverContentDirective], imports: [CommonModule, OverlayModule, NxIconModule], exports: [NxPopoverTriggerDirective, NxPopoverComponent, NxPopoverContentDirective] });
NxPopoverModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPopoverModule, providers: [NxPopoverIntl, NX_POPOVER_SCROLL_STRATEGY_PROVIDER], imports: [[CommonModule, OverlayModule, NxIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: NxPopoverModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, OverlayModule, NxIconModule],
                    declarations: [NxPopoverTriggerDirective, NxPopoverComponent, NxPopoverContentDirective],
                    exports: [NxPopoverTriggerDirective, NxPopoverComponent, NxPopoverContentDirective],
                    providers: [NxPopoverIntl, NX_POPOVER_SCROLL_STRATEGY_PROVIDER],
                }]
        }] });
